// mtk-ui/lib/navigationBar/NavigationItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationItem = exports.container = undefined;

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _compose = require('recompose/compose');

var _compose2 = _interopRequireDefault(_compose);

var _getContext = require('recompose/getContext');

var _getContext2 = _interopRequireDefault(_getContext);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _defaultProps = require('recompose/defaultProps');

var _defaultProps2 = _interopRequireDefault(_defaultProps);

var _mapProps = require('recompose/mapProps');

var _mapProps2 = _interopRequireDefault(_mapProps);

var _setPropTypes = require('recompose/setPropTypes');

var _setPropTypes2 = _interopRequireDefault(_setPropTypes);

var _NavigationItem = require('./styles/NavigationItem.css');

var _NavigationItem2 = _interopRequireDefault(_NavigationItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var container = exports.container = (0, _compose2.default)(_pure2.default, (0, _getContext2.default)({
  selectedValue: _react.PropTypes.any,
  onSelect: _react.PropTypes.func
}), (0, _defaultProps2.default)({
  onClick: _emptyFunction2.default,
  onSelect: _emptyFunction2.default
}), (0, _mapProps2.default)(function (props) {
  return (0, _extends3.default)({}, props, {
    onClick: function onClick(e) {
      props.onClick(e, props.value);
      props.onSelect(e, props.value);
    }
  });
}));

var NavigationItem = exports.NavigationItem = (0, _setPropTypes2.default)({
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  tagName: _react.PropTypes.string,
  active: _react.PropTypes.bool,
  value: _react.PropTypes.any,
  selectedValue: _react.PropTypes.any,
  onClick: _react.PropTypes.func.isRequired
})(function (_ref) {
  var className = _ref.className,
      _ref$tagName = _ref.tagName,
      tagName = _ref$tagName === undefined ? 'div' : _ref$tagName,
      active = _ref.active,
      value = _ref.value,
      selectedValue = _ref.selectedValue,
      onClick = _ref.onClick,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['className', 'tagName', 'active', 'value', 'selectedValue', 'onClick']);

  var isActive = active || value === selectedValue;

  return (0, _react.createElement)(tagName, (0, _extends3.default)({}, otherProps, {
    onClick: onClick,
    className: (0, _classnames2.default)(isActive ? _NavigationItem2.default.active : _NavigationItem2.default.default, className)
  }));
});

exports.default = container(NavigationItem);
