// mtk-ui/lib/navigationBar/NavigationBar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _withContext = require('recompose/withContext');

var _withContext2 = _interopRequireDefault(_withContext);

var _NavigationBar = require('./styles/NavigationBar.css');

var _NavigationBar2 = _interopRequireDefault(_NavigationBar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NavigationBar = (_dec = (0, _withContext2.default)({
  selectedValue: _react.PropTypes.any,
  onSelect: _react.PropTypes.func
}, function (props) {
  return {
    selectedValue: props.value,
    onSelect: props.onSelect
  };
}), _dec(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(NavigationBar, _React$Component);

  function NavigationBar() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, NavigationBar);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = NavigationBar.__proto__ || Object.getPrototypeOf(NavigationBar)).call.apply(_ref, [this].concat(args))), _this), _this.defaultProps = {
      onSelect: _emptyFunction2.default
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(NavigationBar, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          value = _props.value,
          onSelect = _props.onSelect,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'className', 'value', 'onSelect']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, otherProps, { className: (0, _classnames2.default)(_NavigationBar2.default.default, className) }),
        children
      );
    }
  }]);
  return NavigationBar;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  value: _react.PropTypes.any,
  onSelect: _react.PropTypes.func }, _temp2)) || _class);
exports.default = NavigationBar;
module.exports = exports['default'];
