// @mtk/mcs-components/lib/datachannels/integer/ControlCard.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _commonCard = require('../common/card/');

var _commonCard2 = _interopRequireDefault(_commonCard);

var _commonDatachannelInfo = require('../common/datachannelInfo');

var _commonDatachannelInfo2 = _interopRequireDefault(_commonDatachannelInfo);

var _floatControl = require('../float/Control');

var _floatControl2 = _interopRequireDefault(_floatControl);

var displayCard = function displayCard(_ref) {
  var numberOfCards = _ref.numberOfCards;
  var name = _ref.name;
  var description = _ref.description;
  var dataChnId = _ref.dataChnId;
  var configs = _ref.configs;
  var displayType = _ref.displayType;
  var fullScreenConfig = _ref.fullScreenConfig;
  var detailButtonConfig = _ref.detailButtonConfig;
  var recordedAt = _ref.recordedAt;

  var props = _objectWithoutProperties(_ref, ['numberOfCards', 'name', 'description', 'dataChnId', 'configs', 'displayType', 'fullScreenConfig', 'detailButtonConfig', 'recordedAt']);

  return _react2['default'].createElement(
    _commonCard2['default'],
    {
      fullScreenConfig: fullScreenConfig,
      detailButtonConfig: detailButtonConfig,
      displayType: displayType,
      configs: configs,
      numberOfCards: numberOfCards },
    _react2['default'].createElement(
      'div',
      null,
      _react2['default'].createElement(_floatControl2['default'], _extends({ displayType: displayType }, props)),
      _react2['default'].createElement(_commonDatachannelInfo2['default'], {
        name: name,
        dataChnId: dataChnId,
        description: description,
        recordedAt: recordedAt
      })
    )
  );
};

exports['default'] = (0, _recompose.compose)((0, _recompose.defaultProps)({
  numberOfCards: 2,
  displayType: 'none'
}), (0, _recompose.setPropTypes)(_extends({
  configs: _react.PropTypes.object.isRequired,
  numberOfCards: _react.PropTypes.number,
  name: _react.PropTypes.string.isRequired,
  description: _react.PropTypes.string.isRequired,
  fullScreenConfig: _react.PropTypes.object,
  recordedAt: _react.PropTypes.number,
  detailButtonConfig: _react.PropTypes.object,
  values: _react.PropTypes.object
}, _floatControl.propTypes)), _recompose.pure, (0, _recompose.renameProp)('values', 'status'))(displayCard);
module.exports = exports['default'];
