// @mtk/mcs-components/lib/datachannels/common/datachannelInfo/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _stylesMainCss = require('./styles/main.css');

var _stylesMainCss2 = _interopRequireDefault(_stylesMainCss);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var datachannelInfo = function datachannelInfo(_ref) {
  var name = _ref.name;
  var description = _ref.description;
  var recordedAt = _ref.recordedAt;
  var dataChnId = _ref.dataChnId;

  return _react2['default'].createElement(
    'div',
    { className: _stylesMainCss2['default'].base },
    _react2['default'].createElement(
      'h3',
      { className: _stylesMainCss2['default'].deviceName },
      name
    ),
    recordedAt ? _react2['default'].createElement(
      'p',
      { className: _stylesMainCss2['default'].lastDatapointTime },
      'Last data point time: ',
      (0, _moment2['default'])(recordedAt).format('YYYY-MM-DD HH:mm')
    ) : _react2['default'].createElement(
      'p',
      { className: _stylesMainCss2['default'].lastDatapointTime },
      'Data channel Id: ',
      dataChnId
    ),
    _react2['default'].createElement('hr', null),
    _react2['default'].createElement(
      'article',
      { className: _stylesMainCss2['default'].description },
      _react2['default'].createElement(
        'span',
        null,
        description
      )
    )
  );
};

exports['default'] = (0, _recompose.compose)((0, _recompose.defaultProps)({
  active: false,
  numberOfCards: 2,
  description: 'This is a description'
}), (0, _recompose.setPropTypes)({
  recordedAt: _react.PropTypes.number,
  dataChnId: _react.PropTypes.string.isRequired,
  numberOfCards: _react.PropTypes.number.isRequired,
  description: _react.PropTypes.string,
  name: _react.PropTypes.string.isRequired
}), _recompose.pure)(datachannelInfo);
module.exports = exports['default'];
