// @mtk/mcs-components/lib/datachannels/analog/Control.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _mtkUiLibInputRange = require('mtk-ui/lib/InputRange');

var _mtkUiLibInputRange2 = _interopRequireDefault(_mtkUiLibInputRange);

var _commonCardStylesBaseCss = require('../common/card/styles/base.css');

var _commonCardStylesBaseCss2 = _interopRequireDefault(_commonCardStylesBaseCss);

var _stylesMainCss = require('./styles/main.css');

var _stylesMainCss2 = _interopRequireDefault(_stylesMainCss);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _fbjsLibEmptyFunction = require('fbjs/lib/emptyFunction');

var _fbjsLibEmptyFunction2 = _interopRequireDefault(_fbjsLibEmptyFunction);

var control = function control(_ref) {
  var lowerbound = _ref.lowerbound;
  var upperbound = _ref.upperbound;
  var onChange = _ref.onChange;
  var value = _ref.value;
  var displayType = _ref.displayType;
  return _react2['default'].createElement(
    'div',
    { className: (0, _classnames2['default'])(_commonCardStylesBaseCss2['default'].base, _commonCardStylesBaseCss2['default'][displayType], _stylesMainCss2['default'].base) },
    _react2['default'].createElement(
      'div',
      { className: _stylesMainCss2['default'].currentValue },
      ' Current Value: ',
      _react2['default'].createElement(
        'span',
        { style: { color: '#00A1DE' } },
        value
      )
    ),
    _react2['default'].createElement(_mtkUiLibInputRange2['default'], {
      max: lowerbound,
      min: upperbound,
      value: value,
      onChange: onChange,
      style: { width: 240 }
    }),
    _react2['default'].createElement(
      'div',
      { className: _stylesMainCss2['default'].limit },
      _react2['default'].createElement(
        'span',
        null,
        lowerbound
      ),
      _react2['default'].createElement(
        'span',
        null,
        upperbound
      )
    )
  );
};

var propTypes = {
  status: _react.PropTypes.object,
  value: _react.PropTypes.number,
  displayType: _react.PropTypes.oneOf(['none', 'preview', 'embeded']),
  onChange: _react.PropTypes.func,
  lowerbound: _react.PropTypes.number.isRequired,
  upperbound: _react.PropTypes.number.isRequired
};

exports.propTypes = propTypes;
exports['default'] = (0, _recompose.compose)((0, _recompose.defaultProps)({
  status: { value: 0 },
  displayType: 'none',
  onChange: _fbjsLibEmptyFunction2['default']
}), (0, _recompose.setPropTypes)(propTypes), _recompose.pure)(control);
