// @mtk/mcs-components/src/datachannels/common/datachannelInfo/index.js
import { default as React, PropTypes } from 'react';
import { compose, pure, setPropTypes, defaultProps } from 'recompose';
import { default as styles } from './styles/main.css';
import { default as moment } from 'moment';

const datachannelInfo = ({ name, description, recordedAt, dataChnId }) => {
  return (
    <div className={styles.base} >
      <h3 className={styles.deviceName}>{name}</h3>
      {
        recordedAt ?
        <p className={styles.lastDatapointTime}>Last data point time: {moment(recordedAt).format('YYYY-MM-DD HH:mm')}</p> :
        <p className={styles.lastDatapointTime}>Data channel Id: {dataChnId}</p>
      }
      <hr />
      <article className={styles.description}><span>{description}</span></article>
    </div>
  );
};

export default compose(
  defaultProps({
    active: false,
    numberOfCards: 2,
    description: 'This is a description',
  }),
  setPropTypes({
    recordedAt: PropTypes.number,
    dataChnId: PropTypes.string.isRequired,
    numberOfCards: PropTypes.number.isRequired,
    description: PropTypes.string,
    name: PropTypes.string.isRequired,
  }),
  pure,
)(datachannelInfo);
