// mtk-ui/lib/Transitions/SlideDown.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _reactOverlays = require('react-overlays');

var _SlideDown = require('../styles/SlideDown.css');

var _SlideDown2 = _interopRequireDefault(_SlideDown);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SlideDown = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(SlideDown, _React$Component);

  function SlideDown() {
    (0, _classCallCheck3.default)(this, SlideDown);
    return (0, _possibleConstructorReturn3.default)(this, (SlideDown.__proto__ || Object.getPrototypeOf(SlideDown)).apply(this, arguments));
  }

  (0, _createClass3.default)(SlideDown, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children']);


      return _react2.default.createElement(
        _reactOverlays.Transition,
        (0, _extends3.default)({}, otherProps, {
          timeout: 100,
          className: _SlideDown2.default.slide,
          enteredClassName: _SlideDown2.default.in,
          enteringClassName: _SlideDown2.default.in
        }),
        children
      );
    }
  }]);
  return SlideDown;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node
}, _temp)) || _class;

exports.default = SlideDown;
module.exports = exports['default'];
