// @mtk/mcs-components/src/datachannels/common/card/index.js
import { default as React, PropTypes } from 'react';
import { withState, mapProps, compose, pure, setPropTypes, defaultProps } from 'recompose';
import { default as DetailButton } from '../detailButton';
import { default as Card } from 'mtk-ui/lib/Card';
import { default as FullScreenButton } from '../fullScreenButton';
import { default as styles } from './styles/main.css';

const displayCard = ({
  children,
  fullScreenConfig,
  detailButtonConfig,
  block,
}) => (
  <Card
    className={styles.base}
    style={{ width: 280 * block }}
  >
    <div className={styles.info}>
      {fullScreenConfig && <FullScreenButton config={fullScreenConfig} />}
      {detailButtonConfig && <DetailButton config={detailButtonConfig} />}
    </div>
    <div style={{ marginTop: -36 }}>
      {children}
    </div>
  </Card>
);

export default compose(
  defaultProps({
    numberOfCards: 2,
  }),
  withState('block', 'setBlock', (props) =>
    props.numberOfCards > props.configs.card ? props.configs.card : props.numberOfCards),
  setPropTypes({
    block: PropTypes.number.isRequired,
    numberOfCards: PropTypes.number,
    children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    infoConfig: PropTypes.object,
    fullScreenConfig: PropTypes.object,
    detailButtonConfig: PropTypes.object,
    configs: PropTypes.object,
  }),
  mapProps( props => ({
    ...props,
    block: (props.numberOfCards > props.configs.card) ? props.configs.card : props.numberOfCards,
  })),
  pure,
)(displayCard);
