// @mtk/mcs-components/src/ButtonGroup.jsx
import React from 'react';
import Radium from 'radium';
import scrollTo from './utils/scrollTo.js';
import Button from './Button';

class ButtonGroup extends React.Component {
  constructor(props) {
    super(props);
    this.state = this.state || {};
    this.state.sectionList = [];
  }

  render() {
    let {
      style: customizedStyle,
      kind,
      leftTitle,
      rightTitle,
      ...otherProps
    } = this.props;

    return (
      <div {...otherProps} style={customizedStyle}>
        <Button kind={kind}
          style={styles.buttonLeft}>
          {leftTitle}
        </Button>
        <Button kind={kind}
          style={styles.buttonRight}>
          {rightTitle}
        </Button>
      </div>
    );
  }
}

ButtonGroup.propTypes = {
  style: React.PropTypes.object,
  kind: React.PropTypes.oneOf(['primary', 'cancel' /*, 'warning', 'danger', 'primary' */]).isRequired,
  leftTitle: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.object
    ]),
  rightTitle: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.object
    ])
};

ButtonGroup.defaultProps = {
  kind: 'primary',
  leftTitle: 'Left Title',
  rightTitle: 'Right Title'
};

export default Radium(ButtonGroup);

const styles = {
  buttonLeft:{
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0
  },
  buttonRight:{
    minWidth: 0,
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
    borderLeft: '1px solid #1fadea'
  }
};
