// @mtk/mcs-components/src/Button.jsx
import React from 'react';
import Radium from 'radium';

class Button extends React.Component {

  render() {
    const {
      style: customizedStyle,
      ...otherProps
    } = this.props;

    return (
      <button
        {...otherProps}
        style={[
          styles.base,
          styles[this.props.kind],
          customizedStyle
        ]}>
        {this.props.children}
      </button>
    );
  }

}

Button.propTypes = {
  kind: React.PropTypes.oneOf(['primary', 'cancel' /*, 'warning', 'danger', 'primary' */]).isRequired
};

Button.defaultProps = {
  kind: 'primary'
};

export default Radium(Button);

const styles = {
  base: {
    boxSizing: 'border-box',
    display: 'inline-block',
    padding: '6px 12px',
    textAlign: 'center',
    lineHeight: 1,
    whiteSpace: 'nowrap',
    verticalAlign: 'middle',
    touchAction: 'manipulation',
    border: 'none',
    borderRadius: 4,
    color: '#fafafa',
    cursor: 'pointer',
    fontSize: 13,
    minWidth: 80,
    height: 32,
    fontFamily: 'RionaSans-Regular',
    ':focus': {
      outline: 'none'
    },
    transition: 'all 0.4s ease'
  },

  primary: {
    background: '#00a1de',
    ':hover': {
      background: '#0195CC'
    },
    ':active': {
      background: '#0B86BC'
    }
  },

  cancel: {
    background: '#fafafa',
    border: '1px solid #d1d2d3',
    color: '#999a94',
    ':hover': {
      background: '#EAEAEA'
    },
    ':active': {
      background: '#DDDDDD'
    }
  }
};
