// @mtk/mcs-components/src/Footer.jsx
import React from 'react';
import StyleSheet from 'react-style';
import * as Screen from './utils/screen';
import * as I18n from './utils/i18n';
import getLocale from './utils/getLocale.js';
import Radium from 'radium';

import socialMedia from './configs/socialMedia.js';

class Footer extends React.Component {
  constructor(props) {
    super(props);
    this.state = this.state || {};
  }

  componentWillMount() {
    // footer 置底監聽
    window.addEventListener('resize', this._handleResize.bind(this));
    window.addEventListener('load', this._handleResize.bind(this));

    // 取得現在網址代表的語言
    let i18nPrefix = getLocale(window.location.href);

    // i18n 設定
    I18n.setup({
      lng: i18nPrefix
    })
    .then((i18n) => {
      this.setState({ i18n: i18n.t });
    })

    .catch((error) => {
      console.log(error);
    });

    // Media Query
    let mql = window.matchMedia('(max-width: 767px)');
    mql.addListener(this._handleMediaQueryChange.bind(this));

    this.setState({
      hasScrollbar: true,
      isMobileMode: mql.matches ? true : false,
      isCN: /.cn/.test(window.location.hostname),
      socialMedia: socialMedia,
      i18n: function(text) { return text; }
    });
  }

  componentDidMount() {
    // 第一次載入觸發調整一次
    this.setState({ hasScrollbar: Screen.hasScrollbar() });
  }

  render() {
    const { isMobileMode } = this.state;

    return (
      <footer style={[
        styles.footer,
        !this.state.hasScrollbar && styles.footerFixToBottom
        ]}>
        <div style={[
          styles.container,
          isMobileMode && { flexDirection: 'column', justifyContent: 'space-around' }
        ]}>

          <ul style={styles.list}>
            {
              this.state.socialMedia[this.state.isCN ? '.cn' : '.com'].map((entry, index) => {
                return (
                  <li key={index} style={styles.li}>
                    <a href={entry.url}
                      target="_blank"><i style={styles.icon} className={`fa ${entry.icon}`}></i></a>
                  </li>
                );
              })
            }
          </ul>

          <div style={styles.rightPart}>
            <div style={styles.rightPartUpper}>
              <a style={[
                styles.legalLink,
                {marginLeft: 0},
                isMobileMode && {marginBottom: 5, display: 'block'}]}
                href="http://www.mediatek.com/">
                MediaTek.com
              </a>
              { !isMobileMode && '|' }
              <a href="http://www.mediatek.com/en/legal-notice/"
                style={[
                styles.legalLink,
                isMobileMode && {marginLeft: 0}
              ]}>
                { this.state.i18n('Legal Notice') }
              </a>
              |
              <a style={styles.legalLink}
                href="http://www.mediatek.com/en/privacy-policy/">
                { this.state.i18n('Privacy Policy') }
              </a>
              |
              <a style={[
                styles.legalLink,
                {marginRight: 0}]}
                href="http://status.mcs.mediatek.com/">
                { this.state.i18n('System Status') }
              </a>
            </div>
            <div style={[
              styles.rightPartLower,
              isMobileMode && { marginTop: 5, marginBottom: 10 }
            ]}>
              <span>&copy; 2015 </span>
              <span>
                { this.state.i18n('MediaTek Inc All Rights Reserved') }
                {this.state.isCN && '京ICP备15035206号'}
              </span>
            </div>
          </div>

        </div>
      </footer>
    );
  }

  /**
  * hasScrollbar
  * 判斷 Footer 如何置底，解決少部分中間內容不夠把 Footer 撐到至底
  * document.readyState complete 避免閃爍。
  *
  * @author Michael Hsu
  */

  _handleResize(event) {
    if (document.readyState === 'complete') {
      this.setState({ hasScrollbar: Screen.hasScrollbar() });
    }
  }

  /**
  * 處理 media query
  * ref: https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Testing_media_queries
  *
  * @input  {object} mql
  *
  * @author Michael Hsu
  */

  _handleMediaQueryChange(mql) {
    this.setState({
      isMobileMode: mql.matches ? true : false
    });
  }

}

export default Radium(Footer);

const styles = {
  footer: {
    width: '100%',
    overflow: 'hidden',
    marginBottom: -6,
    minHeight: 160,
    background: '#353630',
    display: 'inline-block',
    fontSize: 13,
    zIndex: 2,
    position: 'relative'
  },
  footerFixToBottom: {
    position: 'absolute',
    left: 0,
    bottom: 0,
    marginBottom: 0
  },
  container: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    boxSizing: 'border-box',
    minHeight: 160,
    maxWidth: 954,
    color: '#fff',
    margin: '0 auto'
  },
  list: {
    listStyleType: 'none',
    marginLeft: 12,
    paddingLeft: 0
  },
  li: {
    display: 'inline-block',
    marginRight: 20
  },
  icon: {
    color: '#fff',
    fontSize: 60
  },
  rightPart: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    minHeight: 50,
    marginRight: 12,
    marginLeft: 12,
    fontFamily: 'Helvetica, sans-serif'
  },
  rightPartUpper: {
    fontSize: 15
  },
  rightPartLower: {
    color:'#D1D2D3',
    fontSize: 13,
    fontWeight: 'bold'
  },
  legalLink: {
    color: '#fff',
    fontSize: 15,
    fontWeight: 'bold',
    textDecoration: 'none',
    outline: 0,
    marginRight: 10,
    marginLeft: 10
  }
};
