// @mtk/mcs-components/src/Nav.jsx
import React from 'react';
import Radium from 'radium';
import {PropTypes} from 'react';

class Nav extends React.Component {

  render() {
    let {
      style: listStyle,
      children,
      dropdownMenu,
      ...props
    } = this.props;

    return (
      <ul {...props} style={[
        dropdownMenu ? styles.dropdownMenu : styles.list,
        listStyle
      ]}>
        {React.Children.map(children, this._renderChildren)}
      </ul>
    );
  }

  _renderChildren(child, index) {
    return React.cloneElement(child, {key: index});
  }
}

export default Radium(Nav);

Nav.propTypes = {
  style: PropTypes.object,
  dropdownMenu: PropTypes.bool
};
Nav.defaultProps = {};

const styles = {
  list: {
    display: 'inline-block',
    padding: 0,
    margin: 0
  },
  dropdownMenu: {
    display: 'none',
    position: 'absolute',
    top: '50px',
    minWidth: 100,
    padding: 0,
    zIndex: 1000,
    float: 'left',
    textAlign: 'left',
    listStyle: 'none',
    backgroundColor: '#1995cc',
    borderRadius: '0 0 4px 4px'
  }
};
