// mtk-ui/karma.conf.js
/* eslint no-var: 0 */
var path = require('path');

module.exports = function setupKarmaConfig(config) {
  config.set({
    browsers: ['Chrome'],
    frameworks: ['mocha', 'sinon-chai'],
    reporters: ['mocha', 'coverage'],

    files: [
      'tests.webpack.js',
    ],

    preprocessors: {
      'tests.webpack.js': ['webpack', 'sourcemap'],
    },

    webpack: {
      devtool: 'inline-source-map',
      module: {
        loaders: [
          {
            test: /\.jsx?$/,
            exclude: /node_modules/,
            loader: 'babel',
          },
          {
            test: /\.js$/,
            exclude: /__tests__/,
            include: path.resolve('src/'),
            loader: 'isparta',
          },
          {
            test: /\.css$/,
            exclude: /font-awesome/,
            loader: 'style!css?modules&importLoaders=1&localIdentName=[name]__[local]',
          },
          {
            include: /font-awesome\.css/,
            loader: 'css',
          },
          {
            test: /\.(otf|eot|svg|ttf|woff)\??/,
            loader: 'url',
          },
        ],
      },
      resolve: {
        alias: {
          'mtk-ui': path.resolve(__dirname, './src/index'),
          'react': path.resolve(__dirname, './node_modules/react'),
        },
      },
    },

    webpackServer: {
      noInfo: true,
    },

    coverageReporter: {
      dir: 'coverage/',
      reporters: [
        { type: 'html' },
      ],
    },

    singleRun: !process.env.TESTTEST,
  });

  if (process.env.BAMBOO) {
    config.browsers = ['PhantomJS'];
  }
};
