// mtk-ui/lib/ActionButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _EventListener = require('fbjs/lib/EventListener');

var _EventListener2 = _interopRequireDefault(_EventListener);

var _getEventTargetFromPropsSafely = require('./utils/getEventTargetFromPropsSafely');

var _getEventTargetFromPropsSafely2 = _interopRequireDefault(_getEventTargetFromPropsSafely);

var _ActionButton = require('./styles/ActionButton.css');

var _ActionButton2 = _interopRequireDefault(_ActionButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['keyCode', 'enableKeyCode', 'onRelease', 'onPress'];

var ActionButton = (_temp = _class = function (_React$Component) {
  (0, _inherits3.default)(ActionButton, _React$Component);

  function ActionButton() {
    (0, _classCallCheck3.default)(this, ActionButton);
    return (0, _possibleConstructorReturn3.default)(this, (ActionButton.__proto__ || Object.getPrototypeOf(ActionButton)).apply(this, arguments));
  }

  (0, _createClass3.default)(ActionButton, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.props.enableKeyCode) {
        var eventTarget = this.getEventTarget() || document;
        this.keyDownEventListener = _EventListener2.default.listen(eventTarget, 'keydown', this.onKeyDown.bind(this));
        this.keyUpEventListener = _EventListener2.default.listen(eventTarget, 'keyup', this.onKeyUp.bind(this));
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.enableKeyCode !== this.props.enableKeyCode) {
        if (nextProps.enableKeyCode) {
          var eventTarget = this.getEventTarget() || document;
          this.keyDownEventListener = _EventListener2.default.listen(eventTarget, 'keydown', this.onKeyDown.bind(this));
          this.keyUpEventListener = _EventListener2.default.listen(eventTarget, 'keyup', this.onKeyUp.bind(this));
        } else {
          this.keyDownEventListener.remove();
          this.keyUpEventListener.remove();
        }
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.props.enableKeyCode) {
        this.keyDownEventListener.remove();
        this.keyUpEventListener.remove();
      }
    }
  }, {
    key: 'onKeyDown',
    value: function onKeyDown(e) {
      if (e.keyCode === this.props.keyCode) this.props.onPress(e);
    }
  }, {
    key: 'onKeyUp',
    value: function onKeyUp(e) {
      if (e.keyCode === this.props.keyCode) this.props.onRelease(e);
    }
  }, {
    key: 'onMouseDown',
    value: function onMouseDown(e) {
      var _this2 = this;

      this.props.onPress(e);
      this.props.onMouseDown(e);

      this.mouseUpEventListener = _EventListener2.default.listen(document, 'mouseup', function () {
        _this2.props.onRelease(e);
        _this2.mouseUpEventListener.remove();
      });
    }
  }, {
    key: 'getEventTarget',
    value: function getEventTarget() {
      return (0, _getEventTargetFromPropsSafely2.default)(this.props);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          active = _props.active,
          children = _props.children,
          className = _props.className,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['active', 'children', 'className']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), {
          className: (0, _classnames2.default)(_ActionButton2.default.frame, className),
          onMouseDown: this.onMouseDown.bind(this) }),
        (0, _jsx3.default)('div', {
          className: _ActionButton2.default.core
        }, void 0, (0, _jsx3.default)('div', {
          className: _ActionButton2.default.child
        }, void 0, children), active && (0, _jsx3.default)('div', {
          className: _ActionButton2.default.active
        }))
      );
    }
  }]);
  return ActionButton;
}(_react2.default.Component), _class.propTypes = {
  className: _react.PropTypes.string,
  active: _react.PropTypes.bool,
  keyCode: _react.PropTypes.number,
  enableKeyCode: _react.PropTypes.bool,
  eventTarget: _react.PropTypes.func,
  onRelease: _react.PropTypes.func,
  onPress: _react.PropTypes.func,
  onMouseDown: _react.PropTypes.func,
  children: _react.PropTypes.node.isRequired
}, _class.defaultProps = {
  active: false,
  keyCode: 13,
  enableKeyCode: false,
  onRelease: _emptyFunction2.default,
  onPress: _emptyFunction2.default,
  onMouseDown: _emptyFunction2.default
}, _temp);
exports.default = ActionButton;
module.exports = exports['default'];
