// mtk-ui/lib/Arc.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Arc = function Arc(_ref) {
  var max = _ref.max,
      now = _ref.now,
      offset = _ref.offset,
      radius = _ref.radius,
      strokeWidth = _ref.strokeWidth,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['max', 'now', 'offset', 'radius', 'strokeWidth']);

  var startRadian = offset / max * 2 * Math.PI;
  var startX = strokeWidth / 2 + radius + radius * Math.sin(startRadian);
  var startY = strokeWidth / 2 + radius - radius * Math.cos(startRadian);

  var endRadian = startRadian + now / max * 2 * Math.PI;
  var endX = strokeWidth / 2 + radius + radius * Math.sin(endRadian) - 0.00001; // '- 0.00001' is for now === max
  var endY = strokeWidth / 2 + radius - radius * Math.cos(endRadian);

  var largeArcFlag = now / max > 0.5 ? 1 : 0;

  return _react2.default.createElement('path', (0, _extends3.default)({}, otherProps, {
    d: 'M ' + startX + ' ' + startY + ' A ' + radius + ',' + radius + ' 0 ' + largeArcFlag + ' 1 ' + endX + ' ' + endY,
    strokeWidth: strokeWidth,
    fillOpacity: '0'
  }));
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  radius: 48,
  now: 100,
  max: 100,
  offset: 0,
  stroke: '#F1F2F7',
  strokeWidth: 4
}), (0, _setPropTypes3.default)({
  radius: _react.PropTypes.number,
  now: _react.PropTypes.number,
  max: _react.PropTypes.number,
  offset: _react.PropTypes.number,
  stroke: _react.PropTypes.string,
  strokeWidth: _react.PropTypes.number
}))(Arc);
module.exports = exports['default'];
