// mtk-ui/lib/Arrow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Arrow = require('./styles/Arrow.css');

var _Arrow2 = _interopRequireDefault(_Arrow);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Arrow = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Arrow, _React$Component);

  function Arrow() {
    (0, _classCallCheck3.default)(this, Arrow);
    return (0, _possibleConstructorReturn3.default)(this, (Arrow.__proto__ || Object.getPrototypeOf(Arrow)).apply(this, arguments));
  }

  (0, _createClass3.default)(Arrow, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          placement = _props.placement,
          className = _props.className,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['placement', 'className']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, otherProps, { className: (0, _classnames2.default)(_Arrow2.default[placement], className) }),
        (0, _jsx3.default)('svg', {
          viewBox: '0 0 12 8',
          className: _Arrow2.default.arrow
        }, void 0, (0, _jsx3.default)('polygon', {
          points: '6,0 12,8 0,8'
        }))
      );
    }
  }]);
  return Arrow;
}(_react2.default.Component), _class2.propTypes = {
  className: _react.PropTypes.string,
  placement: _react.PropTypes.string.isRequired
}, _class2.defaultProps = {
  placement: 'bottom'
}, _temp)) || _class;

exports.default = Arrow;
module.exports = exports['default'];
