// mtk-ui/lib/Breadcrumb.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyObject = require('fbjs/lib/emptyObject');

var _emptyObject2 = _interopRequireDefault(_emptyObject);

var _BreadcrumbItem = require('./BreadcrumbItem');

var _BreadcrumbItem2 = _interopRequireDefault(_BreadcrumbItem);

var _MiChevronRight = require('mtk-icon/lib/MiChevronRight');

var _MiChevronRight2 = _interopRequireDefault(_MiChevronRight);

var _Breadcrumb = require('./styles/Breadcrumb.css');

var _Breadcrumb2 = _interopRequireDefault(_Breadcrumb);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['items', 'iconProps'];

var Breadcrumb = function Breadcrumb(_ref) {
  var className = _ref.className,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['className']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), { className: (0, _classnames2.default)(_Breadcrumb2.default.container, className) }),
    renderItems(otherProps)
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  iconProps: _emptyObject2.default,
  items: []
}), (0, _setPropTypes3.default)({
  children: _react.PropTypes.node,
  items: _react.PropTypes.arrayOf(_react.PropTypes.shape((0, _extends3.default)({}, _BreadcrumbItem2.default.propTypes))),
  className: _react.PropTypes.string,
  iconProps: _react.PropTypes.object
}))(Breadcrumb);

/**
 * Render <BreadcrumbItems> components
 *
 * @param {object} Object
 * @param {node}   Object.children
 * @param {array}  Object.items
 * @param {object} Object.iconProps - custom icon props.
 */

function renderItems(_ref2) {
  var children = _ref2.children,
      items = _ref2.items,
      iconProps = _ref2.iconProps;

  var itemFactory = (0, _react.createFactory)(_BreadcrumbItem2.default);

  if (children) {
    return _react.Children.map(children, function (child, index) {
      return renderChildWithIcon({
        child: (0, _react.cloneElement)(child),
        index: index,
        iconProps: iconProps
      });
    });
  }

  return items.map(function (childProps, index) {
    return renderChildWithIcon({
      child: itemFactory(childProps),
      index: index,
      iconProps: iconProps
    });
  });
}

/**
 * Wrap child with icon.
 *
 * @param {object} Object
 * @param {node}   Object.child
 * @param {number} Object.index - index as react key.
 * @param {object} Object.iconProps - custom icon props.
 */
function renderChildWithIcon(_ref3) {
  var child = _ref3.child,
      index = _ref3.index,
      iconProps = _ref3.iconProps;
  var _iconProps$className = iconProps.className,
      className = _iconProps$className === undefined ? '' : _iconProps$className,
      otherProps = (0, _objectWithoutProperties3.default)(iconProps, ['className']);


  return (0, _jsx3.default)('div', {
    className: _Breadcrumb2.default.itemWrap
  }, index, index > 0 && _react2.default.createElement(_MiChevronRight2.default, (0, _extends3.default)({}, otherProps, { className: (0, _classnames2.default)(_Breadcrumb2.default.icon, className) })), child);
}
module.exports = exports['default'];
