// mtk-ui/lib/ButtonGroupSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _withHandlers2 = require('recompose/withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _defaultPropHandler = require('./decorators/defaultPropHandler');

var _defaultPropHandler2 = _interopRequireDefault(_defaultPropHandler);

var _InputGroup = require('./InputGroup');

var _InputGroup2 = _interopRequireDefault(_InputGroup);

var _omitProps = require('./utils/omitProps');

var _omitProps2 = _interopRequireDefault(_omitProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ButtonGroupSelect = function ButtonGroupSelect(_ref) {
  var children = _ref.children,
      onClick = _ref.onClick,
      value = _ref.value,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['children', 'onClick', 'value']);
  return _react2.default.createElement(
    _InputGroup2.default,
    otherProps,
    _react.Children.map(children, function (child, index) {
      return (0, _react.cloneElement)(child, {
        key: child.props.key === null ? child.props.key : index,
        onClick: onClick(child.props.value),
        active: value === child.props.value
      });
    })
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  onClick: _emptyFunction2.default,
  required: false
}), (0, _setPropTypes3.default)({
  children: _react.PropTypes.node.isRequired,
  onClick: _react.PropTypes.func, // (e) => {}
  onChange: _react.PropTypes.func, // (e, value) => {}
  defaultValue: _react.PropTypes.any,
  value: _react.PropTypes.any,
  required: _react.PropTypes.bool
}), (0, _defaultPropHandler2.default)('value', 'onChange', function (setValue) {
  return function (e, value) {
    return setValue(value);
  };
}, function (props) {
  return props.defaultValue;
}), (0, _withHandlers3.default)({
  onClick: function onClick(props) {
    return function (value) {
      return function (e) {
        props.onClick(e); /* origin onClick props */
        if (!props.required && props.value === value) {
          props.onChange(e, null); /* click activated one */
        } else {
          props.onChange(e, value); /* click non-activated one */
        }
      };
    };
  }
}), (0, _omitProps2.default)(['setvalue']))(ButtonGroupSelect);
module.exports = exports['default'];
