// mtk-ui/lib/Button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _withHandlers2 = require('recompose/withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _Button = require('./styles/Button.css');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Button = function Button(_ref) {
  var children = _ref.children,
      kind = _ref.kind,
      size = _ref.size,
      disabled = _ref.disabled,
      block = _ref.block,
      round = _ref.round,
      square = _ref.square,
      active = _ref.active,
      readOnly = _ref.readOnly,
      className = _ref.className,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['children', 'kind', 'size', 'disabled', 'block', 'round', 'square', 'active', 'readOnly', 'className']);
  return _react2.default.createElement(
    'button',
    (0, _extends3.default)({}, otherProps, {
      className: (0, _classnames2.default)(disabled ? _Button2.default.disabled : _Button2.default[active ? kind + 'Active' : kind], size && _Button2.default[size], block && _Button2.default.block, round && _Button2.default.round, square && _Button2.default.square, readOnly && _Button2.default.readOnly, className)
    }),
    (0, _jsx3.default)('div', {
      className: _Button2.default.innerWrapper
    }, void 0, children)
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  kind: 'primary',
  size: 'large',
  onClick: _emptyFunction2.default
}), (0, _setPropTypes3.default)({
  children: _react.PropTypes.node,
  kind: _react.PropTypes.oneOf(['primary', 'secondary', 'cancel', 'text']),
  size: _react.PropTypes.oneOf(['small', 'large']),
  block: _react.PropTypes.bool,
  disabled: _react.PropTypes.bool,
  round: _react.PropTypes.bool,
  square: _react.PropTypes.bool,
  active: _react.PropTypes.bool,
  readOnly: _react.PropTypes.bool,
  className: _react.PropTypes.string
}), (0, _withHandlers3.default)({
  onClick: function onClick(props) {
    return function () {
      return !props.disabled && props.onClick.apply(props, arguments);
    };
  }
}))(Button);
module.exports = exports['default'];
