// mtk-ui/lib/Carousel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _dec, _dec2, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _withContext = require('recompose/withContext');

var _withContext2 = _interopRequireDefault(_withContext);

var _defaultProps = require('recompose/defaultProps');

var _defaultProps2 = _interopRequireDefault(_defaultProps);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _emptyObject = require('fbjs/lib/emptyObject');

var _emptyObject2 = _interopRequireDefault(_emptyObject);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactSwipe = require('react-swipe');

var _reactSwipe2 = _interopRequireDefault(_reactSwipe);

var _MiChevronLeft = require('mtk-icon/lib/MiChevronLeft');

var _MiChevronLeft2 = _interopRequireDefault(_MiChevronLeft);

var _MiChevronRight = require('mtk-icon/lib/MiChevronRight');

var _MiChevronRight2 = _interopRequireDefault(_MiChevronRight);

var _MiImageGallery = require('mtk-icon/lib/MiImageGallery');

var _MiImageGallery2 = _interopRequireDefault(_MiImageGallery);

var _ControllablePaginationDots = require('./ControllablePaginationDots');

var _ControllablePaginationDots2 = _interopRequireDefault(_ControllablePaginationDots);

var _Pagination = require('./Pagination');

var _Pagination2 = _interopRequireDefault(_Pagination);

var _Carousel = require('./styles/Carousel.css');

var _Carousel2 = _interopRequireDefault(_Carousel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['focusedIndex', 'onFocusChange', 'showDots', 'description', 'continuous', 'pagination', 'controlProps'];

var Carousel = (_dec = (0, _withContext2.default)({
  focusedIndex: _react.PropTypes.number.isRequired
}, function (props) {
  return {
    focusedIndex: props.focusedIndex
  };
}), _dec2 = (0, _defaultProps2.default)({
  speed: 400,
  description: null,
  continuous: false,
  pagination: true,
  showDots: false,
  onFocusChange: _emptyFunction2.default,
  controlProps: _emptyObject2.default
}), (0, _pure2.default)(_class = _dec(_class = _dec2(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Carousel, _React$Component);

  function Carousel() {
    (0, _classCallCheck3.default)(this, Carousel);
    return (0, _possibleConstructorReturn3.default)(this, (Carousel.__proto__ || Object.getPrototypeOf(Carousel)).apply(this, arguments));
  }

  (0, _createClass3.default)(Carousel, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      var focusedIndex = this.props.focusedIndex;


      return this.refs.reactSwipe.swipe.slide(focusedIndex);
    }
  }, {
    key: 'onPrevClick',
    value: function onPrevClick(e) {
      var currentPosition = this.getCurrentPosition();
      var isFirst = currentPosition === 0;
      var focusedIndex = isFirst ? _react.Children.count(this.props.children) - 1 : currentPosition - 1;

      this.props.onFocusChange(e, focusedIndex);
    }
  }, {
    key: 'onNextClick',
    value: function onNextClick(e) {
      var currentPosition = this.getCurrentPosition();
      var isLast = currentPosition === _react.Children.count(this.props.children) - 1;
      var focusedIndex = isLast ? 0 : currentPosition + 1;

      this.props.onFocusChange(e, focusedIndex);
    }

    /**
     * onSlideChange()
     *
     * 讓 swipe 的操作行為把 focusedIndex 也傳出去。
     */

  }, {
    key: 'onSlideChange',
    value: function onSlideChange(currentPosition) {
      var focusedIndex = this.props.focusedIndex;


      if (currentPosition !== focusedIndex) {
        this.props.onFocusChange({}, currentPosition);
      }
    }
  }, {
    key: 'getCurrentPosition',
    value: function getCurrentPosition() {
      return this.refs.reactSwipe.swipe.getPos();
    }
  }, {
    key: 'renderChild',
    value: function renderChild() {
      var _props = this.props,
          children = _props.children,
          continuous = _props.continuous,
          focusedIndex = _props.focusedIndex,
          swipeProps = _props.swipeProps,
          speed = _props.speed;

      // TODO: i18n

      if (_react.Children.count(children) === 0) {
        return (0, _jsx3.default)('div', {
          className: _Carousel2.default.noImageWrap
        }, void 0, (0, _jsx3.default)(_MiImageGallery2.default, {
          className: _Carousel2.default.noImageIcon
        }), (0, _jsx3.default)('div', {}, void 0, 'No image yet.'));
      }

      return (0, _jsx3.default)('div', {
        className: _Carousel2.default.reactSwipeWrap
      }, void 0, _react2.default.createElement(
        _reactSwipe2.default,
        (0, _extends3.default)({
          ref: 'reactSwipe',
          key: _react.Children.count(children),
          className: _Carousel2.default.reactSwipe,
          swipeOptions: {
            speed: speed,
            continuous: continuous,
            startSlide: focusedIndex,
            callback: this.onSlideChange.bind(this)
          }
        }, swipeProps),
        _react.Children.map(children, function (child, index) {
          return (0, _jsx3.default)('div', {
            className: _Carousel2.default.childWrap
          }, child.props.key || index, (0, _react.cloneElement)(child, { className: (0, _classnames2.default)(_Carousel2.default.child, child.props.className) }));
        })
      ));
    }
  }, {
    key: 'renderPaginationDots',
    value: function renderPaginationDots() {
      var _props2 = this.props,
          children = _props2.children,
          focusedIndex = _props2.focusedIndex,
          showDots = _props2.showDots;

      return showDots && (0, _jsx3.default)('div', {
        className: _Carousel2.default.PaginationDotsWrap
      }, void 0, (0, _jsx3.default)(_ControllablePaginationDots2.default, {
        now: focusedIndex + 1,
        max: children.length
      }));
    }
  }, {
    key: 'renderFooter',
    value: function renderFooter() {
      var _props3 = this.props,
          description = _props3.description,
          children = _props3.children;


      if (_react.Children.count(children) === 0) return (0, _jsx3.default)('noscript', {});

      return description && (0, _jsx3.default)('div', {
        className: _Carousel2.default.footer
      }, void 0, description);
    }
  }, {
    key: 'renderPagination',
    value: function renderPagination() {
      var _props4 = this.props,
          pagination = _props4.pagination,
          children = _props4.children,
          focusedIndex = _props4.focusedIndex;


      if (_react.Children.count(children) === 0) return (0, _jsx3.default)('noscript', {});

      return pagination && (0, _jsx3.default)(_Pagination2.default, {
        className: _Carousel2.default.pagination,
        max: _react.Children.count(children),
        now: focusedIndex + 1
      });
    }
  }, {
    key: 'renderControl',
    value: function renderControl() {
      var _props5 = this.props,
          _props5$controlProps = _props5.controlProps,
          style = _props5$controlProps.style,
          className = _props5$controlProps.className,
          continuous = _props5.continuous,
          children = _props5.children,
          focusedIndex = _props5.focusedIndex;


      if (_react.Children.count(children) === 0) return (0, _jsx3.default)('noscript', {});

      var isFirst = focusedIndex === 0;
      var isLast = focusedIndex === _react.Children.count(children) - 1;

      return (0, _jsx3.default)('div', {
        className: _Carousel2.default.controlWrap
      }, void 0, !continuous && isFirst ? (0, _jsx3.default)('div', {
        className: _Carousel2.default.control
      }) : _react2.default.createElement(
        'div',
        {
          ref: 'controlLeft',
          style: style,
          className: (0, _classnames2.default)(_Carousel2.default.controlLeft, className),
          onClick: this.onPrevClick.bind(this)
        },
        (0, _jsx3.default)(_MiChevronLeft2.default, {
          className: _Carousel2.default.controlIcon
        })
      ), !continuous && isLast ? (0, _jsx3.default)('div', {
        className: _Carousel2.default.control
      }) : _react2.default.createElement(
        'div',
        {
          ref: 'controlRight',
          style: style,
          className: (0, _classnames2.default)(_Carousel2.default.controlRight, className),
          onClick: this.onNextClick.bind(this)
        },
        (0, _jsx3.default)(_MiChevronRight2.default, {
          className: _Carousel2.default.controlIcon
        })
      ));
    }
  }, {
    key: 'render',
    value: function render() {
      var _props6 = this.props,
          className = _props6.className,
          otherProps = (0, _objectWithoutProperties3.default)(_props6, ['className']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), { className: (0, _classnames2.default)(_Carousel2.default.container, className) }),
        this.renderChild(),
        this.renderFooter(),
        this.renderPaginationDots(),
        this.renderPagination(),
        this.renderControl()
      );
    }
  }]);
  return Carousel;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node.isRequired,
  className: _react.PropTypes.string,
  continuous: _react.PropTypes.bool,
  description: _react.PropTypes.node,
  pagination: _react.PropTypes.bool,
  showDots: _react.PropTypes.bool,
  /**
   * Only support Controlled component now
   */
  focusedIndex: _react.PropTypes.number.isRequired,
  onFocusChange: _react.PropTypes.func,

  /**
   * react-swipe props
   * ref: https://github.com/jed/react-swipe#props
   */
  swipeProps: _react.PropTypes.shape((0, _extends3.default)({}, _reactSwipe2.default.propTypes)),

  /**
   * For custom control style
   */
  controlProps: _react.PropTypes.shape({
    className: _react.PropTypes.string,
    style: _react.PropTypes.object
  })
}, _temp)) || _class) || _class) || _class);
exports.default = Carousel;
module.exports = exports['default'];
