// mtk-ui/lib/ControllablePaginationDots.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultDotComponent = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _withHandlers2 = require('recompose/withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _ControllablePaginationDots = require('./styles/ControllablePaginationDots.css');

var _ControllablePaginationDots2 = _interopRequireDefault(_ControllablePaginationDots);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DefaultDotComponent = exports.DefaultDotComponent = (0, _compose3.default)((0, _withHandlers3.default)({
  onDotClick: function onDotClick(props) {
    return function (e) {
      return props.onPageChange(e, props.page);
    };
  }
}), _pure3.default)(function (_ref) {
  var onDotClick = _ref.onDotClick,
      active = _ref.active,
      readOnly = _ref.readOnly;
  return (0, _jsx3.default)('div', {
    className: active ? _ControllablePaginationDots2.default.activeDot : _ControllablePaginationDots2.default.dot,
    onClick: readOnly ? _emptyFunction2.default : onDotClick
  });
});

var ControllablePaginationDots = function ControllablePaginationDots(_ref2) {
  var max = _ref2.max,
      now = _ref2.now,
      className = _ref2.className,
      onPageChange = _ref2.onPageChange,
      _ref2$readOnly = _ref2.readOnly,
      readOnly = _ref2$readOnly === undefined ? true : _ref2$readOnly,
      _ref2$dotComponent = _ref2.dotComponent,
      dotComponent = _ref2$dotComponent === undefined ? DefaultDotComponent : _ref2$dotComponent,
      otherProps = (0, _objectWithoutProperties3.default)(_ref2, ['max', 'now', 'className', 'onPageChange', 'readOnly', 'dotComponent']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, otherProps, {
      className: (0, _classnames2.default)(_ControllablePaginationDots2.default.wrap, className)
    }),
    Array.from({ length: max }, function (e, index) {
      return (0, _react.createElement)(dotComponent, {
        key: index,
        page: index + 1,
        active: now === index + 1,
        onPageChange: onPageChange,
        readOnly: readOnly
      });
    })
  );
};

exports.default = (0, _compose3.default)((0, _setPropTypes3.default)({
  className: _react.PropTypes.string,
  max: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]).isRequired,
  now: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]).isRequired,
  onChange: _react.PropTypes.func.isRequired
}), (0, _withHandlers3.default)({
  onPageChange: function onPageChange(_ref3) {
    var _ref3$onChange = _ref3.onChange,
        onChange = _ref3$onChange === undefined ? _emptyFunction2.default : _ref3$onChange;
    return function (e, page) {
      return onChange(e, page);
    };
  }
}), _pure3.default)(ControllablePaginationDots);
