// mtk-ui/lib/CopyButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _withHandlers2 = require('recompose/withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _withState2 = require('recompose/withState');

var _withState3 = _interopRequireDefault(_withState2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _isFunction2 = require('lodash/isFunction');

var _isFunction3 = _interopRequireDefault(_isFunction2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _copyToClipboard = require('copy-to-clipboard');

var _copyToClipboard2 = _interopRequireDefault(_copyToClipboard);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _LoadingMorph = require('./LoadingMorph');

var _LoadingMorph2 = _interopRequireDefault(_LoadingMorph);

var _CopyButton = require('./styles/CopyButton.css');

var _CopyButton2 = _interopRequireDefault(_CopyButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['text', 'setStatus'];

var CopyButton = function CopyButton(_ref) {
  var children = _ref.children,
      className = _ref.className,
      status = _ref.status,
      onClick = _ref.onClick,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['children', 'className', 'status', 'onClick']);
  return _react2.default.createElement(
    _Button2.default,
    (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), {
      size: 'small',
      onClick: onClick,
      className: (0, _classnames2.default)(_CopyButton2.default.button, className)
    }),
    status === 'default' && children,
    (0, _jsx3.default)(_LoadingMorph2.default, {
      status: status,
      size: 12
    })
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  onClick: _emptyFunction2.default,
  children: 'Copy'
}), (0, _setPropTypes3.default)({
  onClick: _react.PropTypes.func,
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  text: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.func]).isRequired
}), (0, _withState3.default)('status', 'setStatus', 'default'), (0, _withHandlers3.default)({
  onClick: function onClick(props) {
    return function (e) {
      if (props.status !== 'default') return;
      props.onClick(e);
      props.setStatus('loading');
      (0, _copyToClipboard2.default)((0, _isFunction3.default)(props.text) ? props.text(props) : props.text);
      setTimeout(function () {
        return props.setStatus('success');
      }, 400);
      setTimeout(function () {
        return props.setStatus('default');
      }, 1800);
    };
  }
}))(CopyButton);
module.exports = exports['default'];
