// mtk-ui/lib/Day.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactCssThemr = require('react-css-themr');

var _Item = require('./Item');

var _Item2 = _interopRequireDefault(_Item);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _Day = require('./styles/Day.css');

var _Day2 = _interopRequireDefault(_Day);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Day = (_dec = (0, _reactCssThemr.themr)('Day', _Day2.default), _dec(_class = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Day, _React$Component);

  function Day() {
    (0, _classCallCheck3.default)(this, Day);
    return (0, _possibleConstructorReturn3.default)(this, (Day.__proto__ || Object.getPrototypeOf(Day)).apply(this, arguments));
  }

  (0, _createClass3.default)(Day, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          currentPageDate = _props.currentPageDate,
          theme = _props.theme,
          kind = _props.kind,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['currentPageDate', 'theme', 'kind']);
      var _props2 = this.props,
          value = _props2.value,
          selectedValue = _props2.selectedValue,
          today = _props2.today,
          max = _props2.max,
          min = _props2.min;

      /* could be Date object or moment object, need momentify */

      value = (0, _moment2.default)(value);
      selectedValue = (0, _moment2.default)(selectedValue);
      today = (0, _moment2.default)(today);
      max = (0, _moment2.default)(max);
      min = (0, _moment2.default)(min);

      var isDisabled = kind === 'disabled' || currentPageDate && !value.isSame(currentPageDate, 'month') || !currentPageDate && !value.isSame(selectedValue, 'month') || value.isAfter(max, 'day') || value.isBefore(min, 'day');

      return _react2.default.createElement(
        _Item2.default,
        (0, _extends3.default)({}, otherProps, {
          className: (0, _classnames2.default)(theme.base, (!currentPageDate || currentPageDate.isSame(today, 'month')) && value.isSame(today, 'day') && theme.today, !isDisabled && value.isSame(selectedValue, 'day') && theme.selected, isDisabled && theme.disabled, kind === 'inactive' && theme.inactive),
          value: value,
          selectedValue: selectedValue,
          disabled: isDisabled
        }),
        value.get('date')
      );
    }
  }]);
  return Day;
}(_react2.default.Component), _class2.propTypes = {
  value: _react.PropTypes.object.isRequired, /* could be Date object or moment object, need momentify */
  selectedValue: _react.PropTypes.object, /* could be Date object or moment object, need momentify */
  today: _react.PropTypes.object.isRequired, /* could be Date object or moment object, need momentify */
  max: _react.PropTypes.object.isRequired, /* could be Date object or moment object, need momentify */
  min: _react.PropTypes.object.isRequired, /* could be Date object or moment object, need momentify */
  currentPageDate: _react.PropTypes.object,
  kind: _react.PropTypes.oneOf(['default', 'disabled', 'inactive']),
  theme: _react.PropTypes.object
}, _class2.defaultProps = {
  selectedValue: null,
  today: new Date(), /* Today */
  max: new Date(), /* Today */
  min: new Date(-8640000000000000)
}, _temp)) || _class) || _class);
exports.default = Day;
module.exports = exports['default'];
