// mtk-ui/lib/Dialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _Modal = require('react-overlays/lib/Modal');

var _Modal2 = _interopRequireDefault(_Modal);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _EventListener = require('fbjs/lib/EventListener');

var _EventListener2 = _interopRequireDefault(_EventListener);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _removeEventListenerSafely = require('./utils/removeEventListenerSafely.js');

var _removeEventListenerSafely2 = _interopRequireDefault(_removeEventListenerSafely);

var _throttle = require('core-decorators/lib/throttle');

var _throttle2 = _interopRequireDefault(_throttle);

var _Dialog = require('./styles/Dialog.css');

var _Dialog2 = _interopRequireDefault(_Dialog);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Dialog = (_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Dialog, _React$Component);

  function Dialog() {
    var _ref;

    (0, _classCallCheck3.default)(this, Dialog);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = (0, _possibleConstructorReturn3.default)(this, (_ref = Dialog.__proto__ || Object.getPrototypeOf(Dialog)).call.apply(_ref, [this].concat(args)));

    _this.onResize = function () {
      return _this.resize();
    };

    _this.handleBackdropClick = function (e) {
      if (e.target !== e.currentTarget) {
        return;
      }

      _this.props.onHide();
    };

    _this.state = {
      hasScrollbar: false
    };
    _this.windowResizeListener = null;
    return _this;
  }

  (0, _createClass3.default)(Dialog, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.props.show && !this.windowResizeListener) {
        this.windowResizeListener = _EventListener2.default.listen(window, 'resize', this.onResize);
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var toOpen = !this.props.show && nextProps.show;
      var toClose = this.props.show && !nextProps.show;
      if (toOpen) {
        this.windowResizeListener = _EventListener2.default.listen(window, 'resize', this.onResize);
      } else if (toClose) {
        (0, _removeEventListenerSafely2.default)(this.windowResizeListener);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      (0, _removeEventListenerSafely2.default)(this.windowResizeListener);
    }

    /**
     * onResize
     *
     * 判斷 dialog 視窗的位置用
     * 當有出現 Scroller bar fixed 置頂
     * 當沒有出現 Scroller bar 置中
     *
     * @return (boolean) hasScrollbar
     *
     * @author Michael Hsu
     */

  }, {
    key: 'resize',
    value: function resize() {
      var rootElem = document.documentElement || document.body;
      var dialog = (0, _reactDom.findDOMNode)(this.refs.dialog);

      this.props.onResize();
      this.setState({
        hasScrollbar: dialog.offsetHeight >= rootElem.clientHeight
      });
    }

    /**
     * handleBackdropClick
     *
     * 因為包起來的 div 擋到後面的黑色背景 backdrop
     * 因此這邊自行處理
     * 如果用 CSS pointer-events，在 Safari 上會不能 scroll
     *
     * @author Michael Hsu
     */

  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          show = _props.show,
          size = _props.size,
          onHide = _props.onHide,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'show', 'size', 'onHide']);


      return _react2.default.createElement(
        _Modal2.default,
        (0, _extends3.default)({}, otherProps, {
          backdropClassName: _Dialog2.default.backdrop,
          show: show,
          onShow: this.onResize,
          onHide: onHide
        }),
        (0, _jsx3.default)('div', {
          onClick: this.handleBackdropClick,
          className: _Dialog2.default.wrap
        }, void 0, _react2.default.createElement(
          'div',
          {
            ref: 'dialog',
            className: (0, _classnames2.default)(_Dialog2.default.content, _Dialog2.default[size], this.state.hasScrollbar ? _Dialog2.default.top : _Dialog2.default.center)
          },
          children
        ))
      );
    }
  }]);
  return Dialog;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node,
  show: _react.PropTypes.bool.isRequired,
  size: _react.PropTypes.oneOf(['large', 'small']).isRequired,
  onHide: _react.PropTypes.func.isRequired,
  onResize: _react.PropTypes.func
}, _class2.defaultProps = {
  size: 'small',
  onResize: _emptyFunction2.default,
  onHide: _emptyFunction2.default
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'resize', [_throttle2.default], Object.getOwnPropertyDescriptor(_class.prototype, 'resize'), _class.prototype)), _class);
exports.default = (0, _compose3.default)(_pure3.default)(Dialog);
module.exports = exports['default'];
