// mtk-ui/lib/DropdownButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _dec, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactOverlays = require('react-overlays');

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _withState = require('recompose/withState');

var _withState2 = _interopRequireDefault(_withState);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _SlideDown = require('./Transitions/SlideDown');

var _SlideDown2 = _interopRequireDefault(_SlideDown);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _Menu = require('./Menu');

var _Menu2 = _interopRequireDefault(_Menu);

var _DropdownButton = require('./styles/DropdownButton.css');

var _DropdownButton2 = _interopRequireDefault(_DropdownButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['setIsMenuShow'];

var DropdownButton = (_dec = (0, _withState2.default)('isMenuShow', 'setIsMenuShow', false), (0, _pure2.default)(_class = _dec(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(DropdownButton, _React$Component);

  function DropdownButton() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, DropdownButton);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = DropdownButton.__proto__ || Object.getPrototypeOf(DropdownButton)).call.apply(_ref, [this].concat(args))), _this), _this.onCloseMenu = function () {
      return _this.props.setIsMenuShow(false);
    }, _this.onToggleMenu = function () {
      return _this.props.setIsMenuShow(!_this.props.isMenuShow);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(DropdownButton, [{
    key: 'onSelectChange',
    value: function onSelectChange(e, value) {
      this.props.onChange(e, value);
      this.onCloseMenu();
    }
  }, {
    key: 'onButtonClick',
    value: function onButtonClick(e) {
      this.props.onClick(e);
      this.onToggleMenu();
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          items = _props.items,
          selectedValue = _props.selectedValue,
          onChange = _props.onChange,
          button = _props.button,
          children = _props.children,
          className = _props.className,
          isMenuShow = _props.isMenuShow,
          buttonProps = _props.buttonProps,
          menuClassName = _props.menuClassName,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['items', 'selectedValue', 'onChange', 'button', 'children', 'className', 'isMenuShow', 'buttonProps', 'menuClassName']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), {
          className: (0, _classnames2.default)(_DropdownButton2.default.wrap, className) }),
        _react2.default.createElement(_Button2.default, (0, _extends3.default)({
          kind: 'cancel',
          onClick: this.onButtonClick.bind(this),
          active: isMenuShow,
          children: children || button
        }, buttonProps)),
        (0, _jsx3.default)(_reactOverlays.Overlay, {
          show: isMenuShow,
          onHide: this.onCloseMenu.bind(this),
          rootClose: true,
          placement: 'bottom',
          container: document.body,
          target: function target() {
            return (0, _reactDom.findDOMNode)(_this2);
          },
          transition: _SlideDown2.default
        }, void 0, (0, _jsx3.default)(_Menu2.default, {
          className: (0, _classnames2.default)(_DropdownButton2.default.menu, menuClassName),
          onChange: this.onSelectChange.bind(this),
          selectedValue: selectedValue,
          items: items
        }))
      );
    }
  }]);
  return DropdownButton;
}(_react2.default.Component), _class2.propTypes = {
  items: _react.PropTypes.arrayOf(_react.PropTypes.shape({
    value: _react.PropTypes.any,
    children: _react.PropTypes.node
  })),
  selectedValue: _react.PropTypes.any,
  onChange: _react.PropTypes.func,
  className: _react.PropTypes.string,
  onClick: _react.PropTypes.func,
  isMenuShow: _react.PropTypes.bool.isRequired,
  setIsMenuShow: _react.PropTypes.func.isRequired,
  menuClassName: _react.PropTypes.string,

  // button
  children: _react.PropTypes.node,
  buttonProps: _react.PropTypes.shape((0, _extends3.default)({}, _Button2.default.propTypes)),

  // deprecated
  button: _react.PropTypes.node
}, _class2.defaultProps = {
  onClick: _emptyFunction2.default
}, _temp2)) || _class) || _class);
exports.default = DropdownButton;
module.exports = exports['default'];
