// mtk-ui/lib/ImageStream.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _MiError = require('mtk-icon/lib/MiError');

var _MiError2 = _interopRequireDefault(_MiError);

var _MiWifi = require('mtk-icon/lib/MiWifi');

var _MiWifi2 = _interopRequireDefault(_MiWifi);

var _MiVideocamOff = require('mtk-icon/lib/MiVideocamOff');

var _MiVideocamOff2 = _interopRequireDefault(_MiVideocamOff);

var _MiPlay = require('mtk-icon/lib/MiPlay');

var _MiPlay2 = _interopRequireDefault(_MiPlay);

var _MiPause = require('mtk-icon/lib/MiPause');

var _MiPause2 = _interopRequireDefault(_MiPause);

var _ImageStream = require('./styles/ImageStream.css');

var _ImageStream2 = _interopRequireDefault(_ImageStream);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['status', 'message'];

// TODO: i18n
var MESSAGES = {
  connecting: 'Connecting...',
  disconnected: 'Disconnected',
  error: 'Please check your wifi connection.'
};

var ImageStream = function ImageStream(_ref) {
  var children = _ref.children,
      className = _ref.className,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['children', 'className']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), { className: (0, _classnames2.default)(_ImageStream2.default.wrap, className) }),
    renderMask(otherProps),
    (0, _jsx3.default)('div', {
      className: _ImageStream2.default.contentWrap
    }, void 0, children)
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  status: 'pause',
  onPlay: _emptyFunction2.default,
  onPause: _emptyFunction2.default
}), (0, _setPropTypes3.default)({
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  status: _react.PropTypes.oneOf(['connecting', 'disconnected', 'error', 'playing', 'pause']),
  onPlay: _react.PropTypes.func,
  onPause: _react.PropTypes.func,
  message: _react.PropTypes.string
}))(ImageStream);

/**
 * Render Mask
 * @param {object} props - same as PropTypes
 */

function renderMask(_ref2) {
  var status = _ref2.status,
      message = _ref2.message,
      onPlay = _ref2.onPlay,
      onPause = _ref2.onPause;

  var childContent = void 0;

  switch (status) {
    case 'connecting':
      childContent = (0, _jsx3.default)('div', {
        className: _ImageStream2.default.messageWrap
      }, void 0, (0, _jsx3.default)(_MiWifi2.default, {
        className: _ImageStream2.default.statusIcon
      }), (0, _jsx3.default)('div', {}, void 0, message || MESSAGES.connecting));
      break;
    case 'disconnected':
      childContent = (0, _jsx3.default)('div', {
        className: _ImageStream2.default.messageWrap
      }, void 0, (0, _jsx3.default)(_MiVideocamOff2.default, {
        className: _ImageStream2.default.statusIcon
      }), (0, _jsx3.default)('div', {}, void 0, message || MESSAGES.disconnected));
      break;
    case 'error':
      childContent = (0, _jsx3.default)('div', {
        className: _ImageStream2.default.messageWrap
      }, void 0, (0, _jsx3.default)(_MiError2.default, {
        className: _ImageStream2.default.statusIcon
      }), (0, _jsx3.default)('div', {}, void 0, message || MESSAGES.error));
      break;
    case 'pause':
      childContent = (0, _jsx3.default)('div', {
        className: _ImageStream2.default.playMask
      }, void 0, (0, _jsx3.default)('div', {
        className: _ImageStream2.default.buttonWrap
      }, void 0, (0, _jsx3.default)(_MiPlay2.default, {
        className: _ImageStream2.default.playerButton,
        onClick: onPlay
      })));
      break;
    case 'playing':
    default:
      childContent = (0, _jsx3.default)('div', {
        className: _ImageStream2.default.pauseMask
      }, void 0, (0, _jsx3.default)('div', {
        className: _ImageStream2.default.buttonWrap
      }, void 0, (0, _jsx3.default)(_MiPause2.default, {
        className: _ImageStream2.default.playerButton,
        onClick: onPause
      })));
      break;
  }

  return childContent;
}
module.exports = exports['default'];
