// mtk-ui/lib/InputRadio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Label = require('./Label');

var _Label2 = _interopRequireDefault(_Label);

var _InputRadio = require('./styles/InputRadio.css');

var _InputRadio2 = _interopRequireDefault(_InputRadio);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var InputRadio = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(InputRadio, _React$Component);

  function InputRadio() {
    (0, _classCallCheck3.default)(this, InputRadio);
    return (0, _possibleConstructorReturn3.default)(this, (InputRadio.__proto__ || Object.getPrototypeOf(InputRadio)).apply(this, arguments));
  }

  (0, _createClass3.default)(InputRadio, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          label = _props.label,
          style = _props.style,
          className = _props.className,
          selectedValue = _props.selectedValue,
          value = _props.value,
          labelProps = _props.labelProps,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['label', 'style', 'className', 'selectedValue', 'value', 'labelProps']);


      return (0, _jsx3.default)('label', {
        style: style,
        className: (0, _classnames2.default)(_InputRadio2.default.wrap, className)
      }, void 0, _react2.default.createElement('input', (0, _extends3.default)({}, otherProps, {
        className: _InputRadio2.default.input,
        checked: selectedValue === value,
        value: value,
        type: 'radio' })), (0, _jsx3.default)('svg', {
        viewBox: '0 0 14 14',
        className: _InputRadio2.default.checkbox
      }, void 0, (0, _jsx3.default)('circle', {
        cx: '7',
        cy: '7',
        r: '6',
        className: _InputRadio2.default.box
      }), (0, _jsx3.default)('circle', {
        cx: '7',
        cy: '7',
        r: '6',
        className: _InputRadio2.default.check
      })), label && _react2.default.createElement(
        _Label2.default,
        labelProps,
        label
      ));
    }
  }]);
  return InputRadio;
}(_react2.default.Component), _class2.propTypes = {
  style: _react.PropTypes.object,
  className: _react.PropTypes.string,
  selectedValue: _react.PropTypes.oneOfType([_react.PropTypes.object, _react.PropTypes.string, _react.PropTypes.number]),
  value: _react.PropTypes.oneOfType([_react.PropTypes.object, _react.PropTypes.string, _react.PropTypes.number]),
  label: _react.PropTypes.node,
  labelProps: _react.PropTypes.shape((0, _extends3.default)({}, _Label2.default.propTypes))
}, _temp)) || _class;

exports.default = InputRadio;
module.exports = exports['default'];
