// mtk-ui/lib/InputCheckbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Label = require('./Label');

var _Label2 = _interopRequireDefault(_Label);

var _InputCheckbox = require('./styles/InputCheckbox.css');

var _InputCheckbox2 = _interopRequireDefault(_InputCheckbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var InputCheckbox = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(InputCheckbox, _React$Component);

  function InputCheckbox() {
    (0, _classCallCheck3.default)(this, InputCheckbox);
    return (0, _possibleConstructorReturn3.default)(this, (InputCheckbox.__proto__ || Object.getPrototypeOf(InputCheckbox)).apply(this, arguments));
  }

  (0, _createClass3.default)(InputCheckbox, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          label = _props.label,
          style = _props.style,
          className = _props.className,
          checked = _props.checked,
          labelProps = _props.labelProps,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['label', 'style', 'className', 'checked', 'labelProps']);


      return (0, _jsx3.default)('label', {
        style: style,
        className: (0, _classnames2.default)(_InputCheckbox2.default.wrap, className)
      }, void 0, _react2.default.createElement('input', (0, _extends3.default)({}, otherProps, {
        className: _InputCheckbox2.default.input,
        checked: checked,
        type: 'checkbox' })), (0, _jsx3.default)('svg', {
        viewBox: '0 0 140 140',
        className: _InputCheckbox2.default.checkbox
      }, void 0, (0, _jsx3.default)('rect', {
        className: _InputCheckbox2.default.box,
        x: '10',
        y: '10',
        width: '120',
        height: '120',
        rx: '20'
      }), (0, _jsx3.default)('path', {
        className: _InputCheckbox2.default.check,
        d: 'M62.2139985,103.499997 L57.3818988,97.6411076 L30,64.498635\n              L38.9202746,54.4710496 L61.1151736,71.4932266 L109.627271,30\n              L118.656746,40.000863 L67.2508483,97.8482063 L62.2139985,103.499997 Z'
      })), label && _react2.default.createElement(
        _Label2.default,
        labelProps,
        label
      ));
    }
  }]);
  return InputCheckbox;
}(_react2.default.Component), _class2.propTypes = {
  style: _react.PropTypes.object,
  className: _react.PropTypes.string,
  checked: _react.PropTypes.bool,
  label: _react.PropTypes.node,
  labelProps: _react.PropTypes.shape((0, _extends3.default)({}, _Label2.default.propTypes))
}, _temp)) || _class;

exports.default = InputCheckbox;
module.exports = exports['default'];
