// mtk-ui/lib/InputTypeAhead.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp; // This component is derived from hellojwilde/react-pick
// https://github.com/hellojwilde/react-pick

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _InputText = require('./InputText');

var _InputText2 = _interopRequireDefault(_InputText);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var KEY_BACKSPACE = 8;
var KEY_RIGHT = 39;
var KEY_LEFT = 37;
var KEY_ENTER = 13;

function getTypeahead(value, typeaheadValue) {
  var normalizedValue = (value || '').toLowerCase();
  var normalizedTypeaheadValue = (typeaheadValue || '').toLowerCase();

  if (normalizedValue === '' || normalizedValue === normalizedTypeaheadValue) {
    return null;
  }
  if (normalizedTypeaheadValue.indexOf(normalizedValue) !== 0) {
    return null;
  }

  var start = value.length;
  var end = typeaheadValue.length;
  var valueWithTypeahead = value + typeaheadValue.slice(start, end);

  return { valueWithTypeahead: valueWithTypeahead, start: start, end: end };
}

var InputTypeAhead = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(InputTypeAhead, _React$Component);

  function InputTypeAhead() {
    (0, _classCallCheck3.default)(this, InputTypeAhead);
    return (0, _possibleConstructorReturn3.default)(this, (InputTypeAhead.__proto__ || Object.getPrototypeOf(InputTypeAhead)).apply(this, arguments));
  }

  (0, _createClass3.default)(InputTypeAhead, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps) {
      if (this.props.typeaheadValue !== prevProps.typeaheadValue) {
        this.updateTypeahead();
      }
    }
  }, {
    key: 'onBlur',
    value: function onBlur(e) {
      this.props.onChange(e);
    }
  }, {
    key: 'onKeyUp',
    value: function onKeyUp(e) {
      this.isTypingForward = ![KEY_BACKSPACE, KEY_LEFT].includes(e.keyCode);
      this.updateTypeahead(e);

      if (this.props.onKeyUp) {
        this.props.onKeyUp(e);
      }
    }
  }, {
    key: 'updateTypeahead',
    value: function updateTypeahead(e) {
      if (this.props.typeaheadValue === null || this.isTypingForward !== true) {
        if (e.keyCode === KEY_LEFT) {
          this.props.onChange(e);
        }
        return;
      }

      var input = (0, _reactDom.findDOMNode)(this.refs.input).getElementsByTagName('input')[0];
      var _props = this.props,
          value = _props.value,
          typeaheadValue = _props.typeaheadValue;

      var typeahead = getTypeahead(value, typeaheadValue);

      if (typeahead === null) return;

      // update view value
      input.value = typeahead.valueWithTypeahead;
      input.setSelectionRange(typeahead.start, typeahead.end);

      if ([KEY_RIGHT, KEY_ENTER].includes(e.keyCode)) {
        e.target.value = typeahead.valueWithTypeahead;
        this.props.onChange(e);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          typeaheadValue = _props2.typeaheadValue,
          onKeyUp = _props2.onKeyUp,
          otherProps = (0, _objectWithoutProperties3.default)(_props2, ['typeaheadValue', 'onKeyUp']);


      return _react2.default.createElement(_InputText2.default, (0, _extends3.default)({}, otherProps, { onKeyUp: this.onKeyUp.bind(this), onBlur: this.onBlur.bind(this), ref: 'input' }));
    }
  }]);
  return InputTypeAhead;
}(_react2.default.Component), _class2.propTypes = {
  typeaheadValue: _react.PropTypes.string,
  value: _react.PropTypes.string,
  onKeyUp: _react.PropTypes.func,
  onChange: _react.PropTypes.func.isRequired,
  onBlur: _react.PropTypes.func
}, _temp)) || _class;

exports.default = InputTypeAhead;
module.exports = exports['default'];
