// mtk-ui/lib/InputDate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _withPropsOnChange2 = require('recompose/withPropsOnChange');

var _withPropsOnChange3 = _interopRequireDefault(_withPropsOnChange2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _withHandlers2 = require('recompose/withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _defaultPropHandler = require('./decorators/defaultPropHandler');

var _defaultPropHandler2 = _interopRequireDefault(_defaultPropHandler);

var _reactCssThemr = require('react-css-themr');

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _Calendar = require('./Calendar');

var _Calendar2 = _interopRequireDefault(_Calendar);

var _MiArrowLeft = require('mtk-icon/lib/MiArrowLeft');

var _MiArrowLeft2 = _interopRequireDefault(_MiArrowLeft);

var _MiArrowRight = require('mtk-icon/lib/MiArrowRight');

var _MiArrowRight2 = _interopRequireDefault(_MiArrowRight);

var _clampDate = require('./utils/clampDate');

var _clampDate2 = _interopRequireDefault(_clampDate);

var _InputDate = require('./styles/InputDate.css');

var _InputDate2 = _interopRequireDefault(_InputDate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Header = (0, _pure3.default)(function (_ref) {
  var max = _ref.max,
      min = _ref.min,
      currentPageDate = _ref.currentPageDate,
      locale = _ref.locale,
      addOneMonth = _ref.addOneMonth,
      subtractOneMonth = _ref.subtractOneMonth,
      theme = _ref.theme;

  _moment2.default.locale(locale);
  var monthNames = _moment2.default.months();

  var isLowerReach = (0, _moment2.default)(min).isSame(currentPageDate, 'month');
  var isUpperReach = (0, _moment2.default)(max).isSame(currentPageDate, 'month');

  return (0, _jsx3.default)('div', {
    className: theme.header
  }, void 0, (0, _jsx3.default)('div', {
    onClick: isLowerReach ? undefined : subtractOneMonth
  }, void 0, (0, _jsx3.default)(_MiArrowLeft2.default, {
    size: 24,
    className: isLowerReach ? theme.arrowDisable : theme.arrow
  })), (0, _jsx3.default)('span', {}, void 0, monthNames[currentPageDate.get('month')], '\xA0', currentPageDate.get('year')), (0, _jsx3.default)('div', {
    onClick: isUpperReach ? undefined : addOneMonth
  }, void 0, (0, _jsx3.default)(_MiArrowRight2.default, {
    size: 24,
    className: isUpperReach ? theme.arrowDisable : theme.arrow
  })));
});

var InputDate = function InputDate(_ref2) {
  var max = _ref2.max,
      min = _ref2.min,
      className = _ref2.className,
      currentValue = _ref2.currentValue,
      currentPageDate = _ref2.currentPageDate,
      locale = _ref2.locale,
      addOneMonth = _ref2.addOneMonth,
      subtractOneMonth = _ref2.subtractOneMonth,
      theme = _ref2.theme,
      calendarTheme = _ref2.calendarTheme,
      dayTheme = _ref2.dayTheme,
      otherProps = (0, _objectWithoutProperties3.default)(_ref2, ['max', 'min', 'className', 'currentValue', 'currentPageDate', 'locale', 'addOneMonth', 'subtractOneMonth', 'theme', 'calendarTheme', 'dayTheme']);
  return (0, _jsx3.default)('div', {
    className: className
  }, void 0, (0, _jsx3.default)(Header, {
    locale: locale,
    max: max,
    min: min,
    addOneMonth: addOneMonth,
    subtractOneMonth: subtractOneMonth,
    currentPageDate: (0, _moment2.default)(currentPageDate),
    theme: theme
  }), _react2.default.createElement(_Calendar2.default, (0, _extends3.default)({
    locale: locale,
    max: max,
    min: min,
    theme: calendarTheme,
    dayTheme: dayTheme,
    selectedValue: currentValue,
    currentPageDate: (0, _moment2.default)(currentPageDate)
  }, (0, _omit3.default)(['onPageChange'], otherProps))));
};

exports.default = (0, _compose3.default)((0, _reactCssThemr.themr)('InputDate', _InputDate2.default), _pure3.default, (0, _defaultProps3.default)({
  max: (0, _moment2.default)(),
  min: (0, _moment2.default)('2000/01/01', 'YYYY/MM/DD')
}), (0, _defaultPropHandler2.default)('currentPageDate', 'onPageChange', function (setCurrentPageDate) {
  return function (val) {
    return setCurrentPageDate(val);
  };
}, function (_ref3) {
  var currentValue = _ref3.currentValue,
      max = _ref3.max,
      min = _ref3.min;
  return (0, _clampDate2.default)({
    currentValue: (0, _moment2.default)(currentValue),
    max: max,
    min: min
  });
}), (0, _withHandlers3.default)({
  addOneMonth: function addOneMonth(props) {
    return function () {
      return props.onPageChange((0, _moment2.default)(props.currentPageDate).add(1, 'M'));
    };
  },
  subtractOneMonth: function subtractOneMonth(props) {
    return function () {
      return props.onPageChange((0, _moment2.default)(props.currentPageDate).subtract(1, 'M'));
    };
  }
}), (0, _withPropsOnChange3.default)(['max', 'min', 'currentValue'], function (_ref4) {
  var currentValue = _ref4.currentValue,
      max = _ref4.max,
      min = _ref4.min;
  return {
    currentValue: (0, _clampDate2.default)({
      currentValue: (0, _moment2.default)(currentValue),
      max: max,
      min: min
    })
  };
}), (0, _setPropTypes3.default)({
  max: _react.PropTypes.object.isRequired,
  min: _react.PropTypes.object.isRequired,
  currentValue: _react.PropTypes.object.isRequired,
  onChange: _react.PropTypes.func.isRequired,
  onPageChange: _react.PropTypes.func,
  locale: _react.PropTypes.string,
  weekNames: _react.PropTypes.arrayOf(_react.PropTypes.string),
  className: _react.PropTypes.string,
  disabledDays: _react.PropTypes.arrayOf(_react.PropTypes.string),
  inactiveDays: _react.PropTypes.arrayOf(_react.PropTypes.string),
  theme: _react.PropTypes.shape({
    header: _react.PropTypes.string,
    arrow: _react.PropTypes.string,
    arrowDisable: _react.PropTypes.string
  }),
  calendarTheme: _react.PropTypes.shape({
    base: _react.PropTypes.string,
    weekName: _react.PropTypes.string,
    sundayName: _react.PropTypes.string
  }),
  dayTheme: _react.PropTypes.shape({
    base: _react.PropTypes.string,
    today: _react.PropTypes.string,
    selected: _react.PropTypes.string,
    disabled: _react.PropTypes.string,
    enabled: _react.PropTypes.string,
    inactive: _react.PropTypes.string
  })
}))(InputDate);
module.exports = exports['default'];
