// mtk-ui/lib/InputValidation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputValidation = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _withPropsOnChange2 = require('recompose/withPropsOnChange');

var _withPropsOnChange3 = _interopRequireDefault(_withPropsOnChange2);

var _withHandlers2 = require('recompose/withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _withState2 = require('recompose/withState');

var _withState3 = _interopRequireDefault(_withState2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _InputText = require('./InputText');

var _InputText2 = _interopRequireDefault(_InputText);

var _LoadingMorph = require('./LoadingMorph');

var _LoadingMorph2 = _interopRequireDefault(_LoadingMorph);

var _debounce = require('lodash/debounce');

var _debounce2 = _interopRequireDefault(_debounce);

var _InputValidation = require('./styles/InputValidation.css');

var _InputValidation2 = _interopRequireDefault(_InputValidation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var testCallbacks = function testCallbacks(props, value) {
  return {
    success: function success() {
      if (!!value) {
        props.setStatus('success');
        props.setError('');
        props.onSuccess(value);
      }
    },
    fail: function fail(error) {
      props.setStatus('fail');
      props.setError(error);
      props.onFail(value, error);
    },
    loading: function loading() {
      if (!!value) {
        if (props.status !== 'default') return; // avoid animate twice

        props.setStatus('loading');
        props.setError('');
        props.onLoading(value);
      }
    }
  };
};

var InputValidation = exports.InputValidation = function InputValidation(_ref) {
  var status = _ref.status,
      onChange = _ref.onChange,
      onBlur = _ref.onBlur,
      error = _ref.error,
      controlError = _ref.controlError,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['status', 'onChange', 'onBlur', 'error', 'controlError']);

  var controlStatus = controlError ? 'fail' : status;

  return _react2.default.createElement(_InputText2.default, (0, _extends3.default)({}, otherProps, {
    onChange: onChange,
    onBlur: onBlur,
    className: _InputValidation2.default['input' + controlStatus],
    icon: (0, _jsx3.default)(_LoadingMorph2.default, {
      status: controlStatus,
      size: 18,
      className: _InputValidation2.default['icon' + controlStatus]
    }),
    iconProps: { placement: 'right' },
    error: error || controlError
  }));
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _setPropTypes3.default)({
  validations: _react.PropTypes.func.isRequired, // (value, t) => voild
  debounceWait: _react.PropTypes.number, // millisecond
  onChange: _react.PropTypes.func, // (e) => voild
  onBlur: _react.PropTypes.func, // (e) => voild
  onSuccess: _react.PropTypes.func, // (value) => voild
  onFail: _react.PropTypes.func, // (value, error) => voild
  onLoading: _react.PropTypes.func, // (value) => voild
  defaultStatus: _react.PropTypes.oneOf(['default', 'success', 'fail', 'loading']),
  controlError: _react.PropTypes.any }), (0, _defaultProps3.default)({
  validations: _emptyFunction2.default,
  debounceWait: 1000,
  onChange: _emptyFunction2.default,
  onBlur: _emptyFunction2.default,
  onSuccess: _emptyFunction2.default,
  onFail: _emptyFunction2.default,
  onLoading: _emptyFunction2.default,
  defaultStatus: 'default'
}), (0, _withState3.default)('status', 'setStatus', function (props) {
  return props.defaultStatus;
}), (0, _withState3.default)('error', 'setError', ''), (0, _withPropsOnChange3.default)(['validations', 'debounceWait'], function (_ref2) {
  var validations = _ref2.validations,
      debounceWait = _ref2.debounceWait;
  return {
    debounceValidations: (0, _debounce2.default)(validations, debounceWait)
  };
}), (0, _withHandlers3.default)({
  onChange: function onChange(props) {
    return function (e) {
      props.onChange(e);
      props.setStatus('default');
      props.setError('');

      props.debounceValidations(e.target.value, testCallbacks(props, e.target.value));
    };
  },
  onBlur: function onBlur(props) {
    return function (e) {
      props.onBlur(e);

      if (props.status === 'default') {
        props.validations(e.target.value, testCallbacks(props, e.target.value));
      }
    };
  }
}))(InputValidation);
