// mtk-ui/lib/InputForm.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputForm = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Label = require('./Label');

var _Label2 = _interopRequireDefault(_Label);

var _snakeCase = require('lodash/snakeCase');

var _snakeCase2 = _interopRequireDefault(_snakeCase);

var _InputForm = require('./styles/InputForm.css');

var _InputForm2 = _interopRequireDefault(_InputForm);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var labelFactory = (0, _react.createFactory)(_Label2.default);

var InputForm = exports.InputForm = function InputForm(_ref) {
  var children = _ref.children,
      className = _ref.className,
      labelPropsFromContainer = _ref.labelProps,
      kind = _ref.kind,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['children', 'className', 'labelProps', 'kind']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, otherProps, {
      className: className
    }),
    _react.Children.map(children, function (child, index) {
      var id = child.props.id || child.props.label && (0, _snakeCase2.default)(child.props.label) || index;

      return (0, _jsx3.default)('label', {
        htmlFor: id,
        className: _InputForm2.default['formWrap' + kind]
      }, id, child.props.label && labelFactory((0, _extends3.default)({}, child.props.labelProps, labelPropsFromContainer, {
        children: (0, _jsx3.default)('span', {
          className: _InputForm2.default.lableContent
        }, void 0, child.props.label),
        required: child.props.label && child.props.required,
        className: (0, _classnames2.default)(_InputForm2.default.label, labelPropsFromContainer.className)
      })), (0, _react.cloneElement)(child, (0, _extends3.default)({
        // Override child's default label and labelProps
        label: ''
      }, child.props.labelProps && { labelProps: {} }, {
        className: (0, _classnames2.default)(_InputForm2.default.input, child.props.className),
        id: id
      })));
    })
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _setPropTypes3.default)({
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  kind: _react.PropTypes.oneOf(['horizontal', 'vertical']),
  labelProps: _react.PropTypes.object
}), (0, _defaultProps3.default)({
  kind: 'horizontal',
  labelProps: {}
}))(InputForm);
