// mtk-ui/lib/InputTime.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _InputNumber = require('./InputNumber');

var _InputNumber2 = _interopRequireDefault(_InputNumber);

var _InputTime = require('./styles/InputTime.css');

var _InputTime2 = _interopRequireDefault(_InputTime);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var wrappedInputNumberPropTypes = {
  min: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]),
  max: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]),
  value: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]),
  error: _react.PropTypes.oneOfType([_react.PropTypes.bool, _react.PropTypes.node]),
  onChange: _react.PropTypes.func,
  placeholder: _react.PropTypes.string,
  defaultValue: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string])
};

var WrappedInputNumber = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  placeholder: '',
  defaultValue: ''
}), (0, _setPropTypes3.default)(wrappedInputNumberPropTypes))(function (_ref) {
  var props = (0, _objectWithoutProperties3.default)(_ref, []);
  return _react2.default.createElement(_InputNumber2.default, (0, _extends3.default)({}, props, {
    fitOnBlur: true,
    step: '1',
    className: _InputTime2.default.inputNumber,
    leftpad: 2
  }));
});

var InputTime = function InputTime(_ref2) {
  var hourProps = _ref2.hourProps,
      minuteProps = _ref2.minuteProps,
      className = _ref2.className,
      otherProps = (0, _objectWithoutProperties3.default)(_ref2, ['hourProps', 'minuteProps', 'className']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, otherProps, { className: (0, _classnames2.default)(_InputTime2.default.base, className) }),
    _react2.default.createElement(WrappedInputNumber, (0, _extends3.default)({
      min: 0,
      max: 23
    }, hourProps)),
    (0, _jsx3.default)('div', {}, void 0, ':'),
    _react2.default.createElement(WrappedInputNumber, (0, _extends3.default)({
      min: 0,
      max: 59
    }, minuteProps))
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _setPropTypes3.default)({
  hourProps: _react.PropTypes.shape(wrappedInputNumberPropTypes),
  minuteProps: _react.PropTypes.shape(wrappedInputNumberPropTypes),
  className: _react.PropTypes.string
}))(InputTime);
module.exports = exports['default'];
