// mtk-ui/lib/Label.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _defaultProps = require('recompose/defaultProps');

var _defaultProps2 = _interopRequireDefault(_defaultProps);

var _Label = require('./styles/Label.css');

var _Label2 = _interopRequireDefault(_Label);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Label = (_dec = (0, _defaultProps2.default)({
  kind: 'default',
  required: false
}), (0, _pure2.default)(_class = _dec(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Label, _React$Component);

  function Label() {
    (0, _classCallCheck3.default)(this, Label);
    return (0, _possibleConstructorReturn3.default)(this, (Label.__proto__ || Object.getPrototypeOf(Label)).apply(this, arguments));
  }

  (0, _createClass3.default)(Label, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          kind = _props.kind,
          required = _props.required,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'className', 'kind', 'required']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, otherProps, {
          className: (0, _classnames2.default)(_Label2.default[kind], required && _Label2.default.required, className)
        }),
        children
      );
    }
  }]);
  return Label;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  kind: _react.PropTypes.oneOf(['error', 'success', 'default']),
  required: _react.PropTypes.bool
}, _temp)) || _class) || _class);
exports.default = Label;
module.exports = exports['default'];
