// mtk-ui/lib/LoadingMorph.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactSvgMorph = require('react-svg-morph');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _MiCircleLoading = require('mtk-icon/lib/MiCircleLoading');

var _MiCircleLoading2 = _interopRequireDefault(_MiCircleLoading);

var _MiCheck = require('mtk-icon/lib/MiCheck');

var _MiCheck2 = _interopRequireDefault(_MiCheck);

var _MiFail = require('mtk-icon/lib/MiFail');

var _MiFail2 = _interopRequireDefault(_MiFail);

var _LoadingMorph = require('./styles/LoadingMorph.css');

var _LoadingMorph2 = _interopRequireDefault(_LoadingMorph);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LoadingMorph = function LoadingMorph(_ref) {
  var className = _ref.className,
      status = _ref.status,
      size = _ref.size,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['className', 'status', 'size']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, otherProps, {
      className: (0, _classnames2.default)(status === 'loading' ? _LoadingMorph2.default.spin : _LoadingMorph2.default.default, className)
    }),
    !status || status === 'default' ? null : (0, _jsx3.default)(_reactSvgMorph.MorphReplace, {
      width: size,
      height: size
    }, void 0, getIcon(status))
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({ size: 24, status: 'loading' }), (0, _setPropTypes3.default)({
  status: _react.PropTypes.oneOf(['success', 'fail', 'loading', 'default']),
  size: _react.PropTypes.number
}))(LoadingMorph);


function getIcon(status) {
  switch (status) {
    case 'success':
      return (0, _jsx3.default)(_MiCheck2.default, {}, 'success');
    case 'fail':
      return (0, _jsx3.default)(_MiFail2.default, {}, 'fail');
    case 'loading':
    default:
      return (0, _jsx3.default)(_MiCircleLoading2.default, {}, 'loading');
  }
}
module.exports = exports['default'];
