// mtk-ui/lib/MenuItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _reactCssThemr = require('react-css-themr');

var _getContext = require('recompose/getContext');

var _getContext2 = _interopRequireDefault(_getContext);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _MenuItem = require('./styles/MenuItem.css');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MenuItem = (_dec = (0, _getContext2.default)({
  selectedValue: _react.PropTypes.any,
  onSelect: _react.PropTypes.func
}), _dec2 = (0, _reactCssThemr.themr)('MenuItem', _MenuItem2.default), (0, _pure2.default)(_class = _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(MenuItem, _React$Component);

  function MenuItem() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, MenuItem);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = MenuItem.__proto__ || Object.getPrototypeOf(MenuItem)).call.apply(_ref, [this].concat(args))), _this), _this.onClick = function (e) {
      if (!_this.props.disabled) {
        _this.props.onClick(e, _this.props.value, _this.props.index);
        _this.props.onSelect(e, _this.props.value, _this.props.index);
      }
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(MenuItem, [{
    key: 'render',
    value: function render() {
      var _c;

      var _props = this.props,
          children = _props.children,
          active = _props.active,
          focus = _props.focus,
          hoverDisabled = _props.hoverDisabled,
          value = _props.value,
          selectedValue = _props.selectedValue,
          className = _props.className,
          theme = _props.theme,
          index = _props.index,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'active', 'focus', 'hoverDisabled', 'value', 'selectedValue', 'className', 'theme', 'index']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, otherProps, {
          onClick: this.onClick,
          className: (0, _classnames2.default)(_MenuItem2.default.base, (_c = {}, (0, _defineProperty3.default)(_c, theme.active, !this.props.disabled && (active || value === selectedValue)), (0, _defineProperty3.default)(_c, theme.focus, focus), (0, _defineProperty3.default)(_c, theme.disabled, this.props.disabled), (0, _defineProperty3.default)(_c, theme.hover, !hoverDisabled && !this.props.disabled), _c))
        }),
        children
      );
    }
  }]);
  return MenuItem;
}(_react2.default.Component), _class2.propTypes = {
  className: _react.PropTypes.string,
  children: _react.PropTypes.node,
  active: _react.PropTypes.bool,
  focus: _react.PropTypes.bool,
  selectedValue: _react.PropTypes.any,
  value: _react.PropTypes.any,
  disabled: _react.PropTypes.bool,
  hoverDisabled: _react.PropTypes.bool,
  index: _react.PropTypes.number,
  onClick: _react.PropTypes.func, // (e, value, index) => {}
  onSelect: _react.PropTypes.func, // (e, value, index) => {}
  theme: _react.PropTypes.object
}, _class2.defaultProps = {
  selectedValue: null,
  onClick: _emptyFunction2.default,
  onSelect: _emptyFunction2.default
}, _temp2)) || _class) || _class) || _class);
exports.default = MenuItem;
module.exports = exports['default'];
