// mtk-ui/lib/PopoverOverlay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _defaultProps = require('recompose/defaultProps');

var _defaultProps2 = _interopRequireDefault(_defaultProps);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactOverlays = require('react-overlays');

var _Popover = require('./Popover');

var _Popover2 = _interopRequireDefault(_Popover);

var _getOverlayTransition = require('./Transitions/getOverlayTransition');

var _getOverlayTransition2 = _interopRequireDefault(_getOverlayTransition);

var _PopoverOverlay = require('./styles/PopoverOverlay.css');

var _PopoverOverlay2 = _interopRequireDefault(_PopoverOverlay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PopoverOverlay = (_dec = (0, _defaultProps2.default)({
  placement: 'bottom'
}), (0, _pure2.default)(_class = _dec(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(PopoverOverlay, _React$Component);

  function PopoverOverlay() {
    (0, _classCallCheck3.default)(this, PopoverOverlay);
    return (0, _possibleConstructorReturn3.default)(this, (PopoverOverlay.__proto__ || Object.getPrototypeOf(PopoverOverlay)).apply(this, arguments));
  }

  (0, _createClass3.default)(PopoverOverlay, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          cardClassName = _props.cardClassName,
          arrowClassName = _props.arrowClassName,
          placement = _props.placement,
          style = _props.style,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'className', 'cardClassName', 'arrowClassName', 'placement', 'style']);


      return _react2.default.createElement(
        _reactOverlays.Overlay,
        (0, _extends3.default)({}, otherProps, {
          placement: placement,
          transition: (0, _getOverlayTransition2.default)(placement)
        }),
        (0, _jsx3.default)(_Popover2.default, {
          placement: placement,
          style: style,
          className: (0, _classnames2.default)(_PopoverOverlay2.default.popover, _PopoverOverlay2.default[placement], className),
          cardClassName: cardClassName,
          arrowClassName: arrowClassName
        }, void 0, children)
      );
    }
  }]);
  return PopoverOverlay;
}(_react2.default.Component), _class2.propTypes = (0, _extends3.default)({}, _reactOverlays.Overlay.PropTypes, {
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  cardClassName: _react.PropTypes.string,
  arrowClassName: _react.PropTypes.string,
  placement: _react.PropTypes.string,
  style: _react.PropTypes.object
}), _temp)) || _class) || _class);
exports.default = PopoverOverlay;
module.exports = exports['default'];
