// mtk-ui/lib/Popover.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Card = require('./Card');

var _Card2 = _interopRequireDefault(_Card);

var _Arrow = require('./Arrow');

var _Arrow2 = _interopRequireDefault(_Arrow);

var _Popover = require('./styles/Popover.css');

var _Popover2 = _interopRequireDefault(_Popover);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Popover = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Popover, _React$Component);

  function Popover() {
    (0, _classCallCheck3.default)(this, Popover);
    return (0, _possibleConstructorReturn3.default)(this, (Popover.__proto__ || Object.getPrototypeOf(Popover)).apply(this, arguments));
  }

  (0, _createClass3.default)(Popover, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          cardClassName = _props.cardClassName,
          arrowClassName = _props.arrowClassName,
          placement = _props.placement,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'className', 'cardClassName', 'arrowClassName', 'placement']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, otherProps, { className: (0, _classnames2.default)(_Popover2.default[placement], className) }),
        (0, _jsx3.default)(_Card2.default, {
          className: (0, _classnames2.default)(_Popover2.default.card, cardClassName)
        }, void 0, children),
        (0, _jsx3.default)('div', {
          className: (0, _classnames2.default)(_Popover2.default.arrowWrap, arrowClassName)
        }, void 0, (0, _jsx3.default)(_Arrow2.default, {
          placement: placement
        }))
      );
    }
  }]);
  return Popover;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  cardClassName: _react.PropTypes.string,
  arrowClassName: _react.PropTypes.string,
  placement: _react.PropTypes.string.isRequired
}, _class2.defaultProps = {
  placement: 'bottom'
}, _temp)) || _class;

exports.default = Popover;
module.exports = exports['default'];
