// mtk-ui/lib/ProgressCircle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _WithTooltip = require('./decorators/WithTooltip');

var _WithTooltip2 = _interopRequireDefault(_WithTooltip);

var _Arc = require('./Arc');

var _Arc2 = _interopRequireDefault(_Arc);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ProgressCircle = (0, _pureRenderDecorator2.default)(_class = (0, _WithTooltip2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(ProgressCircle, _React$Component);

  function ProgressCircle() {
    (0, _classCallCheck3.default)(this, ProgressCircle);
    return (0, _possibleConstructorReturn3.default)(this, (ProgressCircle.__proto__ || Object.getPrototypeOf(ProgressCircle)).apply(this, arguments));
  }

  (0, _createClass3.default)(ProgressCircle, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        isChild: true
      };
    }
  }, {
    key: 'renderBaseCircle',
    value: function renderBaseCircle() {
      return (0, _jsx3.default)(ProgressCircle, {});
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'className']);


      if (this.context.isChild) {
        return _react2.default.createElement(_Arc2.default, this.props);
      }

      var haveChildren = _react2.default.Children.toArray(children).some(function (child) {
        return !!child;
      });

      return _react2.default.createElement(
        'svg',
        (0, _extends3.default)({}, (0, _omit3.default)(['now'], this.props), { viewBox: '0 0 100 100', className: className }),
        this.renderBaseCircle(),
        haveChildren ? children : _react2.default.createElement(_Arc2.default, this.props)
      );
    }
  }]);
  return ProgressCircle;
}(_react2.default.Component), _class2.propTypes = {
  className: _react.PropTypes.string,
  children: _react.PropTypes.node,
  now: _react.PropTypes.number,
  max: _react.PropTypes.number,
  offset: _react.PropTypes.number,
  stroke: _react.PropTypes.string
}, _class2.contextTypes = {
  isChild: _react.PropTypes.bool
}, _class2.childContextTypes = {
  isChild: _react.PropTypes.bool
}, _class2.defaultProps = {
  now: 100,
  max: 100,
  offset: 0,
  radius: 48,
  stroke: '#F1F2F7', // default gray
  strokeWidth: 4
}, _temp)) || _class) || _class;

exports.default = ProgressCircle;
module.exports = exports['default'];
