// mtk-ui/lib/Progressbar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _withPropsOnChange2 = require('recompose/withPropsOnChange');

var _withPropsOnChange3 = _interopRequireDefault(_withPropsOnChange2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _isFunction2 = require('lodash/isFunction');

var _isFunction3 = _interopRequireDefault(_isFunction2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Progressbar = require('./styles/Progressbar.css');

var _Progressbar2 = _interopRequireDefault(_Progressbar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['now'];

var Progressbar = function Progressbar(_ref) {
  var label = _ref.label,
      percentage = _ref.percentage,
      className = _ref.className,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['label', 'percentage', 'className']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), { className: (0, _classnames2.default)(_Progressbar2.default.wrap, className) }),
    (0, _jsx3.default)('div', {
      style: { width: percentage + '%' },
      className: (0, _classnames2.default)(_Progressbar2.default.bar, _Progressbar2.default[getProgressClassName(percentage)])
    }, void 0, getLabel(label, percentage))
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _defaultProps3.default)({
  now: 0,
  max: 100,
  min: 0
}), (0, _setPropTypes3.default)({
  className: _react.PropTypes.string,
  now: _react.PropTypes.number,
  max: _react.PropTypes.number,
  min: _react.PropTypes.number,
  label: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.func])
}), (0, _withPropsOnChange3.default)(['min', 'max', 'now'], function (props) {
  return {
    percentage: getPercentage(props)
  };
}))(Progressbar);


function getPercentage(_ref2) {
  var min = _ref2.min,
      max = _ref2.max,
      now = _ref2.now;

  return now / (max - min) * 100;
}

function getProgressClassName(percentage) {
  if (percentage <= 25) return 'progress25';else if (percentage <= 50) return 'progress50';else if (percentage <= 75) return 'progress75';
  return 'progress100';
}

function getLabel(label, percentage) {
  if (!label) {
    return ~~percentage + '%';
  }

  if ((0, _isFunction3.default)(label)) {
    return label(percentage);
  }

  return label;
}
module.exports = exports['default'];
