// mtk-ui/lib/Select.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['labelProps'];

var Select = (0, _pure2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Select, _React$Component);

  function Select() {
    (0, _classCallCheck3.default)(this, Select);
    return (0, _possibleConstructorReturn3.default)(this, (Select.__proto__ || Object.getPrototypeOf(Select)).apply(this, arguments));
  }

  (0, _createClass3.default)(Select, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          onChange = _props.onChange,
          selectedValue = _props.selectedValue,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'onChange', 'selectedValue']);


      return _react2.default.createElement(
        'div',
        (0, _omit3.default)(omitList, otherProps),
        _react.Children.map(children, function (child, index) {
          return (0, _react.cloneElement)(child, {
            selectedValue: selectedValue,
            onClick: function onClick(e, value) {
              onChange(e, value, index);
            }
          });
        })
      );
    }
  }]);
  return Select;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node,
  selectedValue: _react.PropTypes.oneOfType([_react.PropTypes.object, _react.PropTypes.string, _react.PropTypes.number]),
  onChange: _react.PropTypes.func.isRequired }, _temp)) || _class;

exports.default = Select;
module.exports = exports['default'];
