// mtk-ui/lib/SortButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _withHandlers2 = require('recompose/withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _setPropTypes2 = require('recompose/setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _pure2 = require('recompose/pure');

var _pure3 = _interopRequireDefault(_pure2);

var _defaultProps2 = require('recompose/defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _compose2 = require('recompose/compose');

var _compose3 = _interopRequireDefault(_compose2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _emptyFunction = require('fbjs/lib/emptyFunction');

var _emptyFunction2 = _interopRequireDefault(_emptyFunction);

var _MiUnfold = require('mtk-icon/lib/MiUnfold');

var _MiUnfold2 = _interopRequireDefault(_MiUnfold);

var _MiFold = require('mtk-icon/lib/MiFold');

var _MiFold2 = _interopRequireDefault(_MiFold);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _defaultPropHandler = require('./decorators/defaultPropHandler');

var _defaultPropHandler2 = _interopRequireDefault(_defaultPropHandler);

var _SortButton = require('./styles/SortButton.css');

var _SortButton2 = _interopRequireDefault(_SortButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['threeStates', 'defaultStatus', 'setstatus'];

var SortButton = function SortButton(_ref) {
  var status = _ref.status,
      className = _ref.className,
      onClick = _ref.onClick,
      otherProps = (0, _objectWithoutProperties3.default)(_ref, ['status', 'className', 'onClick']);
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), { className: (0, _classnames2.default)(_SortButton2.default.wrap, className), onClick: onClick }),
    (0, _jsx3.default)(_MiUnfold2.default, {
      size: 12,
      className: (0, _classnames2.default)(status === 'desc' && _SortButton2.default.hidden)
    }),
    (0, _jsx3.default)(_MiFold2.default, {
      size: 12,
      className: (0, _classnames2.default)(status === 'asc' && _SortButton2.default.hidden, status === 'normal' && _SortButton2.default.miFold)
    })
  );
};

exports.default = (0, _compose3.default)(_pure3.default, (0, _setPropTypes3.default)({
  className: _react.PropTypes.string,
  status: _react.PropTypes.oneOf(['normal', 'asc', 'desc']),
  defaultStatus: _react.PropTypes.oneOf(['normal', 'asc', 'desc']),
  threeStates: _react.PropTypes.bool,
  onChange: _react.PropTypes.func
}), (0, _defaultPropHandler2.default)('status', 'onChange', function (setStatus) {
  return function (e, status) {
    return setStatus(status);
  };
}, function (_ref2) {
  var defaultStatus = _ref2.defaultStatus;
  return defaultStatus || 'normal';
}), (0, _defaultProps3.default)({
  onClick: _emptyFunction2.default,
  threeStates: false
}), (0, _withHandlers3.default)({
  onClick: function onClick(props) {
    return function (e) {
      props.onClick(e);
      switch (props.status) {
        case 'normal':
          props.onChange(e, 'asc');
          break;
        case 'asc':
          props.onChange(e, 'desc');
          break;
        case 'desc':
          props.onChange(e, props.threeStates ? 'normal' : 'asc');
          break;
        default:
          props.onChange(e, 'normal');
          break;
      }
    };
  }
}))(SortButton);
module.exports = exports['default'];
