// mtk-ui/lib/ToastContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _jsx2 = require('babel-runtime/helpers/jsx');

var _jsx3 = _interopRequireDefault(_jsx2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _omit2 = require('lodash/fp/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactAddonsCssTransitionGroup = require('react-addons-css-transition-group');

var _reactAddonsCssTransitionGroup2 = _interopRequireDefault(_reactAddonsCssTransitionGroup);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _reactOverlays = require('react-overlays');

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _uniqueId = require('lodash/uniqueId');

var _uniqueId2 = _interopRequireDefault(_uniqueId);

var _ToastContainer = require('./styles/ToastContainer.css');

var _ToastContainer2 = _interopRequireDefault(_ToastContainer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var omitList = ['timeout'];

var ToastContainer = (0, _pureRenderDecorator2.default)(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(ToastContainer, _React$Component);

  function ToastContainer() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ToastContainer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ToastContainer.__proto__ || Object.getPrototypeOf(ToastContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      /* eslint new-cap: 0 */
      toasts: _immutable2.default.OrderedMap()
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ToastContainer, [{
    key: 'notify',


    // This methis is usually used by the owner componet to create a notifying toast
    value: function notify(toastElement) {
      var _this2 = this;

      if (!toastElement) {
        return;
      }

      var id = (0, _uniqueId2.default)();

      this.setState({ toasts: this.state.toasts.set(id, toastElement) });

      setTimeout(function () {
        _this2.setState({ toasts: _this2.state.toasts.delete(id) });
      }, this.props.timeout);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          position = _props.position,
          className = _props.className,
          newestOnTop = _props.newestOnTop,
          container = _props.container,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['position', 'className', 'newestOnTop', 'container']);


      return (0, _jsx3.default)(_reactOverlays.Portal, {
        container: container
      }, void 0, _react2.default.createElement(
        _reactAddonsCssTransitionGroup2.default,
        (0, _extends3.default)({}, (0, _omit3.default)(omitList, otherProps), {
          component: 'div',
          className: (0, _classnames2.default)(_ToastContainer2.default.base, _ToastContainer2.default[position], newestOnTop && _ToastContainer2.default.newestOnTop, className),
          transitionName: {
            enter: _ToastContainer2.default.toastAnimationWrapEnter,
            enterActive: _ToastContainer2.default.toastAnimationWrapEnterActive,
            leave: _ToastContainer2.default.toastAnimationWrapLeave,
            leaveActive: _ToastContainer2.default.toastAnimationWrapLeaveActive
          },
          transitionEnterTimeout: 500,
          transitionLeaveTimeout: 500
        }),
        this.state.toasts.map(function (toastElement, key) {
          return (0, _jsx3.default)('div', {
            className: _ToastContainer2.default.toastAnimationWrap
          }, key, toastElement);
        }).toArray()
      ));
    }
  }]);
  return ToastContainer;
}(_react2.default.Component), _class2.propTypes = {
  position: _react.PropTypes.oneOf(['top', 'bottom']),
  timeout: _react.PropTypes.number,
  newestOnTop: _react.PropTypes.bool,
  className: _react.PropTypes.string,
  container: _reactOverlays.Portal.propTypes.container
}, _class2.defaultProps = {
  position: 'top', // Where toasts will display
  timeout: 4000, // How long toasts will display
  newestOnTop: true }, _temp2)) || _class;

exports.default = ToastContainer;
module.exports = exports['default'];
