// mtk-ui/lib/Toast.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _Toast = require('./styles/Toast.css');

var _Toast2 = _interopRequireDefault(_Toast);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Toast = (0, _pureRenderDecorator2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Toast, _React$Component);

  function Toast() {
    (0, _classCallCheck3.default)(this, Toast);
    return (0, _possibleConstructorReturn3.default)(this, (Toast.__proto__ || Object.getPrototypeOf(Toast)).apply(this, arguments));
  }

  (0, _createClass3.default)(Toast, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          kind = _props.kind,
          size = _props.size,
          className = _props.className,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'kind', 'size', 'className']);


      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({}, otherProps, {
          className: (0, _classnames2.default)(_Toast2.default[kind], _Toast2.default[size], className)
        }),
        children
      );
    }
  }]);
  return Toast;
}(_react2.default.Component), _class2.propTypes = {
  kind: _react.PropTypes.oneOf(['info', 'success', 'error']),
  size: _react.PropTypes.oneOf(['large', 'small']),
  className: _react.PropTypes.string,
  children: _react.PropTypes.node
}, _class2.defaultProps = {
  kind: 'success',
  size: 'large'
}, _temp)) || _class;

exports.default = Toast;
module.exports = exports['default'];
