// mtk-ui/lib/Tooltip.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _Popover = require('./Popover');

var _Popover2 = _interopRequireDefault(_Popover);

var _Tooltip = require('./styles/Tooltip.css');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Tooltip = (0, _pure2.default)(_class = (_temp = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Tooltip, _React$Component);

  function Tooltip() {
    (0, _classCallCheck3.default)(this, Tooltip);
    return (0, _possibleConstructorReturn3.default)(this, (Tooltip.__proto__ || Object.getPrototypeOf(Tooltip)).apply(this, arguments));
  }

  (0, _createClass3.default)(Tooltip, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          otherProps = (0, _objectWithoutProperties3.default)(_props, ['children', 'className']);


      return _react2.default.createElement(
        _Popover2.default,
        (0, _extends3.default)({}, otherProps, {
          className: (0, _classnames2.default)(_Tooltip2.default.base, className),
          cardClassName: _Tooltip2.default.card,
          arrowClassName: _Tooltip2.default.arrow
        }),
        children
      );
    }
  }]);
  return Tooltip;
}(_react2.default.Component), _class2.propTypes = {
  children: _react.PropTypes.node.isRequired,
  className: _react.PropTypes.string
}, _temp)) || _class;

exports.default = Tooltip;
module.exports = exports['default'];
