// @mtk/mcs-components/lib/datachannels/float/Control.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _mtkUiLibInputText = require('mtk-ui/lib/InputText');

var _mtkUiLibInputText2 = _interopRequireDefault(_mtkUiLibInputText);

var _fbjsLibEmptyFunction = require('fbjs/lib/emptyFunction');

var _fbjsLibEmptyFunction2 = _interopRequireDefault(_fbjsLibEmptyFunction);

var _mtkUiLibButton = require('mtk-ui/lib/Button');

var _mtkUiLibButton2 = _interopRequireDefault(_mtkUiLibButton);

var _mtkIcon = require('mtk-icon');

var _commonCardStylesBaseCss = require('../common/card/styles/base.css');

var _commonCardStylesBaseCss2 = _interopRequireDefault(_commonCardStylesBaseCss);

var _stylesMainCss = require('./styles/main.css');

var _stylesMainCss2 = _interopRequireDefault(_stylesMainCss);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var control = function control(_ref) {
  var value = _ref.value;
  var onDelete = _ref.onDelete;
  var onClick = _ref.onClick;
  var displayType = _ref.displayType;
  var onChange = _ref.onChange;
  return _react2['default'].createElement(
    'div',
    { className: (0, _classnames2['default'])(_commonCardStylesBaseCss2['default'].base, _commonCardStylesBaseCss2['default'][displayType], _stylesMainCss2['default'].base) },
    _react2['default'].createElement(
      'div',
      { style: { height: 74, marginTop: 16 } },
      _react2['default'].createElement(_mtkUiLibInputText2['default'], { value: value, className: _stylesMainCss2['default'].input, placeholder: 'Float only', onChange: onChange }),
      _react2['default'].createElement(
        'div',
        { className: _stylesMainCss2['default'].nav },
        _react2['default'].createElement(
          'a',
          { onDelete: onDelete, className: _stylesMainCss2['default'].clear },
          _react2['default'].createElement(_mtkIcon.MiDelete, { className: _stylesMainCss2['default'].MiDelete }),
          _react2['default'].createElement(
            'span',
            null,
            'Clear'
          )
        ),
        _react2['default'].createElement(
          _mtkUiLibButton2['default'],
          { onClick: onClick },
          'OK'
        )
      )
    )
  );
};

var propTypes = {
  status: _react.PropTypes.object,
  displayType: _react.PropTypes.oneOf(['none', 'preview', 'embeded']),
  dataChnId: _react.PropTypes.string,
  unit: _react.PropTypes.string,
  values: _react.PropTypes.object,
  onClick: _react.PropTypes.func,
  onDelete: _react.PropTypes.func,
  onChange: _react.PropTypes.func
};

exports.propTypes = propTypes;
exports['default'] = (0, _recompose.compose)((0, _recompose.defaultProps)({
  status: { value: '0.0' },
  onClick: _fbjsLibEmptyFunction2['default'],
  onDelete: _fbjsLibEmptyFunction2['default'],
  displayType: 'none',
  onChange: _fbjsLibEmptyFunction2['default']
}), (0, _recompose.setPropTypes)(propTypes), _recompose.pure)(control);
