// @mtk/mcs-components/src/utils/sectionHelper.js
import ScrollTo from './scrollTo.js';
import _ from 'lodash';

/**
* getDynamicSection
* 動態產生文章的段落 Section list
*
* @input  {string} scopeClassName 文章的範圍
*
* @output {array} sectionList {innerText, element}
*
* innerText: 段落標題，取用 <h1> 標籤。
* element: (HTMLCollections array) 之後取用 offsetTop 屬性用。
*
* @author Michael Hsu
*/

export const getDynamicSection = (scopeClassName) => {
  let scopeElement = scopeClassName ? document.getElementsByClassName(scopeClassName)[0] : document.getElementsByTagName('body')[0];
  scopeElement = scopeElement || document.getElementsByTagName('body')[0];
  const sectionHTMLCollections = scopeElement.getElementsByTagName('h1');

  const sectionList = _.map(sectionHTMLCollections, (elem) => {
    const element = elem.children[0] || elem; /* Pick the first child as content when there is multiple children in h1 tag. */

    // firefox 使用 innerHTML 屬性，其他瀏覽器使用 innerText
    return {
      innerText: element.innerText || element.innerHTML,
      element
    };
  });

  return sectionList;
};

/**
* Scroll 滾動特效
* offset 60: header height
*
* @input {integer} index，判斷第一個要滾到頂。
* @input {HTMLCollection} element，取目前章節的 offetTop。
* @input {integer} offset，偏差值。
*
* @author Michael Hsu
*/

export const scrollTo = ({ index, element, offset }) => {
  if (index === 0) {
    // first one scroll to top
    ScrollTo(0);
  }
  else {
    ScrollTo(element.offsetTop - Number(offset));
  }
};


/**
* 取得目前位置的 Offset
*
* @input {integer} offset 60 為 Header 高度
*
* @output {integer} Y + offset 60
*
* @author Michael Hsu
*/

export const getCurrentOffset = (offset) => {
  return window.scrollY + Number(offset);
};

/**
* 動態產生 Section list 文章段落間的差距
*
* @input {string} scopeClassName 文章的範圍
*
* @return Promise
*
* @output {array} [77, 269, 3706, 3777]
*
* @author Michael Hsu
*/

export const getSectionIntervalsAsync = (scopeClassName) => {
  return new Promise((resolve, reject) => {
    let scopeElement = scopeClassName ? document.getElementsByClassName(scopeClassName)[0] : document.getElementsByTagName('body')[0];
    scopeElement = scopeElement || document.getElementsByTagName('body')[0];
    const sectionHTMLCollections = scopeElement.getElementsByTagName('h1');

    // imgList: 內容總共有多少張圖片
    let imgList = document.getElementsByTagName('img');

    // imgLoadedCounter: 目前已載入幾張圖片
    let imgLoadedCounter = 0;

    let sectionIntervals = [];

    for (let i = 0; i < imgList.length; i++) {
      checkAllImagesLoaded(imgList[i]);
    }

    function checkAllImagesLoaded(img) {
      img.addEventListener('load', () => {
        imgLoadedCounter += 1;

        // 圖片已經全部載入完成
        if (imgLoadedCounter === imgList.length) {
          // sectionIntervals: 紀錄 section 之間的距離
          const sectionIntervals = _.map(sectionHTMLCollections, (element) => {
            return element.offsetTop;
          });

          resolve(sectionIntervals);
        }
      });
    }
  });

};

/**
* 回傳目前應該 Active 第幾個 li 的 index
*
* sensitive: 觸發敏感度微調。
*
* @input {array} sectionIntervals
*
* @output {integer} index
*
* @author Michael Hsu
*/

export const getLiActiveIndex = ({ sectionIntervals, sensitive, offset }) => {
  let currentOffset = getCurrentOffset(offset);

  let liActiveIndex = 0;
  for (let i = 0; i < sectionIntervals.length; i++) {
    if (currentOffset + Number(sensitive) > sectionIntervals[i + 1]) {
      liActiveIndex = liActiveIndex + 1;
    }
    else {
      break;
    }
  }

  return liActiveIndex;
};
