// @mtk/mcs-components/src/NavItem.jsx
import React from 'react';
import Radium from 'radium';
import {PropTypes} from 'react';
import Clickable from './internal/Clickable';

class NavItem extends Clickable {
  constructor(props) {
    super(props);
  }

  render() {
    let {
      href,
      children,
      style,
      activeStyle,
      linkStyle,
      activeLinkStyle,
      target,
      ...props
    } = this.props;

    return (
      <li {...props}
        style={[
          styles.item,
          style,
          this.state.isHover && activeStyle
      ]}>
        <a target={this.props.target} href={this.props.href} style={[
          styles.link,
          linkStyle,
          this.state.isHover && activeLinkStyle
        ]}>
          {children}
        </a>
      </li>
    );
  }
}

NavItem.propTypes = {
  href: PropTypes.string.isRequired,
  style: PropTypes.object,
  linkStyle: PropTypes.object
};
NavItem.defaultProps = {
  href: '#'
};

export default Radium(NavItem);

const styles = {
  item: {
    display: 'inline-block',
    listStyleType: 'none'
  },
  link: {
    textDecoration: 'none',
    outline: 0,
    whiteSpace: 'nowrap'
  }
};
