// @mtk/mcs-components/src/SideBarMenu.jsx
import React from 'react';
import Radium from 'radium';
import * as SectionHelper from './utils/sectionHelper.js';

class SideBarMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = this.state || {};
    this.state.sectionList = [];
    this.state.liActiveIndex = 0;
  }

  componentDidMount() {
    const scopeClassName = this.props.scopeClassName;
    const sensitive = this.props.sensitive;
    const offset = this.props.offset;

    // Menu li 的 Active 狀態
    SectionHelper.getSectionIntervalsAsync(scopeClassName)
    .then((sectionIntervals) => {
      this._setLiActiveIndex({ sectionIntervals, sensitive, offset });
      window.addEventListener('scroll', this._setLiActiveIndex.bind(this, { sectionIntervals, sensitive, offset }));
    })

    .catch((error) => {
      console.log(error);
    });

    this.setState({
      sectionList: SectionHelper.getDynamicSection(scopeClassName)
    });
  }

  render() {
    let {
      offset,
      ...otherProps
    } = this.props;

    return (
      <div {...otherProps} style={styles.container}>
        <div style={styles.sectionGroup}>
          <div style={styles.overview}>Overview</div>
          <ul style={styles.ul}>
            {
              this.state.sectionList.map((entry, index) => {
                return (
                  <li key={index}
                    style={[
                      styles.li,
                      this.state.liActiveIndex === index ? styles.liActive : {}
                    ]}
                    onClick={e => SectionHelper.scrollTo({ index, element: entry.element, offset })}>
                    {entry.innerText}
                  </li>
                );
              })
            }
          </ul>
        </div>
      </div>
    );
  }

  /**
  * setState 目前 Active li 的 index
  *
  * @author Michael Hsu
  */

  _setLiActiveIndex({ sectionIntervals, sensitive, offset }) {
    this.setState({
      liActiveIndex: SectionHelper.getLiActiveIndex({ sectionIntervals, sensitive, offset })
    });
  }

}

export default Radium(SideBarMenu);

SideBarMenu.propTypes = {
  scopeClassName: React.PropTypes.string,
  sensitive: React.PropTypes.number,
  offset: React.PropTypes.number
};
SideBarMenu.defaultProps = {
  scopeClassName: 'resource',
  sensitive: 5, // 觸發敏感度微調。
  offset: 0 // 保留 Header 高度
};

const styles = {
  container: {
    position: 'fixed',
    height: '100%',
    backgroundColor: '#FAFAFA',
    marginTop: '50px',
    boxShadow: '-1px 0px 0px 0px #F1F2F7',
    width: '200px'
  },
  sectionGroup: {
    marginTop: 79
  },
  overview: {
    fontSize: 18,
    color: '#353630',
    lineHeight: '23px',
    marginLeft: 20,
    marginBottom: 10
  },
  ul:{
    listStyleType: 'none',
    paddingLeft: 0,
    fontSize: 13,
    color: '#353630',
    lineHeight: '16px',
    marginTop: 0
  },
  li: {
    paddingTop: 8,
    paddingBottom: 8,
    paddingLeft: 20,
    cursor: 'pointer',
    ':hover':{
      color: '#00A1DE'
    }
  },
  liActive: {
    background: '#FFFFFF',
    outline: '1px solid #F1F2F7',
    color: '#00A1DE',
    borderLeft: '3px solid #00A1DE',
    paddingLeft: 17
  }
};
