// @mtk/mcs-components/src/Toast.jsx
import React from 'react';
import Radium from 'radium';
import Button from './Button';

class Toast extends React.Component {
  render() {
    const {
      isShow,
      handleYes,
      handleNo,
      primary,
      cancel,
      children,
      ...otherProps
    } = this.props;

    return (
      <div {...otherProps}
        style={[
          styles.wrap,
          !isShow && { marginTop: -height }
        ]}
      >
        <div style={styles.content}>
          <div style={styles.buttonGroup}>
            <Button
              onMouseUp={handleNo}
              kind="cancel">
              {cancel}
            </Button>
            <Button
              onMouseUp={handleYes}
              style={{ marginLeft: 13 }}>
              {primary}
            </Button>
          </div>

          <div style={styles.message}>
            {children}
          </div>
        </div>
      </div>
    );
  }

}

Toast.defaultProps = {
  isShow: true
};

Toast.propTypes = {
  isShow: React.PropTypes.bool,
  handleYes: React.PropTypes.func,
  handleNo: React.PropTypes.func,
  primary: React.PropTypes.string,
  cancel: React.PropTypes.string
};

export default Radium(Toast);

const height = 62;

const styles = {
  wrap: {
    transition: 'margin-top 0.4s ease',
    position: 'fixed',
    zIndex: 998,
    height: height,
    width: '100%',
    color: '#FFF'
  },
  content: {
    boxSizing: 'border-box',
    width: 618,
    padding: '15px 28px 15px 28px',
    fontSize: '13px',
    borderRadius: '0 0 3px 3px',
    backgroundColor: 'rgba(0, 0, 0, 0.8)',
    margin: '0 auto',
    fontFamily: 'RionaSans-Regular'
  },
  buttonGroup: {
    float: 'right'
  },
  message: {
    marginRight: 218,
    lineHeight: '32px'
  }
};
