// @mtk/mcs-components/src/datachannels/string/Display.js
import { default as React, PropTypes } from 'react';
import { compose, pure, setPropTypes, defaultProps, withAttachedProps } from 'recompose';
import { default as InputTextarea } from 'mtk-ui/lib/InputTextarea';
import { default as elementStyle } from '../common/card/styles/base.css';
import { default as styles } from './styles/main.css';
import { default as c } from 'classnames';

const display = ({ status, displayType, placeholder }) => {
  return (
    <div className={c(elementStyle.base, elementStyle[displayType], styles.base )}>
      <InputTextarea
        rows="4"
        placeholder={placeholder}
        value={status.value}
        disabled
        className={styles.displayTextarea}
      />
    </div>
  );
};

export const propTypes = {
  status: PropTypes.object,
  channelType: PropTypes.oneOf(['String', 'Hex']),
  unit: PropTypes.string,
  displayType: PropTypes.oneOf(['none', 'preview', 'embeded']),
};

export default compose(
  defaultProps({
    status: { value: '' },
    channelType: 'String',
    displayType: 'none',
  }),
  setPropTypes(propTypes),
  pure,
  withAttachedProps((getProps) => ({
    placeholder: getProps().channelType === 'Hex' ? 'This place holds Hex value.' : 'This place holds String value.',
  })),
)(display);
