// @mtk/mcs-components/lib/datachannels/pwm/Control.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _mtkUiLibInputText = require('mtk-ui/lib/InputText');

var _mtkUiLibInputText2 = _interopRequireDefault(_mtkUiLibInputText);

var _mtkUiLibButton = require('mtk-ui/lib/Button');

var _mtkUiLibButton2 = _interopRequireDefault(_mtkUiLibButton);

var _mtkIcon = require('mtk-icon');

var _commonCardStylesBaseCss = require('../common/card/styles/base.css');

var _commonCardStylesBaseCss2 = _interopRequireDefault(_commonCardStylesBaseCss);

var _stylesMainCss = require('./styles/main.css');

var _stylesMainCss2 = _interopRequireDefault(_stylesMainCss);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _fbjsLibEmptyFunction = require('fbjs/lib/emptyFunction');

var _fbjsLibEmptyFunction2 = _interopRequireDefault(_fbjsLibEmptyFunction);

var control = function control(_ref) {
  var onClick = _ref.onClick;
  var onDelete = _ref.onDelete;
  var displayType = _ref.displayType;
  return _react2['default'].createElement(
    'div',
    { className: (0, _classnames2['default'])(_commonCardStylesBaseCss2['default'].base, _commonCardStylesBaseCss2['default'][displayType], _stylesMainCss2['default'].base, _stylesMainCss2['default'].controlBase) },
    _react2['default'].createElement(
      'ul',
      { className: _stylesMainCss2['default'].controlList },
      _react2['default'].createElement(
        'li',
        null,
        _react2['default'].createElement(
          'div',
          { className: _stylesMainCss2['default'].title },
          'Period'
        ),
        _react2['default'].createElement(_mtkUiLibInputText2['default'], { placeholder: 'Interger only' })
      ),
      _react2['default'].createElement(
        'li',
        null,
        _react2['default'].createElement(
          'div',
          { className: _stylesMainCss2['default'].title },
          'Value'
        ),
        _react2['default'].createElement(_mtkUiLibInputText2['default'], { placeholder: 'Interger only' })
      )
    ),
    _react2['default'].createElement(
      'div',
      { className: _stylesMainCss2['default'].nav },
      _react2['default'].createElement(
        'a',
        { onDelete: onDelete, className: _stylesMainCss2['default'].clear },
        _react2['default'].createElement(_mtkIcon.MiDelete, { className: _stylesMainCss2['default'].MiDelete }),
        _react2['default'].createElement(
          'span',
          null,
          'Clear'
        )
      ),
      _react2['default'].createElement(
        _mtkUiLibButton2['default'],
        { onClick: onClick },
        'OK'
      )
    )
  );
};

var propTypes = {
  status: _react.PropTypes.object,
  onClick: _react.PropTypes.func,
  onDelete: _react.PropTypes.func,
  displayType: _react.PropTypes.oneOf(['none', 'preview', 'embeded'])
};

exports.propTypes = propTypes;
exports['default'] = (0, _recompose.compose)((0, _recompose.defaultProps)({
  status: { value: 0, period: 100 },
  displayType: 'none',
  onDelete: _fbjsLibEmptyFunction2['default'],
  onClick: _fbjsLibEmptyFunction2['default']
}), (0, _recompose.setPropTypes)(propTypes), _recompose.pure)(control);
