// recompose/index.js
// Higher-order component helpers
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _mapProps2 = require('./mapProps');

var _mapProps3 = _interopRequireDefault(_mapProps2);

exports.mapProps = _mapProps3['default'];

var _mapPropsOnChange2 = require('./mapPropsOnChange');

var _mapPropsOnChange3 = _interopRequireDefault(_mapPropsOnChange2);

exports.mapPropsOnChange = _mapPropsOnChange3['default'];

var _withProps2 = require('./withProps');

var _withProps3 = _interopRequireDefault(_withProps2);

exports.withProps = _withProps3['default'];

var _withAttachedProps2 = require('./withAttachedProps');

var _withAttachedProps3 = _interopRequireDefault(_withAttachedProps2);

exports.withAttachedProps = _withAttachedProps3['default'];

var _defaultProps2 = require('./defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

exports.defaultProps = _defaultProps3['default'];

var _renameProp2 = require('./renameProp');

var _renameProp3 = _interopRequireDefault(_renameProp2);

exports.renameProp = _renameProp3['default'];

var _renameProps2 = require('./renameProps');

var _renameProps3 = _interopRequireDefault(_renameProps2);

exports.renameProps = _renameProps3['default'];

var _flattenProp2 = require('./flattenProp');

var _flattenProp3 = _interopRequireDefault(_flattenProp2);

exports.flattenProp = _flattenProp3['default'];

var _withState2 = require('./withState');

var _withState3 = _interopRequireDefault(_withState2);

exports.withState = _withState3['default'];

var _withReducer2 = require('./withReducer');

var _withReducer3 = _interopRequireDefault(_withReducer2);

exports.withReducer = _withReducer3['default'];

var _branch2 = require('./branch');

var _branch3 = _interopRequireDefault(_branch2);

exports.branch = _branch3['default'];

var _renderComponent2 = require('./renderComponent');

var _renderComponent3 = _interopRequireDefault(_renderComponent2);

exports.renderComponent = _renderComponent3['default'];

var _renderNothing2 = require('./renderNothing');

var _renderNothing3 = _interopRequireDefault(_renderNothing2);

exports.renderNothing = _renderNothing3['default'];

var _shouldUpdate2 = require('./shouldUpdate');

var _shouldUpdate3 = _interopRequireDefault(_shouldUpdate2);

exports.shouldUpdate = _shouldUpdate3['default'];

var _pure2 = require('./pure');

var _pure3 = _interopRequireDefault(_pure2);

exports.pure = _pure3['default'];

var _onlyUpdateForKeys2 = require('./onlyUpdateForKeys');

var _onlyUpdateForKeys3 = _interopRequireDefault(_onlyUpdateForKeys2);

exports.onlyUpdateForKeys = _onlyUpdateForKeys3['default'];

var _onlyUpdateForPropTypes2 = require('./onlyUpdateForPropTypes');

var _onlyUpdateForPropTypes3 = _interopRequireDefault(_onlyUpdateForPropTypes2);

exports.onlyUpdateForPropTypes = _onlyUpdateForPropTypes3['default'];

var _withContext2 = require('./withContext');

var _withContext3 = _interopRequireDefault(_withContext2);

exports.withContext = _withContext3['default'];

var _getContext2 = require('./getContext');

var _getContext3 = _interopRequireDefault(_getContext2);

exports.getContext = _getContext3['default'];

var _doOnReceiveProps2 = require('./doOnReceiveProps');

var _doOnReceiveProps3 = _interopRequireDefault(_doOnReceiveProps2);

exports.doOnReceiveProps = _doOnReceiveProps3['default'];

var _lifecycle2 = require('./lifecycle');

var _lifecycle3 = _interopRequireDefault(_lifecycle2);

exports.lifecycle = _lifecycle3['default'];

var _toClass2 = require('./toClass');

var _toClass3 = _interopRequireDefault(_toClass2);

exports.toClass = _toClass3['default'];

// Static property helpers

var _setStatic2 = require('./setStatic');

var _setStatic3 = _interopRequireDefault(_setStatic2);

exports.setStatic = _setStatic3['default'];

var _setPropTypes2 = require('./setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

exports.setPropTypes = _setPropTypes3['default'];

var _setDisplayName2 = require('./setDisplayName');

var _setDisplayName3 = _interopRequireDefault(_setDisplayName2);

exports.setDisplayName = _setDisplayName3['default'];

// Composition function

var _compose2 = require('./compose');

var _compose3 = _interopRequireDefault(_compose2);

exports.compose = _compose3['default'];

// Other utils

var _getDisplayName2 = require('./getDisplayName');

var _getDisplayName3 = _interopRequireDefault(_getDisplayName2);

exports.getDisplayName = _getDisplayName3['default'];

var _wrapDisplayName2 = require('./wrapDisplayName');

var _wrapDisplayName3 = _interopRequireDefault(_wrapDisplayName2);

exports.wrapDisplayName = _wrapDisplayName3['default'];

var _shallowEqual2 = require('./shallowEqual');

var _shallowEqual3 = _interopRequireDefault(_shallowEqual2);

exports.shallowEqual = _shallowEqual3['default'];

var _isClassComponent2 = require('./isClassComponent');

var _isClassComponent3 = _interopRequireDefault(_isClassComponent2);

exports.isClassComponent = _isClassComponent3['default'];

var _createSink2 = require('./createSink');

var _createSink3 = _interopRequireDefault(_createSink2);

exports.createSink = _createSink3['default'];

var _componentFromProp2 = require('./componentFromProp');

var _componentFromProp3 = _interopRequireDefault(_componentFromProp2);

exports.componentFromProp = _componentFromProp3['default'];

var _nest2 = require('./nest');

var _nest3 = _interopRequireDefault(_nest2);

exports.nest = _nest3['default'];

var _hoistStatics2 = require('./hoistStatics');

var _hoistStatics3 = _interopRequireDefault(_hoistStatics2);

exports.hoistStatics = _hoistStatics3['default'];
