// @mtk/mcs-components/src/datachannels/float/Control.js
import { default as React, PropTypes } from 'react';
import { compose, pure, setPropTypes, defaultProps } from 'recompose';
import { default as InputText } from 'mtk-ui/lib/InputText';
import { default as emptyFunction } from 'fbjs/lib/emptyFunction';
import { default as Button } from 'mtk-ui/lib/Button';
import { MiDelete } from 'mtk-icon';
import { default as elementStyle } from '../common/card/styles/base.css';
import { default as styles } from './styles/main.css';
import { default as c } from 'classnames';

const control = ({ value, onDelete, onClick, displayType, onChange }) => (
  <div className={c(elementStyle.base, elementStyle[displayType], styles.base )}>
    <div style={{ height: 74, marginTop: 16 }}>
      <InputText value={value} className={styles.input} placeholder="Float only" onChange={onChange} />
      <div className={styles.nav}>
        <a onDelete={onDelete} className={styles.clear}>
          <MiDelete className={styles.MiDelete} />
          <span>Clear</span>
        </a>
        <Button onClick={onClick}>OK</Button>
      </div>
    </div>
  </div>
);

export const propTypes = {
  status: PropTypes.object,
  displayType: PropTypes.oneOf(['none', 'preview', 'embeded']),
  dataChnId: PropTypes.string,
  unit: PropTypes.string,
  values: PropTypes.object,
  onClick: PropTypes.func,
  onDelete: PropTypes.func,
  onChange: PropTypes.func,
};

export default compose(
  defaultProps({
    status: { value: '0.0' },
    onClick: emptyFunction,
    onDelete: emptyFunction,
    displayType: 'none',
    onChange: emptyFunction,
  }),
  setPropTypes(propTypes),
  pure,
)(control);
