// @mtk/mcs-components/src/datachannels/float/DisplayCard.js
import { default as React, PropTypes } from 'react';
import { compose, pure, setPropTypes, defaultProps, renameProp } from 'recompose';
import { default as Card } from '../common/card/';
import { default as DatachannelInfo } from '../common/datachannelInfo/';
import { default as Display, propTypes as DisplayPropTypes } from './Display';

const displayCard = ({
  numberOfCards,
  name,
  description,
  dataChnId,
  configs,
  recordedAt,
  fullScreenConfig,
  detailButtonConfig,
  ...props,
}) => (
  <Card
    fullScreenConfig={fullScreenConfig}
    detailButtonConfig={detailButtonConfig}
    configs={configs}
    numberOfCards={numberOfCards} >
    <div>
      <Display {...props} />
      <DatachannelInfo
        name={name}
        recordedAt={recordedAt}
        dataChnId={dataChnId}
        description={description}
      />
    </div>
  </Card>
);

export default compose(
  defaultProps({
    displayType: 'none',
    numberOfCards: 2,
  }),
  setPropTypes({
    configs: PropTypes.object.isRequired,
    numberOfCards: PropTypes.number,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    dataChnId: PropTypes.string.isRequired,
    fullScreenConfig: PropTypes.object,
    detailButtonConfig: PropTypes.object,
    recordedAt: PropTypes.number,
    values: PropTypes.object,
    ...DisplayPropTypes,
  }),
  pure,
  renameProp('values', 'status'),
)(displayCard);
