// @mtk/mcs-components/lib/datachannels/common/detailButton/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _mtkUiLibMenu = require('mtk-ui/lib/Menu');

var _mtkUiLibMenu2 = _interopRequireDefault(_mtkUiLibMenu);

var _mtkUiLibMenuItem = require('mtk-ui/lib/MenuItem');

var _mtkUiLibMenuItem2 = _interopRequireDefault(_mtkUiLibMenuItem);

var _mtkIcon = require('mtk-icon');

var _stylesMainCss = require('./styles/main.css');

var _stylesMainCss2 = _interopRequireDefault(_stylesMainCss);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _iconStylesMainCss = require('../icon/styles/main.css');

var _iconStylesMainCss2 = _interopRequireDefault(_iconStylesMainCss);

var detailButton = function detailButton(_ref) {
  var onClick = _ref.onClick;
  var active = _ref.active;
  var config = _ref.config;
  var defaultSelectedValue = _ref.defaultSelectedValue;

  var props = _objectWithoutProperties(_ref, ['onClick', 'active', 'config', 'defaultSelectedValue']);

  return _react2['default'].createElement(
    'a',
    _extends({ onClick: onClick }, props),
    _react2['default'].createElement(_mtkIcon.MiMoreVert, { className: _iconStylesMainCss2['default'].icon }),
    active ? _react2['default'].createElement(
      _mtkUiLibMenu2['default'],
      {
        className: _stylesMainCss2['default'].menu,
        selectedValue: defaultSelectedValue
      },
      config.map(function (key) {
        return _react2['default'].createElement(_mtkUiLibMenuItem2['default'], { value: key.value, children: key.children });
      })
    ) : ''
  );
};

exports['default'] = (0, _recompose.compose)((0, _recompose.setPropTypes)({
  config: _react.PropTypes.object.isRequired
}), _recompose.pure, (0, _recompose.withState)('active', 'onClick', false), (0, _recompose.withAttachedProps)(function (getProps) {
  return {
    config: getProps().config || [{
      value: 'history',
      children: 'history'
    }],
    className: (0, _classnames2['default'])(_iconStylesMainCss2['default'].base, getProps().className),
    onClick: function onClick() {
      return getProps().onClick(function (value) {
        return !value;
      });
    }
  };
}))(detailButton);
module.exports = exports['default'];
