// @mtk/mcs-components/lib/Breadcrumbs.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _utilsI18n = require('./utils/i18n');

var I18n = _interopRequireWildcard(_utilsI18n);

var _configsResources = require('./configs/resources');

var _configsResources2 = _interopRequireDefault(_configsResources);

var _utilsGetLocaleJs = require('./utils/getLocale.js');

var _utilsGetLocaleJs2 = _interopRequireDefault(_utilsGetLocaleJs);

/**
* 麵包屑導航
* for mcs-resources project
*
* @author Michael Hsu
*/

var Breadcrumbs = (function (_React$Component) {
  _inherits(Breadcrumbs, _React$Component);

  function Breadcrumbs() {
    _classCallCheck(this, Breadcrumbs);

    _get(Object.getPrototypeOf(Breadcrumbs.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(Breadcrumbs, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this = this;

      // 取得現在網址代表的語言
      var i18nPrefix = (0, _utilsGetLocaleJs2['default'])(window.location.href);

      var basePath = '' + (window.location.origin + '/') + (_configsResources2['default'].base + '/') + (i18nPrefix ? i18nPrefix + '/' : '') + (_configsResources2['default'].version + '/');

      // i18n 設定
      I18n.setup({
        lng: i18nPrefix
      }).then(function (i18n) {
        _this.setState({ i18n: i18n.t });
      })['catch'](function (error) {
        console.log(error);
      });

      this.setState({
        basePath: this._adjustPathLocale(basePath),
        resourcesList: _configsResources2['default'].resourcesList.concat(_configsResources2['default'].helpList).concat(_configsResources2['default'].otherList),
        tutorialList: _configsResources2['default'].tutorialList,
        isTutorialBreadcrumbsShow: this._isTutorialBreadcrumbsShow(),
        isMobileBreadcrumbsShow: this._isMobileBreadcrumbsShow(),
        changeLogVersion: this._getChangeLogVersion(),
        i18n: function i18n(text) {
          return text;
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var otherProps = _objectWithoutProperties(this.props, []);

      return _react2['default'].createElement(
        'div',
        _extends({}, otherProps, {
          style: styles.container }),
        _react2['default'].createElement(
          'div',
          { style: styles.breadcrumbs },
          _react2['default'].createElement(
            'a',
            { href: '' + this.state.basePath + this.state.resourcesList[0].path,
              style: styles.text },
            _react2['default'].createElement(
              'span',
              null,
              this.state.i18n('Resources')
            )
          ),
          _react2['default'].createElement(
            'span',
            { style: [styles.hide, this.state.isTutorialBreadcrumbsShow || this.state.isMobileBreadcrumbsShow ? styles.show : {}] },
            _react2['default'].createElement(
              'span',
              { style: [styles.text, styles.textNextIcon] },
              '  ',
              '>',
              '  '
            ),
            _react2['default'].createElement(
              'a',
              { href: '' + this.state.basePath + (this.state.isTutorialBreadcrumbsShow ? 'tutorial' : 'mobile_application'),
                style: styles.text },
              _react2['default'].createElement(
                'span',
                null,
                this.state.isTutorialBreadcrumbsShow ? this.state.i18n('Tutorial') : this.state.i18n('Mobile application')
              )
            )
          ),
          this.state.resourcesList.map(function (entry, index) {
            return _react2['default'].createElement(
              'span',
              { key: index,
                style: [styles.hide, _this2._isItemActive(entry.path) && entry.name !== 'Tutorial' && styles.show] },
              _react2['default'].createElement(
                'span',
                { style: [styles.text, styles.textNextIcon] },
                '  ',
                '>',
                '  '
              ),
              _react2['default'].createElement(
                'a',
                { href: '' + _this2.state.basePath + entry.path,
                  style: styles.text },
                _react2['default'].createElement(
                  'span',
                  null,
                  _this2.state.i18n(entry.name)
                )
              )
            );
          }),
          this.state.tutorialList.map(function (entry, index) {
            return _react2['default'].createElement(
              'span',
              { key: index,
                style: [styles.hide, _this2._isItemActive(entry.path) && styles.show] },
              _react2['default'].createElement(
                'span',
                { style: [styles.text, styles.textNextIcon] },
                '  ',
                '>',
                '  '
              ),
              _react2['default'].createElement(
                'a',
                { href: '' + _this2.state.basePath + entry.path,
                  style: styles.text },
                _react2['default'].createElement(
                  'span',
                  null,
                  _this2.state.i18n(entry.name)
                )
              )
            );
          }),
          _react2['default'].createElement(
            'span',
            { style: [styles.hide, this.state.changeLogVersion ? styles.show : {}] },
            _react2['default'].createElement(
              'span',
              { style: [styles.text, styles.textNextIcon] },
              '  ',
              '>',
              '  '
            ),
            _react2['default'].createElement(
              'a',
              { href: this.state.basePath + 'change_log/' + this.state.changeLogVersion,
                style: styles.text },
              _react2['default'].createElement(
                'span',
                null,
                this.state.changeLogVersion
              )
            )
          )
        )
      );
    }

    /**
    * 判斷該頁 Resources / Help 是否顯示麵包屑
    *
    * @input  {string} regPath (introduction / key_conecpt ...)
    * @output {boolean}
    *
    * @author Michael Hsu
    */

  }, {
    key: '_isItemActive',
    value: function _isItemActive(regPath) {
      return new RegExp(regPath).test(window.location.pathname);
    }

    /**
    * 是否顯示中間的 Tutorial 麵包屑
    * 目前有兩種狀況：tutorial, development_cycle
    *
    * @output {boolean}
    *
    * @author Michael Hsu
    */

  }, {
    key: '_isTutorialBreadcrumbsShow',
    value: function _isTutorialBreadcrumbsShow() {
      return new RegExp('tutorial').test(window.location.pathname) || new RegExp('development_cycle').test(window.location.pathname);
    }

    /**
    * 是否顯示中間的 Mobile 麵包屑
    * 目前有一種狀況：7681_firmware_update
    *
    * @output {boolean}
    *
    * @author Michael Hsu
    */

  }, {
    key: '_isMobileBreadcrumbsShow',
    value: function _isMobileBreadcrumbsShow() {
      return new RegExp('7681_firmware_update').test(window.location.pathname);
    }

    /**
    * 是否顯示中間的 Change log 麵包屑
    * change_log
    *
    * @output {integer} null || version
    *
    * @author Michael Hsu
    */

  }, {
    key: '_getChangeLogVersion',
    value: function _getChangeLogVersion() {
      var changeLogVersion = /change_log\/.*(.html)?/.exec(window.location.pathname);
      changeLogVersion = changeLogVersion && changeLogVersion[0].replace(/(change_log\/)|(.html)/g, '');
      return changeLogVersion;
    }

    /**
    * 處理 basePath 的調整，刪除不需要的 locale
    *
    * @input  (string) basePath
    * @output (string) adjuested basePath
    *
    * Example: 'mcs.mediatek.cn/zh-CN' -> 'mcs.mediatek.cn'
    * Example: 'mcs.mediatek.com/en' -> 'mcs.mediatek.com'
    *
    * @author Michael Hsu
    */

  }, {
    key: '_adjustPathLocale',
    value: function _adjustPathLocale(path) {
      var i18nHost = /\.((cn)|(com)|(io)|(inc))$/.exec(window.location.hostname);
      i18nHost = i18nHost && i18nHost[0];

      switch (i18nHost) {
        case '.cn':
          return path.replace('/zh-CN', '');
        default:
          return path.replace('/en', '');
      }
    }
  }]);

  return Breadcrumbs;
})(_react2['default'].Component);

exports['default'] = (0, _radium2['default'])(Breadcrumbs);

var styles = {
  container: {
    position: 'absolute',
    marginTop: 50,
    width: '100%',
    height: '1em'
  },
  breadcrumbs: {
    background: 'none',
    margin: '0 auto',
    padding: '8px 8px 8px 15px',
    maxWidth: 954
  },
  text: {
    fontSize: 13,
    padding: '0',
    color: '#00a1de',
    textDecoration: 'none',
    outline: 'none'
  },
  textNextIcon: {
    color: 'black'
  },
  hide: {
    display: 'none'
  },
  show: {
    display: 'inline-block'
  }
};
module.exports = exports['default'];
