// @mtk/mcs-components/lib/LayoutHeader.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _reactCookie = require('react-cookie');

var _reactCookie2 = _interopRequireDefault(_reactCookie);

var _utilsUserVoice = require('./utils/UserVoice');

var _utilsUserVoice2 = _interopRequireDefault(_utilsUserVoice);

var _utilsUserReport = require('./utils/UserReport');

var _utilsUserReport2 = _interopRequireDefault(_utilsUserReport);

var _utilsI18n = require('./utils/i18n');

var I18n = _interopRequireWildcard(_utilsI18n);

var _utilsStyleMerger = require('./utils/styleMerger');

var _utilsStyleMerger2 = _interopRequireDefault(_utilsStyleMerger);

var _utilsGetLocaleJs = require('./utils/getLocale.js');

var _utilsGetLocaleJs2 = _interopRequireDefault(_utilsGetLocaleJs);

var _NavItem = require('./NavItem');

var _NavItem2 = _interopRequireDefault(_NavItem);

var _Nav = require('./Nav');

var _Nav2 = _interopRequireDefault(_Nav);

var _DropdownButton = require('./DropdownButton');

var _DropdownButton2 = _interopRequireDefault(_DropdownButton);

var _configsResources = require('./configs/resources');

var _configsResources2 = _interopRequireDefault(_configsResources);

var _configsUserVoiceJson = require('./configs/userVoice.json');

var _configsUserVoiceJson2 = _interopRequireDefault(_configsUserVoiceJson);

var _configsUserReportJson = require('./configs/userReport.json');

var _configsUserReportJson2 = _interopRequireDefault(_configsUserReportJson);

var _imgWeb_logoPng = require('./img/web_logo.png');

var _imgWeb_logoPng2 = _interopRequireDefault(_imgWeb_logoPng);

var LayoutHeader = (function (_React$Component) {
  _inherits(LayoutHeader, _React$Component);

  _createClass(LayoutHeader, null, [{
    key: 'propTypes',
    value: {
      nickname: _react.PropTypes.string,
      imageUrl: _react.PropTypes.string,
      numberOfCards: _react.PropTypes.number,
      logoutFn: _react.PropTypes.func,
      isManager: _react.PropTypes.bool
    },
    enumerable: true
  }, {
    key: 'defaultProps',
    value: {
      isManager: false
    },
    enumerable: true
  }]);

  function LayoutHeader(props) {
    _classCallCheck(this, LayoutHeader);

    _get(Object.getPrototypeOf(LayoutHeader.prototype), 'constructor', this).call(this, props);
    this.state = {
      isChina: false
    };
  }

  _createClass(LayoutHeader, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this = this;

      // 取得現在網址代表的語言
      var i18nPrefix = (0, _utilsGetLocaleJs2['default'])(window.location.href);

      var basePath = '' + (window.location.origin + '/') + (_configsResources2['default'].base + '/') + (i18nPrefix ? i18nPrefix + '/' : '') + (_configsResources2['default'].version + '/');

      var rootPath = '' + (window.location.origin + '/') + (i18nPrefix ? i18nPrefix + '/' : '');

      var isChina = /\.cn$/.test(window.location.hostname);
      if (isChina) {
        // UserReport 設定
        this.userReport = new _utilsUserReport2['default']({
          API_KEY: require('./configs/userReport.json').API_KEY
        }).setup();
        this.setState({ isChina: true });
      } else {
        // UserVoice 設定
        this.userVoice = new _utilsUserVoice2['default']({
          target: '#feedbackUserVoice',
          locale: i18nPrefix,
          API_KEY: _configsUserVoiceJson2['default'].API_KEY,
          API_KEY: require('./configs/userVoice.json').API_KEY
        }).setup();
      }

      // i18n 設定
      I18n.setup({
        lng: i18nPrefix
      }).then(function (i18n) {
        _this.setState({ i18n: i18n.t });
      })['catch'](function (error) {
        console.log(error);
      });

      this.setState({
        resourcesList: _configsResources2['default'].resourcesList,
        helpList: _configsResources2['default'].helpList,
        i18nPrefix: i18nPrefix,
        basePath: this._adjustPathLocale(basePath),
        rootPath: this._adjustPathLocale(rootPath),
        v2consolePath: this._adjustPathLocale('/v2console/' + i18nPrefix),
        isMobileMode: this.props.numberOfCards <= 2,
        i18n: function i18n(text) {
          return text;
        }
      });
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        isMobileMode: nextProps.numberOfCards <= 2
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props;
      var nickname = _props.nickname;
      var imageUrl = _props.imageUrl;
      var numberOfCards = _props.numberOfCards;
      var logoutFn = _props.logoutFn;
      var isManager = _props.isManager;

      var otherProps = _objectWithoutProperties(_props, ['nickname', 'imageUrl', 'numberOfCards', 'logoutFn', 'isManager']);

      return _react2['default'].createElement(
        'header',
        _extends({}, otherProps, { style: styles.header }),
        _react2['default'].createElement('div', { style: styles.mask }),
        _react2['default'].createElement(
          'div',
          { style: [styles.container, { width: 318 * numberOfCards }] },
          _react2['default'].createElement(
            _Nav2['default'],
            null,
            _react2['default'].createElement(
              _NavItem2['default'],
              { href: this.state.v2consolePath,
                linkStyle: styles.logoLink
              },
              _react2['default'].createElement('img', {
                src: _imgWeb_logoPng2['default'],
                alt: 'logo',
                style: styles.logo
              })
            ),
            this.state.isChina && _react2['default'].createElement(
              _NavItem2['default'],
              { href: 'http://ns.csip.cn/',
                target: '_blank',
                linkStyle: styles.logoLink
              },
              _react2['default'].createElement('img', {
                src: require('./img/csip_logo.png'),
                alt: 'csip',
                style: (0, _utilsStyleMerger2['default'])(styles.logo, { width: 105 })
              })
            ),
            _react2['default'].createElement(
              _Nav2['default'],
              null,
              _react2['default'].createElement(
                _DropdownButton2['default'],
                {
                  buttonStyle: (0, _utilsStyleMerger2['default'])(styles.link, this._isMenuActive('developmentList') && styles.activeStyle),
                  activeStyle: styles.activeStyle,
                  title: _react2['default'].createElement(
                    'span',
                    null,
                    _react2['default'].createElement('i', { className: 'fa fa-wrench', style: styles.prefixIcon }),
                    !this.state.isMobileMode && this.state.i18n('Development')
                  )
                },
                _react2['default'].createElement(
                  _NavItem2['default'],
                  {
                    href: this.state.v2consolePath + '/development',
                    style: [styles.menuItem, this._isItemActive('development') ? styles.menuItemActive : {}],
                    linkStyle: styles.menuLink,
                    activeStyle: styles.menuItemActive
                  },
                  this.state.i18n('Prototype')
                ),
                _react2['default'].createElement(
                  _NavItem2['default'],
                  {
                    href: this.state.v2consolePath + '/testdevices',
                    style: [styles.menuItem, this._isItemActive('testdevices') ? styles.menuItemActive : {}, { borderRadius: '0 0 4px 4px' }],
                    linkStyle: styles.menuLink,
                    activeStyle: styles.menuItemActive
                  },
                  this.state.i18n('Test devices')
                )
              )
            ),
            _react2['default'].createElement(
              _NavItem2['default'],
              {
                href: this.state.v2consolePath + '/management',
                style: this._isItemActive('management') ? styles.activeStyle : {},
                linkStyle: (0, _utilsStyleMerger2['default'])(styles.link, this._isItemActive('management') && styles.activeStyle),
                activeStyle: styles.activeStyle,
                activeLinkStyle: styles.activeStyle
              },
              _react2['default'].createElement('i', { className: 'fa fa-clipboard', style: styles.prefixIcon }),
              !this.state.isMobileMode && this.state.i18n('Management')
            ),
            isManager && _react2['default'].createElement(
              _NavItem2['default'],
              {
                href: '/analytics/user' + (this.state.i18nPrefix && '?locale=' + this.state.i18nPrefix),
                style: this._isItemActive('analytics') ? styles.activeStyle : {},
                linkStyle: (0, _utilsStyleMerger2['default'])(styles.link, this._isItemActive('analytics') && styles.activeStyle),
                activeStyle: styles.activeStyle,
                activeLinkStyle: styles.activeStyle
              },
              _react2['default'].createElement('i', { className: 'fa fa-area-chart', style: styles.prefixIcon }),
              !this.state.isMobileMode && this.state.i18n('Analytics')
            )
          ),
          _react2['default'].createElement(
            _Nav2['default'],
            { style: { float: 'right' } },
            _react2['default'].createElement(
              _DropdownButton2['default'],
              {
                buttonStyle: (0, _utilsStyleMerger2['default'])(styles.link, this._isMenuActive('resourcesList') && styles.activeStyle),
                activeStyle: styles.activeStyle,
                title: _react2['default'].createElement(
                  'span',
                  null,
                  _react2['default'].createElement('i', { className: 'fa fa-inbox', style: styles.prefixIcon }),
                  !this.state.isMobileMode && this.state.i18n('Resources')
                )
              },
              this.state.resourcesList.map(function (entry, index, array) {
                return _react2['default'].createElement(
                  _NavItem2['default'],
                  { key: index, href: entry.path.startsWith('http') ? entry.path : _this2.state.basePath + entry.path,
                    style: [styles.menuItem, _this2._isItemActive(entry.path) ? styles.menuItemActive : {}, array.length === index + 1 ? { borderRadius: '0 0 4px 4px' } : {}],
                    linkStyle: styles.menuLink,
                    activeStyle: (0, _utilsStyleMerger2['default'])(styles.menuItemActive, array.length === index + 1 ? { borderRadius: '0 0 4px 4px' } : {})
                  },
                  _this2.state.i18n(entry.name)
                );
              })
            ),
            _react2['default'].createElement(
              _DropdownButton2['default'],
              {
                id: 'feedbackUserVoice',
                buttonStyle: (0, _utilsStyleMerger2['default'])(styles.link, this._isMenuActive('helpList') && styles.activeStyle),
                activeStyle: styles.activeStyle,
                title: _react2['default'].createElement(
                  'span',
                  null,
                  _react2['default'].createElement('i', { className: 'fa fa-comments-o', style: styles.prefixIcon }),
                  !this.state.isMobileMode && this.state.i18n('Help')
                )
              },
              this.state.helpList.map(function (entry, index) {
                return _react2['default'].createElement(
                  _NavItem2['default'],
                  { key: index, href: '' + _this2.state.basePath + entry.path,
                    style: [styles.menuItem, _this2._isItemActive(entry.path) ? styles.menuItemActive : {}],
                    linkStyle: styles.menuLink,
                    activeStyle: styles.menuItemActive
                  },
                  _this2.state.i18n(entry.name)
                );
              }),
              _react2['default'].createElement(
                _NavItem2['default'],
                { key: 'Forum',
                  href: 'http://labs.mediatek.com/forums/forums/show/48.page',
                  target: '_blank',
                  style: styles.menuItem,
                  linkStyle: styles.menuLink,
                  activeStyle: styles.menuItemActive
                },
                this.state.i18n('Forum')
              ),
              _react2['default'].createElement(
                _NavItem2['default'],
                { key: 'Feedback',
                  onClick: this._handleFeedbackClick.bind(this),
                  style: styles.menuItem,
                  linkStyle: styles.menuLink,
                  activeStyle: (0, _utilsStyleMerger2['default'])(styles.menuItemActive, { borderRadius: '0 0 4px 4px' })
                },
                this.state.i18n('Feedback')
              )
            ),
            _react2['default'].createElement(
              _DropdownButton2['default'],
              {
                buttonStyle: styles.link,
                activeStyle: styles.activeStyle,
                title: _react2['default'].createElement(
                  'span',
                  null,
                  _react2['default'].createElement('img', { src: imageUrl, style: styles.userImage })
                )
              },
              _react2['default'].createElement(
                _NavItem2['default'],
                { key: 'Profile',
                  href: this.state.v2consolePath + '/console/profile',
                  style: styles.menuItem,
                  linkStyle: styles.menuLink,
                  activeStyle: styles.menuItemActive },
                this.state.i18n('Profile')
              ),
              _react2['default'].createElement(
                _NavItem2['default'],
                { key: 'TermsOfUse',
                  href: this.state.basePath + 'terms_of_use/',
                  style: styles.menuItem,
                  linkStyle: styles.menuLink,
                  activeStyle: styles.menuItemActive },
                this.state.i18n('Terms of use')
              ),
              _react2['default'].createElement(
                _NavItem2['default'],
                { key: 'SignOut',
                  onClick: function (e) {
                    return logoutFn();
                  },
                  style: styles.menuItem,
                  linkStyle: styles.menuLink,
                  activeStyle: (0, _utilsStyleMerger2['default'])(styles.menuItemActive, { borderRadius: '0 0 4px 4px' })
                },
                this.state.i18n('Sign out')
              )
            )
          )
        )
      );
    }
  }, {
    key: '_handleFeedbackClick',
    value: function _handleFeedbackClick() {
      var isChina = /\.cn$/.test(window.location.hostname);

      isChina ? this.userReport.show() : this.userVoice.show();
    }

    /**
    * Menu (Resources / Help) 是否 active
    *
    * @input  {string} listName (resourcesList / helpList)
    * @output {boolean}
    *
    * @author Michael Hsu
    */

  }, {
    key: '_isMenuActive',
    value: function _isMenuActive(listName) {
      switch (listName) {
        case 'developmentList':
          return (/development/.test(window.location.pathname) || /testdevices/.test(window.location.pathname)
          );
        case 'resourcesList':
          return (/resources/.test(window.location.pathname) && !/faq/.test(window.location.pathname)
          );
        case 'helpList':
          return (/resources/.test(window.location.pathname) && /faq/.test(window.location.pathname)
          );
      }
    }

    /**
    * Item (Resources / Help) 是否 active
    *
    * @input  {string} regPath (introduction / key_conecpt ...)
    * @output {boolean}
    *
    * @author Michael Hsu
    */

  }, {
    key: '_isItemActive',
    value: function _isItemActive(regPath) {
      return new RegExp(regPath).test(window.location.pathname);
    }

    /**
    * 處理 basePath 的調整，刪除不需要的 locale
    *
    * @input  (string) basePath
    * @output (string) adjuested basePath
    *
    * Example: 'mcs.mediatek.cn/zh-CN' -> 'mcs.mediatek.cn'
    * Example: 'mcs.mediatek.com/en' -> 'mcs.mediatek.com'
    *
    * @author Michael Hsu
    */

  }, {
    key: '_adjustPathLocale',
    value: function _adjustPathLocale(path) {
      var i18nHost = /\.((cn)|(com)|(io)|(inc))$/.exec(window.location.hostname);
      i18nHost = i18nHost && i18nHost[0];

      switch (i18nHost) {
        case '.cn':
          return path.replace('/zh-CN', '');
        default:
          return path.replace('/en', '');
      }
    }
  }]);

  return LayoutHeader;
})(_react2['default'].Component);

exports['default'] = (0, _radium2['default'])(LayoutHeader);

LayoutHeader.propTypes = {
  nickname: _react2['default'].PropTypes.string,
  imageUrl: _react2['default'].PropTypes.string,
  numberOfCards: _react2['default'].PropTypes.number
};
LayoutHeader.defaultProps = {
  nickname: 'MCS User',
  numberOfCards: 3 // 預設卡片數量，一般螢幕為三張
};

var styles = {
  header: {
    width: '100%',
    height: 50,
    background: '#00a1de',
    boxShadow: '1px 2px 1px 0 rgba(0,0,0,0.1), 0 0 0 rgba(0,0,0,0.1)',
    position: 'fixed',
    zIndex: 999,
    fontSize: 13,
    userSelect: 'none'
  },
  mask: {
    position: 'absolute',
    width: '50%',
    backgroundColor: '#00a1de',
    height: 49,
    zIndex: 1004,
    right: 0
  },
  container: {
    background: '#00a1de',
    margin: '0 auto',
    padding: '0px 0px 0 3px',
    position: 'relative'
  },
  item: {
    marginLeft: '-0.32em'
  },
  link: {
    lineHeight: '50px',
    color: '#fff',
    position: 'relative',
    padding: '19px 12px 17px 12px',
    backgroundColor: '#00a1de',
    zIndex: 1004
  },
  logoLink: {
    lineHeight: '50px',
    padding: '18px 5px',
    marginRight: 15
  },
  logo: {
    width: 120,
    verticalAlign: 'middle'
  },
  activeStyle: {
    backgroundColor: '#1995cc'
  },
  menuLink: {
    display: 'block',
    clear: 'both',
    fontWeight: '400',
    padding: '10px 20px',
    color: '#fff'
  },
  menuItem: {
    width: '100%',
    height: '40px',
    textAlign: 'left'
  },
  menuItemActive: {
    backgroundColor: '#0c86bc'
  },
  prefixIcon: {
    position: 'relative',
    marginLeft: 5,
    marginRight: 5,
    fontSize: 13
  },
  navbar: {
    position: 'relative',
    marginLeft: 0,
    marginRight: 0,
    fontSize: 25,
    top: 8
  },
  userImage: {
    width: 29,
    height: 29,
    borderRadius: '100%',
    border: '1px solid #d1d3d4',
    padding: 1,
    background: '#fff',
    verticalAlign: 'middle'
  }
};
module.exports = exports['default'];
