// @mtk/mcs-components/lib/Toast.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var Toast = (function (_React$Component) {
  _inherits(Toast, _React$Component);

  function Toast() {
    _classCallCheck(this, Toast);

    _get(Object.getPrototypeOf(Toast.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(Toast, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var isShow = _props.isShow;
      var handleYes = _props.handleYes;
      var handleNo = _props.handleNo;
      var primary = _props.primary;
      var cancel = _props.cancel;
      var children = _props.children;

      var otherProps = _objectWithoutProperties(_props, ['isShow', 'handleYes', 'handleNo', 'primary', 'cancel', 'children']);

      return _react2['default'].createElement(
        'div',
        _extends({}, otherProps, {
          style: [styles.wrap, !isShow && { marginTop: -height }]
        }),
        _react2['default'].createElement(
          'div',
          { style: styles.content },
          _react2['default'].createElement(
            'div',
            { style: styles.buttonGroup },
            _react2['default'].createElement(
              _Button2['default'],
              {
                onMouseUp: handleNo,
                kind: 'cancel' },
              cancel
            ),
            _react2['default'].createElement(
              _Button2['default'],
              {
                onMouseUp: handleYes,
                style: { marginLeft: 13 } },
              primary
            )
          ),
          _react2['default'].createElement(
            'div',
            { style: styles.message },
            children
          )
        )
      );
    }
  }]);

  return Toast;
})(_react2['default'].Component);

Toast.defaultProps = {
  isShow: true
};

Toast.propTypes = {
  isShow: _react2['default'].PropTypes.bool,
  handleYes: _react2['default'].PropTypes.func,
  handleNo: _react2['default'].PropTypes.func,
  primary: _react2['default'].PropTypes.string,
  cancel: _react2['default'].PropTypes.string
};

exports['default'] = (0, _radium2['default'])(Toast);

var height = 62;

var styles = {
  wrap: {
    transition: 'margin-top 0.4s ease',
    position: 'fixed',
    zIndex: 998,
    height: height,
    width: '100%',
    color: '#FFF'
  },
  content: {
    boxSizing: 'border-box',
    width: 618,
    padding: '15px 28px 15px 28px',
    fontSize: '13px',
    borderRadius: '0 0 3px 3px',
    backgroundColor: 'rgba(0, 0, 0, 0.8)',
    margin: '0 auto',
    fontFamily: 'RionaSans-Regular'
  },
  buttonGroup: {
    float: 'right'
  },
  message: {
    marginRight: 218,
    lineHeight: '32px'
  }
};
module.exports = exports['default'];
