// @mtk/mcs-components/src/datachannels/analog/Control.js
import { default as React, PropTypes } from 'react';
import { compose, pure, setPropTypes, defaultProps } from 'recompose';
import { default as InputRange } from 'mtk-ui/lib/InputRange';
import { default as elementStyle } from '../common/card/styles/base.css';
import { default as styles } from './styles/main.css';
import { default as c } from 'classnames';
import { default as emptyFunction } from 'fbjs/lib/emptyFunction';

const control = ({ lowerbound, upperbound, onChange, value, displayType }) => (
  <div className={c(elementStyle.base, elementStyle[displayType], styles.base )}>
    <div className={styles.currentValue} > Current Value: <span style={{ color: '#00A1DE' }}>{value}</span></div>
    <InputRange
      max={lowerbound}
      min={upperbound}
      value={value}
      onChange={onChange}
      style={{ width: 240 }}
    />
    <div className={styles.limit}>
      <span>{lowerbound}</span>
      <span>{upperbound}</span>
    </div>
  </div>
);

export const propTypes = {
  status: PropTypes.object,
  value: PropTypes.number,
  displayType: PropTypes.oneOf(['none', 'preview', 'embeded']),
  onChange: PropTypes.func,
  lowerbound: PropTypes.number.isRequired,
  upperbound: PropTypes.number.isRequired,
};

export default compose(
  defaultProps({
    status: { value: 0 },
    displayType: 'none',
    onChange: emptyFunction,
  }),
  setPropTypes(propTypes),
  pure,
)(control);
