// @mtk/mcs-components/src/datachannels/analog/ControlCard.js
import { default as React, PropTypes } from 'react';
import { compose, pure, setPropTypes, defaultProps, renameProp } from 'recompose';
import { default as Card } from '../common/card/';
import { default as DatachannelInfo } from '../common/datachannelInfo/index.js';
import { default as Control, propTypes as ControlPropTypes } from './Control';

const displayCard = ({
  numberOfCards,
  name,
  description,
  dataChnId,
  configs,
  recordedAt,
  fullScreenConfig,
  detailButtonConfig,
  ...props,
}) => (
  <Card
    fullScreenConfig={fullScreenConfig}
    detailButtonConfig={detailButtonConfig}
    configs={configs}
    numberOfCards={numberOfCards} >
    <Control {...props} />
    <DatachannelInfo
      name={name}
      recordedAt={recordedAt}
      dataChnId={dataChnId}
      description={description}
    />
  </Card>
);

export default compose(
  defaultProps({
    numberOfCards: 2,
    displayType: 'none',
  }),
  setPropTypes({
    configs: PropTypes.object.isRequired,
    numberOfCards: PropTypes.number,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    dataChnId: PropTypes.string.isRequired,
    values: PropTypes.object,
    recordedAt: PropTypes.number,
    fullScreenConfig: PropTypes.object,
    detailButtonConfig: PropTypes.object,
    ...ControlPropTypes,
  }),
  pure,
  renameProp('values', 'status'),
)(displayCard);
