// @mtk/mcs-components/src/datachannels/category/DisplayCard.js
import { default as React, PropTypes } from 'react';
import { renameProp, compose, pure, setPropTypes, defaultProps } from 'recompose';
import { default as Card } from '../common/card/';
import { default as DatachannelInfo } from '../common/datachannelInfo/';
import { default as Display, propTypes as DisplayPropTypes } from './Display';

const displayCard = ({
  numberOfCards,
  name,
  description,
  dataChnId,
  fullScreenConfig,
  detailButtonConfig,
  recordedAt,
  configs,
  ...props,
}) => (
  <Card
    fullScreenConfig={fullScreenConfig}
    detailButtonConfig={detailButtonConfig}
    configs={configs}
    numberOfCards={numberOfCards} >
    <Display {...props} />
    <DatachannelInfo
      name={name}
      recordedAt={recordedAt}
      dataChnId={dataChnId}
      description={description}
    />
  </Card>
);

export default compose(
  defaultProps({
    numberOfCards: 2,
    numberOfCategory: 2,
    displayType: 'none',
  }),
  setPropTypes({
    configs: PropTypes.object.isRequired,
    numberOfCards: PropTypes.number,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    dataChnId: PropTypes.string.isRequired,
    fullScreenConfig: PropTypes.object,
    recordedAt: PropTypes.number,
    values: PropTypes.object,
    detailButtonConfig: PropTypes.object,
    ...DisplayPropTypes,
  }),
  pure,
  renameProp('values', 'status'),
)(displayCard);
