// @mtk/mcs-components/src/utils/UserVoice.js
/**
* 設定 UserVoice
*
* @input  {string} target - DOM ID
* @input  {string} locale - language
* @input  {string} API_KEY
*
* @output {object} UserVoice (global)
*
* @author Michael Hsu
*/

export default class UserVoice {
  constructor({ target, locale, API_KEY }) {
    this.API_KEY = API_KEY;
    this.target = target;
    this.locale = locale;
  }

  renderScript() {
    const uv = document.createElement('script');
    uv.type = 'text/javascript';
    uv.async = true;
    uv.src = `//widget.uservoice.com/${this.API_KEY}.js`;
    const s = document.getElementsByTagName('script')[0];
    s.parentNode.insertBefore(uv, s);
  }

  setup() {
    this.renderScript();

    // Include the UserVoice JavaScript SDK (only needed once on a page)
    global.UserVoice = window.UserVoice || [];

    // locale ISO 不同規則
    // Reference: https://developer.uservoice.com/docs/single-sign-on/single-sign-on/
    if (this.locale === 'zh-CN') this.locale = 'cn';

    global.UserVoice.push(['set', {
      target: this.target,
      accent_color: '#448dd6',
      trigger_color: 'white',
      trigger_background_color: 'rgba(46, 49, 51, 0.6)',
      locale: this.locale,
      right: '155px'
    }]);
    global.UserVoice.push(['identify', {}]);
    global.UserVoice.push(['autoprompt', {}]);

    return this;
  }

  show() {
    global.UserVoice.push(['show', {
      mode: 'contact' // mode: contact, smartvote, satisfaction
    }]);
  }
}
