// @mtk/mcs-components/lib/datachannels/string/Control.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _mtkUiLibInputTextarea = require('mtk-ui/lib/InputTextarea');

var _mtkUiLibInputTextarea2 = _interopRequireDefault(_mtkUiLibInputTextarea);

var _mtkUiLibButton = require('mtk-ui/lib/Button');

var _mtkUiLibButton2 = _interopRequireDefault(_mtkUiLibButton);

var _mtkIcon = require('mtk-icon');

var _commonCardStylesBaseCss = require('../common/card/styles/base.css');

var _commonCardStylesBaseCss2 = _interopRequireDefault(_commonCardStylesBaseCss);

var _stylesMainCss = require('./styles/main.css');

var _stylesMainCss2 = _interopRequireDefault(_stylesMainCss);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _fbjsLibEmptyFunction = require('fbjs/lib/emptyFunction');

var _fbjsLibEmptyFunction2 = _interopRequireDefault(_fbjsLibEmptyFunction);

var control = function control(_ref) {
  var defaultValue = _ref.defaultValue;
  var value = _ref.value;
  var onDelete = _ref.onDelete;
  var onClick = _ref.onClick;
  var onChange = _ref.onChange;
  var displayType = _ref.displayType;
  var placeholder = _ref.placeholder;

  return _react2['default'].createElement(
    'div',
    { className: (0, _classnames2['default'])(_commonCardStylesBaseCss2['default'].base, _commonCardStylesBaseCss2['default'][displayType], _stylesMainCss2['default'].base) },
    _react2['default'].createElement(
      'div',
      { style: { height: 165, marginTop: 23 } },
      _react2['default'].createElement(_mtkUiLibInputTextarea2['default'], {
        rows: '8',
        value: value,
        onChange: onChange,
        defaultValue: defaultValue,
        placeholder: placeholder,
        className: _stylesMainCss2['default'].controlTextarea }),
      _react2['default'].createElement(
        'div',
        { className: _stylesMainCss2['default'].nav },
        _react2['default'].createElement(
          'a',
          { onDelete: onDelete, className: _stylesMainCss2['default'].clear },
          _react2['default'].createElement(_mtkIcon.MiDelete, { className: _stylesMainCss2['default'].MiDelete }),
          _react2['default'].createElement(
            'span',
            null,
            'Clear'
          )
        ),
        _react2['default'].createElement(
          _mtkUiLibButton2['default'],
          { onClick: onClick },
          'OK'
        )
      )
    )
  );
};

var propTypes = {
  displayType: _react.PropTypes.oneOf(['none', 'preview', 'embeded']),
  unit: _react.PropTypes.string,
  value: _react.PropTypes.string,
  channelType: _react.PropTypes.oneOf(['String', 'Hex']),
  defaultValue: _react.PropTypes.string
};

exports.propTypes = propTypes;
exports['default'] = (0, _recompose.compose)((0, _recompose.defaultProps)({
  channelType: 'String',
  displayType: 'none'
}), (0, _recompose.withState)('value', 'setValue', function (props) {
  return props.defaultValue;
}), (0, _recompose.setPropTypes)(propTypes), _recompose.pure, (0, _recompose.withAttachedProps)(function (getProps) {
  return {
    placeholder: getProps().channelType === 'Hex' ? 'Input the hexadecimal value, only 0-9 or A-F.' : 'Input the String.',
    onClick: _fbjsLibEmptyFunction2['default'],
    onChange: function onChange(e) {
      return getProps().setValue(e.target.value);
    }
  };
}), (0, _recompose.mapProps)(function (props) {
  return _extends({}, props, {
    value: props.value
  });
}))(control);
