// @mtk/mcs-components/lib/Button.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var Button = (function (_React$Component) {
  _inherits(Button, _React$Component);

  function Button() {
    _classCallCheck(this, Button);

    _get(Object.getPrototypeOf(Button.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(Button, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var customizedStyle = _props.style;

      var otherProps = _objectWithoutProperties(_props, ['style']);

      return _react2['default'].createElement(
        'button',
        _extends({}, otherProps, {
          style: [styles.base, styles[this.props.kind], customizedStyle] }),
        this.props.children
      );
    }
  }]);

  return Button;
})(_react2['default'].Component);

Button.propTypes = {
  kind: _react2['default'].PropTypes.oneOf(['primary', 'cancel' /*, 'warning', 'danger', 'primary' */]).isRequired
};

Button.defaultProps = {
  kind: 'primary'
};

exports['default'] = (0, _radium2['default'])(Button);

var styles = {
  base: {
    boxSizing: 'border-box',
    display: 'inline-block',
    padding: '6px 12px',
    textAlign: 'center',
    lineHeight: 1,
    whiteSpace: 'nowrap',
    verticalAlign: 'middle',
    touchAction: 'manipulation',
    border: 'none',
    borderRadius: 4,
    color: '#fafafa',
    cursor: 'pointer',
    fontSize: 13,
    minWidth: 80,
    height: 32,
    fontFamily: 'RionaSans-Regular',
    ':focus': {
      outline: 'none'
    },
    transition: 'all 0.4s ease'
  },

  primary: {
    background: '#00a1de',
    ':hover': {
      background: '#0195CC'
    },
    ':active': {
      background: '#0B86BC'
    }
  },

  cancel: {
    background: '#fafafa',
    border: '1px solid #d1d2d3',
    color: '#999a94',
    ':hover': {
      background: '#EAEAEA'
    },
    ':active': {
      background: '#DDDDDD'
    }
  }
};
module.exports = exports['default'];
