// @mtk/mcs-components/lib/ButtonGroup.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _utilsScrollToJs = require('./utils/scrollTo.js');

var _utilsScrollToJs2 = _interopRequireDefault(_utilsScrollToJs);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var ButtonGroup = (function (_React$Component) {
  _inherits(ButtonGroup, _React$Component);

  function ButtonGroup(props) {
    _classCallCheck(this, ButtonGroup);

    _get(Object.getPrototypeOf(ButtonGroup.prototype), 'constructor', this).call(this, props);
    this.state = this.state || {};
    this.state.sectionList = [];
  }

  _createClass(ButtonGroup, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var customizedStyle = _props.style;
      var kind = _props.kind;
      var leftTitle = _props.leftTitle;
      var rightTitle = _props.rightTitle;

      var otherProps = _objectWithoutProperties(_props, ['style', 'kind', 'leftTitle', 'rightTitle']);

      return _react2['default'].createElement(
        'div',
        _extends({}, otherProps, { style: customizedStyle }),
        _react2['default'].createElement(
          _Button2['default'],
          { kind: kind,
            style: styles.buttonLeft },
          leftTitle
        ),
        _react2['default'].createElement(
          _Button2['default'],
          { kind: kind,
            style: styles.buttonRight },
          rightTitle
        )
      );
    }
  }]);

  return ButtonGroup;
})(_react2['default'].Component);

ButtonGroup.propTypes = {
  style: _react2['default'].PropTypes.object,
  kind: _react2['default'].PropTypes.oneOf(['primary', 'cancel' /*, 'warning', 'danger', 'primary' */]).isRequired,
  leftTitle: _react2['default'].PropTypes.oneOfType([_react2['default'].PropTypes.string, _react2['default'].PropTypes.object]),
  rightTitle: _react2['default'].PropTypes.oneOfType([_react2['default'].PropTypes.string, _react2['default'].PropTypes.object])
};

ButtonGroup.defaultProps = {
  kind: 'primary',
  leftTitle: 'Left Title',
  rightTitle: 'Right Title'
};

exports['default'] = (0, _radium2['default'])(ButtonGroup);

var styles = {
  buttonLeft: {
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0
  },
  buttonRight: {
    minWidth: 0,
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
    borderLeft: '1px solid #1fadea'
  }
};
module.exports = exports['default'];
