// @mtk/mcs-components/lib/Footer.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactStyle = require('react-style');

var _reactStyle2 = _interopRequireDefault(_reactStyle);

var _utilsScreen = require('./utils/screen');

var Screen = _interopRequireWildcard(_utilsScreen);

var _utilsI18n = require('./utils/i18n');

var I18n = _interopRequireWildcard(_utilsI18n);

var _utilsGetLocaleJs = require('./utils/getLocale.js');

var _utilsGetLocaleJs2 = _interopRequireDefault(_utilsGetLocaleJs);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _configsSocialMediaJs = require('./configs/socialMedia.js');

var _configsSocialMediaJs2 = _interopRequireDefault(_configsSocialMediaJs);

var Footer = (function (_React$Component) {
  _inherits(Footer, _React$Component);

  function Footer(props) {
    _classCallCheck(this, Footer);

    _get(Object.getPrototypeOf(Footer.prototype), 'constructor', this).call(this, props);
    this.state = this.state || {};
  }

  _createClass(Footer, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this = this;

      // footer 置底監聽
      window.addEventListener('resize', this._handleResize.bind(this));
      window.addEventListener('load', this._handleResize.bind(this));

      // 取得現在網址代表的語言
      var i18nPrefix = (0, _utilsGetLocaleJs2['default'])(window.location.href);

      // i18n 設定
      I18n.setup({
        lng: i18nPrefix
      }).then(function (i18n) {
        _this.setState({ i18n: i18n.t });
      })['catch'](function (error) {
        console.log(error);
      });

      // Media Query
      var mql = window.matchMedia('(max-width: 767px)');
      mql.addListener(this._handleMediaQueryChange.bind(this));

      this.setState({
        hasScrollbar: true,
        isMobileMode: mql.matches ? true : false,
        isCN: /.cn/.test(window.location.hostname),
        socialMedia: _configsSocialMediaJs2['default'],
        i18n: function i18n(text) {
          return text;
        }
      });
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      // 第一次載入觸發調整一次
      this.setState({ hasScrollbar: Screen.hasScrollbar() });
    }
  }, {
    key: 'render',
    value: function render() {
      var isMobileMode = this.state.isMobileMode;

      return _react2['default'].createElement(
        'footer',
        { style: [styles.footer, !this.state.hasScrollbar && styles.footerFixToBottom] },
        _react2['default'].createElement(
          'div',
          { style: [styles.container, isMobileMode && { flexDirection: 'column', justifyContent: 'space-around' }] },
          _react2['default'].createElement(
            'ul',
            { style: styles.list },
            this.state.socialMedia[this.state.isCN ? '.cn' : '.com'].map(function (entry, index) {
              return _react2['default'].createElement(
                'li',
                { key: index, style: styles.li },
                _react2['default'].createElement(
                  'a',
                  { href: entry.url,
                    target: '_blank' },
                  _react2['default'].createElement('i', { style: styles.icon, className: 'fa ' + entry.icon })
                )
              );
            })
          ),
          _react2['default'].createElement(
            'div',
            { style: styles.rightPart },
            _react2['default'].createElement(
              'div',
              { style: styles.rightPartUpper },
              _react2['default'].createElement(
                'a',
                { style: [styles.legalLink, { marginLeft: 0 }, isMobileMode && { marginBottom: 5, display: 'block' }],
                  href: 'http://www.mediatek.com/' },
                'MediaTek.com'
              ),
              !isMobileMode && '|',
              _react2['default'].createElement(
                'a',
                { href: 'http://www.mediatek.com/en/legal-notice/',
                  style: [styles.legalLink, isMobileMode && { marginLeft: 0 }] },
                this.state.i18n('Legal Notice')
              ),
              '|',
              _react2['default'].createElement(
                'a',
                { style: styles.legalLink,
                  href: 'http://www.mediatek.com/en/privacy-policy/' },
                this.state.i18n('Privacy Policy')
              ),
              '|',
              _react2['default'].createElement(
                'a',
                { style: [styles.legalLink, { marginRight: 0 }],
                  href: 'http://status.mcs.mediatek.com/' },
                this.state.i18n('System Status')
              )
            ),
            _react2['default'].createElement(
              'div',
              { style: [styles.rightPartLower, isMobileMode && { marginTop: 5, marginBottom: 10 }] },
              _react2['default'].createElement(
                'span',
                null,
                '© 2015 '
              ),
              _react2['default'].createElement(
                'span',
                null,
                this.state.i18n('MediaTek Inc All Rights Reserved'),
                this.state.isCN && '京ICP备15035206号'
              )
            )
          )
        )
      );
    }

    /**
    * hasScrollbar
    * 判斷 Footer 如何置底，解決少部分中間內容不夠把 Footer 撐到至底
    * document.readyState complete 避免閃爍。
    *
    * @author Michael Hsu
    */

  }, {
    key: '_handleResize',
    value: function _handleResize(event) {
      if (document.readyState === 'complete') {
        this.setState({ hasScrollbar: Screen.hasScrollbar() });
      }
    }

    /**
    * 處理 media query
    * ref: https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Testing_media_queries
    *
    * @input  {object} mql
    *
    * @author Michael Hsu
    */

  }, {
    key: '_handleMediaQueryChange',
    value: function _handleMediaQueryChange(mql) {
      this.setState({
        isMobileMode: mql.matches ? true : false
      });
    }
  }]);

  return Footer;
})(_react2['default'].Component);

exports['default'] = (0, _radium2['default'])(Footer);

var styles = {
  footer: {
    width: '100%',
    overflow: 'hidden',
    marginBottom: -6,
    minHeight: 160,
    background: '#353630',
    display: 'inline-block',
    fontSize: 13,
    zIndex: 2,
    position: 'relative'
  },
  footerFixToBottom: {
    position: 'absolute',
    left: 0,
    bottom: 0,
    marginBottom: 0
  },
  container: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    boxSizing: 'border-box',
    minHeight: 160,
    maxWidth: 954,
    color: '#fff',
    margin: '0 auto'
  },
  list: {
    listStyleType: 'none',
    marginLeft: 12,
    paddingLeft: 0
  },
  li: {
    display: 'inline-block',
    marginRight: 20
  },
  icon: {
    color: '#fff',
    fontSize: 60
  },
  rightPart: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    minHeight: 50,
    marginRight: 12,
    marginLeft: 12,
    fontFamily: 'Helvetica, sans-serif'
  },
  rightPartUpper: {
    fontSize: 15
  },
  rightPartLower: {
    color: '#D1D2D3',
    fontSize: 13,
    fontWeight: 'bold'
  },
  legalLink: {
    color: '#fff',
    fontSize: 15,
    fontWeight: 'bold',
    textDecoration: 'none',
    outline: 0,
    marginRight: 10,
    marginLeft: 10
  }
};
module.exports = exports['default'];
