// @mtk/mcs-components/lib/Nav.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var Nav = (function (_React$Component) {
  _inherits(Nav, _React$Component);

  function Nav() {
    _classCallCheck(this, Nav);

    _get(Object.getPrototypeOf(Nav.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(Nav, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var listStyle = _props.style;
      var children = _props.children;
      var dropdownMenu = _props.dropdownMenu;

      var props = _objectWithoutProperties(_props, ['style', 'children', 'dropdownMenu']);

      return _react2['default'].createElement(
        'ul',
        _extends({}, props, { style: [dropdownMenu ? styles.dropdownMenu : styles.list, listStyle] }),
        _react2['default'].Children.map(children, this._renderChildren)
      );
    }
  }, {
    key: '_renderChildren',
    value: function _renderChildren(child, index) {
      return _react2['default'].cloneElement(child, { key: index });
    }
  }]);

  return Nav;
})(_react2['default'].Component);

exports['default'] = (0, _radium2['default'])(Nav);

Nav.propTypes = {
  style: _react.PropTypes.object,
  dropdownMenu: _react.PropTypes.bool
};
Nav.defaultProps = {};

var styles = {
  list: {
    display: 'inline-block',
    padding: 0,
    margin: 0
  },
  dropdownMenu: {
    display: 'none',
    position: 'absolute',
    top: '50px',
    minWidth: 100,
    padding: 0,
    zIndex: 1000,
    float: 'left',
    textAlign: 'left',
    listStyle: 'none',
    backgroundColor: '#1995cc',
    borderRadius: '0 0 4px 4px'
  }
};
module.exports = exports['default'];
