// @mtk/mcs-components/lib/NavItem.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _internalClickable = require('./internal/Clickable');

var _internalClickable2 = _interopRequireDefault(_internalClickable);

var NavItem = (function (_Clickable) {
  _inherits(NavItem, _Clickable);

  function NavItem(props) {
    _classCallCheck(this, NavItem);

    _get(Object.getPrototypeOf(NavItem.prototype), 'constructor', this).call(this, props);
  }

  _createClass(NavItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var href = _props.href;
      var children = _props.children;
      var style = _props.style;
      var activeStyle = _props.activeStyle;
      var linkStyle = _props.linkStyle;
      var activeLinkStyle = _props.activeLinkStyle;
      var target = _props.target;

      var props = _objectWithoutProperties(_props, ['href', 'children', 'style', 'activeStyle', 'linkStyle', 'activeLinkStyle', 'target']);

      return _react2['default'].createElement(
        'li',
        _extends({}, props, {
          style: [styles.item, style, this.state.isHover && activeStyle] }),
        _react2['default'].createElement(
          'a',
          { target: this.props.target, href: this.props.href, style: [styles.link, linkStyle, this.state.isHover && activeLinkStyle] },
          children
        )
      );
    }
  }]);

  return NavItem;
})(_internalClickable2['default']);

NavItem.propTypes = {
  href: _react.PropTypes.string.isRequired,
  style: _react.PropTypes.object,
  linkStyle: _react.PropTypes.object
};
NavItem.defaultProps = {
  href: '#'
};

exports['default'] = (0, _radium2['default'])(NavItem);

var styles = {
  item: {
    display: 'inline-block',
    listStyleType: 'none'
  },
  link: {
    textDecoration: 'none',
    outline: 0,
    whiteSpace: 'nowrap'
  }
};
module.exports = exports['default'];
