// @mtk/mcs-components/lib/ParagraphSelector.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _ButtonGroup = require('./ButtonGroup');

var _ButtonGroup2 = _interopRequireDefault(_ButtonGroup);

var _fbjsLibEventListener = require('fbjs/lib/EventListener');

var _fbjsLibEventListener2 = _interopRequireDefault(_fbjsLibEventListener);

var _utilsSectionHelperJs = require('./utils/sectionHelper.js');

var SectionHelper = _interopRequireWildcard(_utilsSectionHelperJs);

var _utilsI18n = require('./utils/i18n');

var I18n = _interopRequireWildcard(_utilsI18n);

var _utilsGetLocaleJs = require('./utils/getLocale.js');

var _utilsGetLocaleJs2 = _interopRequireDefault(_utilsGetLocaleJs);

var ParagraphSelector = (function (_React$Component) {
  _inherits(ParagraphSelector, _React$Component);

  function ParagraphSelector(props) {
    _classCallCheck(this, ParagraphSelector);

    _get(Object.getPrototypeOf(ParagraphSelector.prototype), 'constructor', this).call(this, props);
    this.state = this.state || {};
  }

  _createClass(ParagraphSelector, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this = this;

      // 取得現在網址代表的語言
      var i18nPrefix = (0, _utilsGetLocaleJs2['default'])(window.location.href);

      // i18n 設定
      I18n.setup({
        lng: i18nPrefix
      }).then(function (i18n) {
        _this.setState({ i18n: i18n.t });
      })['catch'](function (error) {
        console.log(error);
      });

      this.setState({
        isOpen: false,
        sectionList: [],
        i18n: function i18n(text) {
          return text;
        }
      });
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      var scopeClassName = this.props.scopeClassName;

      this.setState({
        sectionList: SectionHelper.getDynamicSection(scopeClassName)
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props;
      var customizedStyle = _props.style;
      var offset = _props.offset;

      var otherProps = _objectWithoutProperties(_props, ['style', 'offset']);

      return _react2['default'].createElement(
        'div',
        _extends({}, otherProps, { style: customizedStyle }),
        _react2['default'].createElement(_ButtonGroup2['default'], {
          kind: 'primary',
          leftTitle: this.state.i18n('Select the paragraph'),
          rightTitle: _react2['default'].createElement('i', { className: 'fa fa-angle-down' }),
          onClick: function (e) {
            return _this2._handleClick(e);
          } }),
        _react2['default'].createElement(
          'ul',
          { style: [styles.dropdownMenu, this.state.isOpen ? { display: 'block' } : { display: 'none' }] },
          this.state.sectionList.map(function (entry, index) {
            return _react2['default'].createElement(
              'li',
              { key: index,
                style: styles.dropdownList,
                onClick: function (e) {
                  return SectionHelper.scrollTo({ index: index, element: entry.element, offset: offset });
                } },
              entry.innerText
            );
          })
        )
      );
    }
  }, {
    key: '_handleClick',
    value: function _handleClick(e) {
      e.preventDefault();
      if (this.state.isOpen) {
        return this._hide();
      }

      if (this._onDocumentClickListener) {
        this._onDocumentClickListener.remove();
      }

      this.setState({ isOpen: true });
      this._onDocumentClickListener = _fbjsLibEventListener2['default'].listen(document, 'click', this._hide.bind(this));
      this._onDocumentKeyupListener = _fbjsLibEventListener2['default'].listen(document, 'keyup', this._hide.bind(this));
    }
  }, {
    key: '_hide',
    value: function _hide() {
      this.setState({ isOpen: false });

      if (this._onDocumentClickListener) {
        this._onDocumentClickListener.remove();
      }

      if (this._onDocumentKeyupListener) {
        this._onDocumentKeyupListener.remove();
      }
    }
  }]);

  return ParagraphSelector;
})(_react2['default'].Component);

exports['default'] = (0, _radium2['default'])(ParagraphSelector);

ParagraphSelector.propTypes = {
  scopeClassName: _react2['default'].PropTypes.string,
  sensitive: _react2['default'].PropTypes.number,
  offset: _react2['default'].PropTypes.number
};
ParagraphSelector.defaultProps = {
  scopeClassName: 'resource',
  sensitive: 5, // 觸發敏感度微調。
  offset: 0 // 保留 Header 高度
};

var styles = {
  dropdownMenu: {
    position: 'absolute',
    border: '1px solid rgba(0,0,0,.15)',
    borderRadius: 4,
    boxShadow: '0 6px 12px rgba(0,0,0,.175)',
    backgroundClip: 'padding-box',
    backgroundColor: '#fff',
    textAlign: 'left',
    listStyle: 'none',
    minWidth: 160,
    padding: '10px 0px',
    margin: '2px 0 0',
    right: 0
  },
  dropdownList: {
    lineHeight: '20px',
    fontSize: 14,
    padding: '3px 20px',
    whiteSpace: 'nowrap',
    cursor: 'pointer',
    ':hover': {
      color: '#fafafa',
      backgroundColor: '#00A1DE'
    }
  },
  dropdownListActive: {
    color: '#fafafa',
    backgroundColor: '#00A1DE'
  }
};
module.exports = exports['default'];
