// @mtk/mcs-components/lib/SideBarMenu.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _utilsSectionHelperJs = require('./utils/sectionHelper.js');

var SectionHelper = _interopRequireWildcard(_utilsSectionHelperJs);

var SideBarMenu = (function (_React$Component) {
  _inherits(SideBarMenu, _React$Component);

  function SideBarMenu(props) {
    _classCallCheck(this, SideBarMenu);

    _get(Object.getPrototypeOf(SideBarMenu.prototype), 'constructor', this).call(this, props);
    this.state = this.state || {};
    this.state.sectionList = [];
    this.state.liActiveIndex = 0;
  }

  _createClass(SideBarMenu, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this = this;

      var scopeClassName = this.props.scopeClassName;
      var sensitive = this.props.sensitive;
      var offset = this.props.offset;

      // Menu li 的 Active 狀態
      SectionHelper.getSectionIntervalsAsync(scopeClassName).then(function (sectionIntervals) {
        _this._setLiActiveIndex({ sectionIntervals: sectionIntervals, sensitive: sensitive, offset: offset });
        window.addEventListener('scroll', _this._setLiActiveIndex.bind(_this, { sectionIntervals: sectionIntervals, sensitive: sensitive, offset: offset }));
      })['catch'](function (error) {
        console.log(error);
      });

      this.setState({
        sectionList: SectionHelper.getDynamicSection(scopeClassName)
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props;
      var offset = _props.offset;

      var otherProps = _objectWithoutProperties(_props, ['offset']);

      return _react2['default'].createElement(
        'div',
        _extends({}, otherProps, { style: styles.container }),
        _react2['default'].createElement(
          'div',
          { style: styles.sectionGroup },
          _react2['default'].createElement(
            'div',
            { style: styles.overview },
            'Overview'
          ),
          _react2['default'].createElement(
            'ul',
            { style: styles.ul },
            this.state.sectionList.map(function (entry, index) {
              return _react2['default'].createElement(
                'li',
                { key: index,
                  style: [styles.li, _this2.state.liActiveIndex === index ? styles.liActive : {}],
                  onClick: function (e) {
                    return SectionHelper.scrollTo({ index: index, element: entry.element, offset: offset });
                  } },
                entry.innerText
              );
            })
          )
        )
      );
    }

    /**
    * setState 目前 Active li 的 index
    *
    * @author Michael Hsu
    */

  }, {
    key: '_setLiActiveIndex',
    value: function _setLiActiveIndex(_ref) {
      var sectionIntervals = _ref.sectionIntervals;
      var sensitive = _ref.sensitive;
      var offset = _ref.offset;

      this.setState({
        liActiveIndex: SectionHelper.getLiActiveIndex({ sectionIntervals: sectionIntervals, sensitive: sensitive, offset: offset })
      });
    }
  }]);

  return SideBarMenu;
})(_react2['default'].Component);

exports['default'] = (0, _radium2['default'])(SideBarMenu);

SideBarMenu.propTypes = {
  scopeClassName: _react2['default'].PropTypes.string,
  sensitive: _react2['default'].PropTypes.number,
  offset: _react2['default'].PropTypes.number
};
SideBarMenu.defaultProps = {
  scopeClassName: 'resource',
  sensitive: 5, // 觸發敏感度微調。
  offset: 0 // 保留 Header 高度
};

var styles = {
  container: {
    position: 'fixed',
    height: '100%',
    backgroundColor: '#FAFAFA',
    marginTop: '50px',
    boxShadow: '-1px 0px 0px 0px #F1F2F7',
    width: '200px'
  },
  sectionGroup: {
    marginTop: 79
  },
  overview: {
    fontSize: 18,
    color: '#353630',
    lineHeight: '23px',
    marginLeft: 20,
    marginBottom: 10
  },
  ul: {
    listStyleType: 'none',
    paddingLeft: 0,
    fontSize: 13,
    color: '#353630',
    lineHeight: '16px',
    marginTop: 0
  },
  li: {
    paddingTop: 8,
    paddingBottom: 8,
    paddingLeft: 20,
    cursor: 'pointer',
    ':hover': {
      color: '#00A1DE'
    }
  },
  liActive: {
    background: '#FFFFFF',
    outline: '1px solid #F1F2F7',
    color: '#00A1DE',
    borderLeft: '3px solid #00A1DE',
    paddingLeft: 17
  }
};
module.exports = exports['default'];
