// @mtk/mcs-components/lib/datachannels/string/Display.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _recompose = require('recompose');

var _mtkUiLibInputTextarea = require('mtk-ui/lib/InputTextarea');

var _mtkUiLibInputTextarea2 = _interopRequireDefault(_mtkUiLibInputTextarea);

var _commonCardStylesBaseCss = require('../common/card/styles/base.css');

var _commonCardStylesBaseCss2 = _interopRequireDefault(_commonCardStylesBaseCss);

var _stylesMainCss = require('./styles/main.css');

var _stylesMainCss2 = _interopRequireDefault(_stylesMainCss);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var display = function display(_ref) {
  var status = _ref.status;
  var displayType = _ref.displayType;
  var placeholder = _ref.placeholder;

  return _react2['default'].createElement(
    'div',
    { className: (0, _classnames2['default'])(_commonCardStylesBaseCss2['default'].base, _commonCardStylesBaseCss2['default'][displayType], _stylesMainCss2['default'].base) },
    _react2['default'].createElement(_mtkUiLibInputTextarea2['default'], {
      rows: '4',
      placeholder: placeholder,
      value: status.value,
      disabled: true,
      className: _stylesMainCss2['default'].displayTextarea
    })
  );
};

var propTypes = {
  status: _react.PropTypes.object,
  channelType: _react.PropTypes.oneOf(['String', 'Hex']),
  unit: _react.PropTypes.string,
  displayType: _react.PropTypes.oneOf(['none', 'preview', 'embeded'])
};

exports.propTypes = propTypes;
exports['default'] = (0, _recompose.compose)((0, _recompose.defaultProps)({
  status: { value: '' },
  channelType: 'String',
  displayType: 'none'
}), (0, _recompose.setPropTypes)(propTypes), _recompose.pure, (0, _recompose.withAttachedProps)(function (getProps) {
  return {
    placeholder: getProps().channelType === 'Hex' ? 'This place holds Hex value.' : 'This place holds String value.'
  };
}))(display);
